<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">

<!--Converted with LaTeX2HTML 2K.1beta (1.47)
original version by:  Nikos Drakos, CBLU, University of Leeds
* revised and updated by:  Marcus Hennecke, Ross Moore, Herb Swan
* with significant contributions from:
  Jens Lippmann, Marek Rouchal, Martin Wilck and others -->
<HTML>
<HEAD>
<TITLE>5 File types</TITLE>
<META NAME="description" CONTENT="5 File types">
<META NAME="keywords" CONTENT="sam_doc">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">

<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META NAME="Generator" CONTENT="LaTeX2HTML v2K.1beta">
<META HTTP-EQUIV="Content-Style-Type" CONTENT="text/css">

<LINK REL="STYLESHEET" HREF="sam_doc.css">

<LINK REL="next" HREF="node8.html">
<LINK REL="previous" HREF="node6.html">
<LINK REL="up" HREF="sam_doc.html">
<LINK REL="next" HREF="node8.html">
</HEAD>

<BODY >
<!--Navigation Panel-->
<A NAME="tex2html367"
  HREF="node8.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<A NAME="tex2html363"
  HREF="sam_doc.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up.png"></A> 
<A NAME="tex2html357"
  HREF="node6.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev.png"></A> 
<A NAME="tex2html365"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html368"
  HREF="node8.html">6 Parameter specification</A>
<B> Up:</B> <A NAME="tex2html364"
  HREF="sam_doc.html">SAM (Sequence Alignment and</A>
<B> Previous:</B> <A NAME="tex2html358"
  HREF="node6.html">Bibliography</A>
 &nbsp <B>  <A NAME="tex2html366"
  HREF="node1.html">Contents</A></B> 
<BR>
<BR>
<!--End of Navigation Panel-->

<H1><A NAME="SECTION00070000000000000000"></A><A NAME="sec:files"></A>
<BR>
5 File types
</H1>

<P>
SAM produces a variety of files.  This section summarizes both the
extensions SAM uses and the extensions we conventionally use when
naming output that goes to standard output.

<P>
The following file extensions are automatically used:
<DL>
<DT><STRONG>.a2m</STRONG></DT>
<DD>A SAM alignment file, as created by <TT>align2model</TT> or
<TT>hmmscore</TT>.   A FASTA-compatible format.
See Section&nbsp;<A HREF="node12.html#sec:align2model">10.1</A>.
</DD>
<DT><STRONG>.apr</STRONG></DT>
<DD>A SAM alignment probabilities file, as created by <TT>pathprobs</TT>.  An RDB-compatible format.  See Section&nbsp;<A HREF="node12.html#sec:pathprobs">10.7</A>.
</DD>
<DT><STRONG>.cst</STRONG></DT>
<DD>A SAM constraints file containing the definition of
sequence position constraints.
See Section&nbsp;<A HREF="node11.html#sec:constraints">9.6</A>.
</DD>
<DT><STRONG>.data</STRONG></DT>
<DD>A <TT>gnuplot</TT> data file generated by <TT>makehist</TT>.
See Section&nbsp;<A HREF="node12.html#sec:plottingprogs">10.10</A>.
</DD>
<DT><STRONG>.dist</STRONG></DT>
<DD>A scoring file listing sequence identifiers, lengths, and
scores, produced by <TT>hmmscore</TT>. See Section&nbsp;<A HREF="node12.html#sec:hmmscore">10.2</A>.
</DD>
<DT><STRONG>.dist-rdb</STRONG></DT>
<DD>A scoring file in RDB format including sequence
identifiers, lengths, scores, and information about scoring method
produced by <TT>hmmscore</TT>. See Section&nbsp;<A HREF="node12.html#sec:hmmscore">10.2</A>. 
</DD>
<DT><STRONG>.freq</STRONG></DT>
<DD>A frequency model generated by <TT>buildmodel</TT> when
<TT>many_files</TT> is set.  This is
a model that reports the frequency of each letter in each state when 
the training set is evaluated according to the model. See Section&nbsp;<A HREF="node11.html#sec:trainstat">9.3</A>.
</DD>
<DT><STRONG>.hmmer</STRONG></DT>
<DD>An HMMER-format model produced by <TT>sam2hmmer</TT>.
See Section&nbsp;<A HREF="node12.html#sec:convert">10.9.5</A>.
</DD>
<DT><STRONG>.kids</STRONG></DT>
<DD>A Kestrel sequence id file.  The file is an ordered
list of identifiers for sequences in a <TT>.kseq</TT> or <TT>.krseq</TT> file.
See Section&nbsp;<A HREF="node12.html#sec:mkkestreldb">10.11.9</A>.
</DD>
<DT><STRONG>.kseqs</STRONG></DT>
<DD>A Kestrel sequence database, containing an encoding
of the sequences optimized for Kestrel <TT>hmmscore</TT>.
See Section&nbsp;<A HREF="node12.html#sec:mkkestreldb">10.11.9</A>.
</DD>
<DT><STRONG>.kseqs</STRONG></DT>
<DD>A Kestrel reverse null model scoring sequence database.
This contains each sequence in the original database, plus a reverse
of the sequence.
See Section&nbsp;<A HREF="node12.html#sec:mkkestreldb">10.11.9</A>.
</DD>
<DT><STRONG>.match-rdb</STRONG></DT>
<DD>An RDB file created by <TT>grabdp</TT>, listing the posterior
match probabilities of each amino acid type at each node, plus the prior
probabilities of each amino acid type.  Used for diagnostic purposes.
See Section&nbsp;<A HREF="node12.html#sec:grabdp">10.4</A>.
</DD>
<DT><STRONG>.mod</STRONG></DT>
<DD>A model file created by a <TT>buildmodel</TT>, <TT>addfims</TT>, or <TT>modelfromalign</TT>.  When <TT>many_files</TT> is not set
(the default), <TT>buildmodel</TT> places the run statistics, parameter
settings,  and the
frequency model into this file as well. See Section&nbsp;<A HREF="node11.html#sec:trainstat">9.3</A>.
<DL>
<DT><STRONG>.f.mod</STRONG></DT>
<DD>A subfamily model for family number `f'.  See Section&nbsp;<A HREF="node11.html#sec:weights">9.4</A>.
</DD>
<DT><STRONG>.a.mrrr.mod</STRONG></DT>
<DD>If <TT>print_all_models</TT> is set, model number <IMG
 WIDTH="20" HEIGHT="17" ALIGN="BOTTOM" BORDER="0"
 SRC="img33.png"
 ALT="$m$">
during re-estimate cycle 'rrr' will be placed in the file.
</DD>
<DT><STRONG>.s.rrr.mod</STRONG></DT>
<DD>If <TT>print_surg_models</TT> is set, the winning 
model is printed  after each surgery iteration, where `rrr' is the
re-estimation index.
</DD>
</DL>
</DD>
<DT><STRONG>.mult</STRONG></DT>
<DD>One of two multiple domain alignment output files created
by <TT>hmmscore</TT>.  This will contain
all alignments to a motif that were found. See Section&nbsp;<A HREF="node12.html#sec:multdomain">10.2.4</A>.
</DD>
<DT><STRONG>.mlib</STRONG></DT>
<DD>A model library file created by <TT>hmmscore</TT> model
calibration.  Each entry includes a single- or multi-track model
specification and various scoring parameters. See Section&nbsp;<A HREF="node12.html#sec:calibrate">10.2.9</A>.
</DD>
<DT><STRONG>.mstat</STRONG></DT>
<DD>The other multiple domain output file.  This file contains
the sequence identifier and scores for the data in the corresponding
<TT>.mult</TT> file. See Section&nbsp;<A HREF="node12.html#sec:multdomain">10.2.4</A>.
</DD>
<DT><STRONG>.pa2m</STRONG></DT>
<DD>A SAM alignment file that alternates sequences and
posterior probability annotations in sequence format, as created by
<TT>pathprobs</TT>.
See Section&nbsp;<A HREF="node12.html#sec:pathprobs">10.7</A>.
</DD>
<DT><STRONG>.pdoc</STRONG></DT>
<DD>Created by <TT>grabdp</TT>, this file contains the posterior
decoding of the dynamic programming for match states and transitions
into match states.  The files tend to be rather large. See Section&nbsp;<A HREF="node12.html#sec:grabdp">10.4</A>.
</DD>
<DT><STRONG>.plt</STRONG></DT>
<DD>A <TT>gnuplot</TT> command file created by <TT>makehist</TT>. See Section&nbsp;<A HREF="node12.html#sec:plottingprogs">10.10</A>.
</DD>
<DT><STRONG>.ptrack</STRONG></DT>
<DD>Created by <TT>predict_track</TT>, this file contains
secondary structure (or other second alphabet) predictions for a given
sequence based on a primary sequence model, and databases of primary
and secondary structure for a group of sequences.
See Section&nbsp;<A HREF="node12.html#sec:predicttrack">10.8</A>.
</DD>
<DT><STRONG>.samrc</STRONG></DT>
<DD>A file of default parameters either in a home directory
or the current directory. See Section&nbsp;<A HREF="node8.html#params">6</A>.
</DD>
<DT><STRONG>.sel</STRONG></DT>
<DD>Sequences that passed the selection criteria used in <TT>hmmscore</TT>. See Section&nbsp;<A HREF="node12.html#sec:hmmscore">10.2</A>.
</DD>
<DT><STRONG>.seq</STRONG></DT>
<DD>A file of sequences, as for example created by <TT>sampleseqs</TT>.
</DD>
<DT><STRONG>.stat</STRONG></DT>
<DD>The statistics of a <TT>buildmodel</TT> run when <TT>many_files</TT> is set, including re-estimation scores and initial
parameters. See Section&nbsp;<A HREF="node11.html#sec:trainstat">9.3</A>. 
</DD>
<DT><STRONG>.weightoutput</STRONG></DT>
<DD>A list of sequence weights from the internal
weighting option of <TT>buildmodel</TT>.  Present when <TT>print_all_weights</TT> and internal weighting are both enabled.
See Section&nbsp;<A HREF="node11.html#sec:internalweight">9.4.4</A>. 
</DD>
</DL>

<P>
The following file extensions are conventionally used in this manual.
<DL>
<DT><STRONG>.colors</STRONG></DT>
<DD>A color file for the <TT>makelogo</TT> program, usually
located in the same directory as the Dirichlet priors.
See Section&nbsp;<A HREF="node12.html#sec:makelogo">10.9.3</A>. 
</DD>
<DT><STRONG>.log</STRONG></DT>
<DD>Standard error output from running <TT>target99</TT> can be
redirected to this file.  See Section&nbsp;<A HREF="node5.html#sec:samt99">4</A>.
</DD>
<DT><STRONG>.ncomp</STRONG></DT>
<DD>A Dirichlet mixture prior library file, where n is the
number of components to the mixture, as in <TT>null.1comp</TT> or <TT>mall-opt.9comp</TT>.  Information on specific mixtures can be
found at <A NAME="tex2html46"
  HREF="http://www.cse.ucsc.edu/research/compbio/dirichlets/index.html"><TT>http://www.cse.ucsc.edu/research/compbio/dirichlets/index.html</TT></A>.
See Section&nbsp;<A HREF="node10.html#sec:regularizer">8.1</A>.
</DD>
<DT><STRONG>.plib</STRONG></DT>
<DD>A Dirichlet mixture prior library file (old naming
convention).
The same extension is used both for match state regularizers and
HSSP-based transition regularizers.
See Section&nbsp;<A HREF="node10.html#sec:regularizer">8.1</A>.
</DD>
<DT><STRONG>.pretty</STRONG></DT>
<DD>The output (stdout) from <TT>prettyalign</TT>.
See Section&nbsp;<A HREF="node12.html#sec:align2model">10.1</A>.
</DD>
<DT><STRONG>.regularizer</STRONG></DT>
<DD>A transition regularizer, generally used in
conjunction with a Dirichlet mixture regularizer for the match
states.  See Section&nbsp;<A HREF="node10.html#sec:regularizer">8.1</A>.
</DD>
<DT><STRONG>.weights</STRONG></DT>
<DD>A file of sequence weights. See Section&nbsp;<A HREF="node11.html#sec:weights">9.4</A>.
</DD>
</DL>

<P>
In most cases, SAM can read compressed (.gz or .Z) files, specified
either as their complete name or as their root name without the
compression suffix.  If the root name is given and both a compressed
and an uncompressed file exist, the uncompressed file is read.  If
both a .gz and a .Z file is present, the .gz file is used.  Reading
compressed files requires that the GNU decompression program <TT>gunzip</TT> and the Unix decompression program <TT>uncompress</TT> be in the
user's path.

<P>
<HR>
<!--Navigation Panel-->
<A NAME="tex2html367"
  HREF="node8.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<A NAME="tex2html363"
  HREF="sam_doc.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up.png"></A> 
<A NAME="tex2html357"
  HREF="node6.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev.png"></A> 
<A NAME="tex2html365"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html368"
  HREF="node8.html">6 Parameter specification</A>
<B> Up:</B> <A NAME="tex2html364"
  HREF="sam_doc.html">SAM (Sequence Alignment and</A>
<B> Previous:</B> <A NAME="tex2html358"
  HREF="node6.html">Bibliography</A>
 &nbsp <B>  <A NAME="tex2html366"
  HREF="node1.html">Contents</A></B> 
<!--End of Navigation Panel-->
<ADDRESS>
<a href=http://www.cse.ucsc.edu/research/compbio/sam.html>SAM</a><BR><a href=mailto:sam-info@cse.ucsc.edu>sam-info@cse.ucsc.edu</a><BR><A HREF=http://www.cse.ucsc.edu/research/compbio>UCSC Computational Biology Group</A>
</ADDRESS>
</BODY>
</HTML>
