<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">

<!--Converted with LaTeX2HTML 2K.1beta (1.47)
original version by:  Nikos Drakos, CBLU, University of Leeds
* revised and updated by:  Marcus Hennecke, Ross Moore, Herb Swan
* with significant contributions from:
  Jens Lippmann, Marek Rouchal, Martin Wilck and others -->
<HTML>
<HEAD>
<TITLE>10 Related programs</TITLE>
<META NAME="description" CONTENT="10 Related programs">
<META NAME="keywords" CONTENT="sam_doc">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">

<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META NAME="Generator" CONTENT="LaTeX2HTML v2K.1beta">
<META HTTP-EQUIV="Content-Style-Type" CONTENT="text/css">

<LINK REL="STYLESHEET" HREF="sam_doc.css">

<LINK REL="next" HREF="node13.html">
<LINK REL="previous" HREF="node11.html">
<LINK REL="up" HREF="sam_doc.html">
<LINK REL="next" HREF="node13.html">
</HEAD>

<BODY >
<!--Navigation Panel-->
<A NAME="tex2html457"
  HREF="node13.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<A NAME="tex2html453"
  HREF="sam_doc.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up.png"></A> 
<A NAME="tex2html447"
  HREF="node11.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev.png"></A> 
<A NAME="tex2html455"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html458"
  HREF="node13.html">11 System installation</A>
<B> Up:</B> <A NAME="tex2html454"
  HREF="sam_doc.html">SAM (Sequence Alignment and</A>
<B> Previous:</B> <A NAME="tex2html448"
  HREF="node11.html">9 The buildmodel estimation</A>
 &nbsp <B>  <A NAME="tex2html456"
  HREF="node1.html">Contents</A></B> 
<BR>
<BR>
<!--End of Navigation Panel-->
<!--Table of Child-Links-->
<A NAME="CHILD_LINKS"><STRONG>Subsections</STRONG></A>

<UL>
<LI><A NAME="tex2html459"
  HREF="#SECTION000121000000000000000">10.1  <TT>align2model</TT> and <TT>prettyalign</TT></A>
<UL>
<LI><A NAME="tex2html460"
  HREF="#SECTION000121100000000000000">10.1.1 <TT>prettyalign</TT></A>
<LI><A NAME="tex2html461"
  HREF="#SECTION000121200000000000000">10.1.2 Aligning fragments</A>
</UL>
<BR>
<LI><A NAME="tex2html462"
  HREF="#SECTION000122000000000000000">10.2 <TT>hmmscore</TT></A>
<UL>
<LI><A NAME="tex2html463"
  HREF="#SECTION000122100000000000000">10.2.1 NLL-NULL scoring</A>
<LI><A NAME="tex2html464"
  HREF="#SECTION000122200000000000000">10.2.2 Selecting sequences, scores, and alignments</A>
<LI><A NAME="tex2html465"
  HREF="#SECTION000122300000000000000">10.2.3 Scoring Fragments</A>
<LI><A NAME="tex2html466"
  HREF="#SECTION000122400000000000000">10.2.4 Selecting multiple domain alignments</A>
<LI><A NAME="tex2html467"
  HREF="#SECTION000122500000000000000">10.2.5 Multi-track HMM scoring</A>
<LI><A NAME="tex2html468"
  HREF="#SECTION000122600000000000000">10.2.6 Distributed scoring</A>
<LI><A NAME="tex2html469"
  HREF="#SECTION000122700000000000000">10.2.7 Smith &amp; Waterman Scoring and
Alignment</A>
<LI><A NAME="tex2html470"
  HREF="#SECTION000122800000000000000">10.2.8 Scoring using the Kestrel parallel processor</A>
<LI><A NAME="tex2html471"
  HREF="#SECTION000122900000000000000">10.2.9 Calibration and Model libraries</A>
</UL>
<BR>
<LI><A NAME="tex2html472"
  HREF="#SECTION000123000000000000000">10.3 <TT>addfims</TT></A>
<UL>
<LI><A NAME="tex2html473"
  HREF="#SECTION000123100000000000000">10.3.1 Training with FIMs</A>
</UL>
<BR>
<LI><A NAME="tex2html474"
  HREF="#SECTION000124000000000000000">10.4 <TT>grabdp</TT></A>
<LI><A NAME="tex2html475"
  HREF="#SECTION000125000000000000000">10.5 <TT>get_fisher_scores</TT></A>
<LI><A NAME="tex2html476"
  HREF="#SECTION000126000000000000000">10.6 <TT>modelfromalign</TT></A>
<LI><A NAME="tex2html477"
  HREF="#SECTION000127000000000000000">10.7 <TT>pathprobs</TT></A>
<LI><A NAME="tex2html478"
  HREF="#SECTION000128000000000000000">10.8 <TT>predict_track</TT></A>
<LI><A NAME="tex2html479"
  HREF="#SECTION000129000000000000000">10.9 Model manipulation</A>
<UL>
<LI><A NAME="tex2html480"
  HREF="#SECTION000129100000000000000">10.9.1 <TT>drawmodel</TT></A>
<LI><A NAME="tex2html481"
  HREF="#SECTION000129200000000000000">10.9.2 <TT>hmmconvert</TT></A>
<LI><A NAME="tex2html482"
  HREF="#SECTION000129300000000000000">10.9.3 <TT>makelogo</TT></A>
<LI><A NAME="tex2html483"
  HREF="#SECTION000129400000000000000">10.9.4 <TT>modifymodel</TT></A>
<LI><A NAME="tex2html484"
  HREF="#SECTION000129500000000000000">10.9.5 Conversion between SAM and HMMer</A>
<UL>
<LI><A NAME="tex2html485"
  HREF="#SECTION000129510000000000000">10.9.5.1 Internal HMM Structure</A>
<LI><A NAME="tex2html486"
  HREF="#SECTION000129520000000000000">10.9.5.2 Extra Information in HMMer</A>
<LI><A NAME="tex2html487"
  HREF="#SECTION000129530000000000000">10.9.5.3 Extra Information in SAM</A>
</UL>
</UL>
<BR>
<LI><A NAME="tex2html488"
  HREF="#SECTION0001210000000000000000">10.10 Plotting Programs</A>
<UL>
<LI><A NAME="tex2html489"
  HREF="#SECTION0001210100000000000000">10.10.1 <TT>makehist</TT></A>
<LI><A NAME="tex2html490"
  HREF="#SECTION0001210200000000000000">10.10.2 <TT>makeroc</TT></A>
<LI><A NAME="tex2html491"
  HREF="#SECTION0001210300000000000000">10.10.3 <TT>makeroc2</TT></A>
<LI><A NAME="tex2html492"
  HREF="#SECTION0001210400000000000000">10.10.4 <TT>makeroc3</TT></A>
</UL>
<BR>
<LI><A NAME="tex2html493"
  HREF="#SECTION0001211000000000000000">10.11 Sequence manipulation</A>
<UL>
<LI><A NAME="tex2html494"
  HREF="#SECTION0001211100000000000000">10.11.1 <TT>checkseq</TT></A>
<LI><A NAME="tex2html495"
  HREF="#SECTION0001211200000000000000">10.11.2 <TT>genseq</TT></A>
<LI><A NAME="tex2html496"
  HREF="#SECTION0001211300000000000000">10.11.3 <TT>permuteseq</TT></A>
<LI><A NAME="tex2html497"
  HREF="#SECTION0001211400000000000000">10.11.4 <TT>randseq</TT></A>
<LI><A NAME="tex2html498"
  HREF="#SECTION0001211500000000000000">10.11.5 <TT>splitseq</TT></A>
<LI><A NAME="tex2html499"
  HREF="#SECTION0001211600000000000000">10.11.6 <TT>sampleseqs</TT></A>
<LI><A NAME="tex2html500"
  HREF="#SECTION0001211700000000000000">10.11.7 <TT>sortseq</TT></A>
<LI><A NAME="tex2html501"
  HREF="#SECTION0001211800000000000000">10.11.8 <TT>uniqueseq</TT></A>
<LI><A NAME="tex2html502"
  HREF="#SECTION0001211900000000000000">10.11.9 <TT>mk_kestrel_db</TT></A>
</UL></UL>
<!--End of Table of Child-Links-->
<HR>

<H1><A NAME="SECTION000120000000000000000">
10 Related programs</A>
</H1>

<P>

<H2><A NAME="SECTION000121000000000000000"></A><A NAME="sec:align2model"></A>
<BR>
10.1  <TT>align2model</TT> and <TT>prettyalign</TT>
</H2>
After you have obtained a model of your sequence family, <TT>align2model</TT> can be used to give a multiple alignment of sequences.
Often one is just interested in aligning the sequences that were also
used to train the model, but in principle any sequence can be included
in the alignment.

<P>
The multiple alignment is obtained by aligning each of the sequences
to the model by the Viterbi algorithm.  This has the advantage that
it can be done for each sequence independently, and therefore it is
very simple to add new sequences to the multiple alignment.
Also, once the model is found, the multiple alignment is very
fast and easy to produce.

<P>
The program to produce the basic alignment is called <TT>align2model</TT>.
Calling it with no arguments gives a brief explanation.
To align all the sequences in <TT>trna10.seq</TT> use the command:
<PRE>

align2model trna10 -i test.mod -db trna10.seq
</PRE>
This will put an intermediate form of the alignment in the file
<TT>trna10.a2m</TT>.  In this FASTA-compatible intermediate format
deletions are shown as  dashes (`-') and insertions (produced in the
insert states of the model) are shown as lower case characters, while
periods (`.') are used to fill in the sequences that did not have any
insertions if  <TT>a2mdots</TT> is set to 1, the default.

<P>
In case you only want to align a few sequences in a large file,
you can specify the identifiers of these sequences on the command
line.  For instance
<PRE>

align2model trna2 -i test.mod -db trna10.seq -id TRNA1 -id TRNA9
</PRE>
will only align the two specified sequences.  The output (in <TT>trna2.a2m</TT>) would look like this: 
<PRE>

&gt;TRNA1
GGGGAUGUAGCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCC
CcG.GGUUCGAUCCCC--GGCAUCUCCA
&gt;TRNA9
CGGCACGUAGCGCAGCCUGG-UAGCGCACCGUCCUGGGGUUGCGGGGGUC
G.GaGGUUCAAAUCCUCUCGUGCCGACCA
</PRE>

<P>
The Viterbi algorithm finds the single best path for the alignment --
that which has the highest probability.  The SAM system can also find
the single most <EM>likely</EM> path.  Internally, this corresponds to
using the EM algorithm to evaluate all possible paths, calculating the
probability of each residue appear in each model state and the
probability of use for each transition (the posterior probabilities).
A second (Viterbi-style) dynamic programming is then performed on
these probabilities to find the most likely path (see, for example,
Holmes and Durbin in Recomb98).

<P>
SAM includes two flavors of posterior-decoded alignment.  When <TT>adpstyle</TT> is 4, posteriors are calculated using a <EM>global</EM>
(<TT>SW</TT> is 0) forward calculation.  Then, the minimizing path
through the transition <EM>and</EM> character emission posteriors is
calculated using the specified <TT>SW</TT> mode.

<P>
When <TT>adpstyle</TT> is 5, posteriors for character emissions are
calculated using the specified global or local <TT>SW</TT> mode.  Then,
the minimizing path for character emission posteriors <EM>only</EM> is
calculated using a global dynamic programming mode.

<P>
These alignment methods are much slower than standard Viterbi
alignment, and have not yet been implemented for reduced memory use.
If the sequence is too long to perform this extensive alignment
calculation, a warning messages printed and the Viterbi algorithms
used.

<P>

<H3><A NAME="SECTION000121100000000000000">
10.1.1 <TT>prettyalign</TT></A>
</H3>

<P>
To get a nice display of the alignment 
produced by <TT>align2model</TT>,
you can use the <TT>prettyalign</TT> program, which has several display
options.  The program reads from a file like the one
made in the example above:
<PRE>

prettyalign trna10.a2m &gt; trna10.pretty
</PRE>
which would give you an alignment similar to the one shown in
the Section <A HREF="node4.html#quick">3</A>

<P>
<TT>Prettyalign</TT> does not follow SAM's
normal commandline format.
To see an explanation of the various options, run the program with
some invalid option (like <TT>prettyalign -h</TT>).  Some of the most
useful options are
<DL>
<DT><STRONG>-f</STRONG></DT>
<DD>Print in a FASTA-like format.
</DD>
<DT><STRONG>-i</STRONG></DT>
<DD>Do not include sequence identifiers in front of each line.
</DD>
<DT><STRONG>-l <EM>num</EM></STRONG></DT>
<DD>Set the output line length equal to <EM>num</EM>.
</DD>
<DT><STRONG>-n</STRONG></DT>
<DD>Toggle indexing the sequences, as well as labeling them.
</DD>
<DT><STRONG>-c</STRONG></DT>
<DD>Toggle column numbering.
</DD>
<DT><STRONG>-m</STRONG></DT>
<DD>Set maximum insertion length (longer insertions are printed
as their length).
</DD>
<DT><STRONG>-I</STRONG></DT>
<DD>I-G style alignment.  Also sets maximum insertion length
very high.
</DD>
<DT><STRONG>-b</STRONG></DT>
<DD>Generate BLAST-style alignments.
</DD>
</DL>
The commands
<PRE>

align2model  trna3 -i test.mod -db trna10.seq -id TRNA1 -id TRNA2 -id TRNA9
prettyalign trna3.a2m -l 50 &gt; trna3.pretty
</PRE>
gives the following output

<P>
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;GGGGAUGUAGCUCAG-UGGUA.GAGCGCA-UGCUUCGCAUGUA
TRNA2&nbsp;GCGGCCGUCGUCUAGUCUGGAuUAGGACGCUGGCCUCCCAAGCC
TRNA9&nbsp;CGGCACGUAGCGCAGCCUGG-.UAGCGCACCGUCCUGGGGUUGC

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;UGAGGCCCcG.GGUUCGAUCCCC--GGCAUCUCCA
TRNA2&nbsp;AGCAAUCCcG.GGUUCGAAUCCCGGCGGCC--GCA
TRNA9&nbsp;GGGGGUCG.GaGGUUCAAAUCCUCUCGUGCCGACCA

<P>
</PRE>

<P>
The <TT>prettyalign</TT> program can compress long insertions to only the
initial segment of bases in the insertion plus digits representing the
<EM>total</EM> length of the insertion.  For example, the sequence <TT>GacguacguG</TT> could be printed out as <TT>Ga8guG</TT> if 4 was the largest
number of insertions that was to be allowed (note that the character 8
is using up one of the positions).  By default, insertions of up to
length ten thousand are fully printed.  This can be changed with the
<code>-m</code> flag to <TT>prettyalign</TT>, which sets the maximum number of
insertions that are printed.  If set to zero, no insertions are
printed, and no indication of the lack is given.  If less than zero,
insertion characters are not printed, and that number of digits is
used to indicate the length of each insertion.

<P>
For example, the alignment at the end of Section&nbsp;<A HREF="node12.html#sec:fimtrain">10.3.1</A>
could alternatively be created using the command
<PRE>

prettyalign ftrain.mod -m 8 &gt; ftrain2.pretty
</PRE>
to produce the alignment:
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
TRNA1&nbsp;&nbsp;........GGGGAUGUAGCUCAG-.UGG...U.AGAGCGCAUGCUUCGCAUGUAUGAGGCCCC
TRNA2X&nbsp;gcg16aguGCGGCCGUCGUCUAGU.CUGgauU.AGGACGCUGGCCUCCCAAGCCAGCAAUCCC
TRNA3X&nbsp;ccc.....GGCCCUGUGGC-UAGC.UGG...UcAAAGCGCCUGUCUAGUAAACAGGAGAUCCU
TRNA4X&nbsp;ggg15ucaGGGCGAAUAGUGUCAG.CGG...G.AGCACACCAGACUUGCAAUCUGGUAG-GGA
TRNA5X&nbsp;........GCCGGGAUAGCUCAGU.UGG...U.AGAGCAGAGGACUGAAAAUCCUCGUGUCAC
TRNA6&nbsp;&nbsp;........GGGGCCUUAGCUCAGC.UGG...G.AGAGCGCCUGCUUUGCACGCAGGAGGUCAG
TRNA7&nbsp;&nbsp;........GGGCACAUGGCGCAGU.UGG...U.AGCGCGCUUCCCUUGCAAGGAAGAGGUCAU
TRNA8&nbsp;&nbsp;........GGGCCCGUGGCCUAGU.CUGga.U.ACGGCACCGGCCUUCUAAGCCGGGGAUCGG
TRNA9&nbsp;&nbsp;c.......-GGCACGUAGCGCAGCcUGG...U.AGCGCACCGUCCUGGGGUUGCGGGGGUCGG
TRNA10&nbsp;........UCCGUCGUAGUCUAGG.UGGu..U.AGGAUACUCGGCUUUCACCCGAGAGA-CCC

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;GGGUUCGAUCCCCGGCAUCUCCA-.........
TRNA2X&nbsp;GGGUUCGAAUCCCGGCGGCCGCAC-ggc10gca.
TRNA3X&nbsp;GGGUUCGAAUCCCAGCGGGGCCUC-cagggg...
TRNA4X&nbsp;GGGUUCGAGUCCCUCUUUGUCCAC-ca.......
TRNA5X&nbsp;CAGUUCAAAUCUGGUUCCUGGCAU-ggu11gca.
TRNA6&nbsp;&nbsp;CGGU-CGA-CCCGCUAGGCUCCAC-ca.......
TRNA7&nbsp;&nbsp;CGGUUCGAUUCCGGUUGCGUCCA-.........
TRNA8&nbsp;&nbsp;GGGUUCAAAUCCCUCCGGGUCCG-.........
TRNA9&nbsp;&nbsp;AGGUUCAAAUCCUCUCGUGCCGAC-ca.......
TRNA10&nbsp;GGGUUCAAGUCCCGGCGACGGAAC-ca.......

<P>
</PRE>

<P>
The <code>-I</code> switch will create a compatible IG-style alignment file which
may be converted to other formats using the <TT>readseq</TT> package
included as a subdirectory of SAM.  The <code>-I</code> option
automatically sets a high value for the insertion length parameter.

<P>
The <code>-b</code> switch generate a BLAST-style output, comparing the
first sequence in the alignment (by default) to each subsequent
sequence, and generating pairwise alignments with a middle line
highlighting identical residues and conservative substitutions.  These
pairwise alignments do not include terminal inserts: These pairwise
alignments run from the first to last residue aligned by either
sequence.  They do not include terminal inserts; they include insernal
inserts only if some sequence in the pairwise alignment is inserting
residues.  The <code>-R</code> option can be used to specify a different
sequence to use as this reference sequence.

<P>

<H3><A NAME="SECTION000121200000000000000"></A><A NAME="sec:fragalign"></A>
<BR>
10.1.2 Aligning fragments
</H3>

<P>
Consider a model of 100 nodes and a fragment of 25 that very closely
matches some contiguous section of the model.  Even though that
section would align very well, the overall alignment of the fragment could
be quite poor because of its need to use 75 delete states in the
model.  The problem here is that in addition to modeling conserved
regions, the model also models the length of the conserved region.

<P>
SAM has two alignment options for dealing with fragments and one for
domains.  The first option allows a sequence to start matching the
model at any location (rather than only the begin node) and end at any
location (rather than only the end node).  This will improve alignment
for short sequences that match a segment of the model.  This option
can be turned on by setting the <TT>SW</TT> variable to one.  

<P>
The second option is similar to Smith and Waterman method of sequence
comparison, which will find the best alignment for any pair of
subsequences within two sequences.  The same can be done with models,
allowing a submodel to match a subsequence.  This type of dynamic
programming can be specified by setting the <TT>SW</TT> variable to 2.
When this is done, sequences can jump from the initial module
(presumably a FIM, automatically added when <TT>auto_fim</TT> is set)
into the match state of any module in the model, and can also jump
out of the match state of any module within the model to the delete
state of the next-to-last node.  The first and next-to-last module are
assumed to be FIMs, hence the rational is that a sequence will use the
FIM for some period of time to consume characters that do not match
the model, then the sequence will jump to the model node corresponding
to the start of the fragment, use several model nodes, and then jump
to the ending FIM to consume the rest of the sequence.

<P>
The probability of these jumps is set by the variables <TT>jump_in_prob</TT> and <TT>jump_out_prob</TT>, both of which have a
default value of unity.  That is, as in the sequence-to-sequence Smith
and Waterman,  there is no cost associated with jumping in and out of
the model.

<P>
For domains, <TT>SW</TT> can be set to 3 to match part of a model to all
of a sequence.

<P>
The file <TT>trna1frag.seq</TT> contains several sequences that contain
part or all of TRNA1.  The sequence include TRNA1 (72 bases),
TRNA1Long (the complete tRNA with additional characters), Long (58
base segment of TRNA1), Medium (34 base segment), Short (6 base
segment TRNA1), and AAMediumA, an embedding of Medium within several
segments of As to bring it to 176 characters.  Additionally, the file
contains several (obviously) non-tRNAs of various lengths, all of
whose IDs begin with the word `Not'.

<P>
When this file is aligned to the model <TT>test.mod</TT>, created above,
the alignment of the sequence and fragments is reasonable, but the
non-tRNAs still align the entire model and may even use internal
insertion states.  (As shall be seen in Section&nbsp;<A HREF="node12.html#sec:fragments">10.2.3</A>,
the scoring of these fragments with the SW option off is not nearly so
good as their alignments).

<P>
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........GGGGAUGUAGCUCAG-UGGU........AGAGCGCA-UGCUUCGCAUGUAUG
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa11aaaGGGGAUGUAGCUCAG-UGGU........AGAGCGCA-UGCUUCGCAUGUAUG
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........----------........-------------
ShortReverse&nbsp;&nbsp;........UCGUAC-------........-------------
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........G----------........-AGCGCA-UGCUUCGCAUGUAUG
MediumReverse&nbsp;........UUGGGC-CCCGGAGUAUGU........A-----CGCUUCGUACGCGAG
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;........AAAAAAAAAAAAAAAAAAAAaaa48aaaAAAGAGCGCAUGCUUCGCAUGUAUG
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........-----GCUCAG-UGGU........AGAGCGCA-UGCUUCGCAUGUAUG
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........----------........--------AAAAAAAAA
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........AAAAAAAAAAAAAAA--A........AAAAAAAAAAAAAAAAAAAAAAAAA
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........AAAAAAAAAAAAAAAAAAAA........AAAAAAAAAAAAAAAAAAAAAAAAA
NotExtraLong&nbsp;&nbsp;........AAAAAAAAAAAAAAAAAAAA........AAAAAAAAAAAAAAAAAAAAAAAAA

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AGGCCCcGGGUUCGAUCCCC--GGCAUCUCCA.........
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AGGCCCcGGGUUCGAUCCCCGGCAUCUCCAAAAaaaaaaaa.
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;---.----------CAUGCU.........
ShortReverse&nbsp;&nbsp;---.-------------.........
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AGGCCCcGGGUU-----------.........
MediumReverse&nbsp;---.-------------.........
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;AGGCCCcGGGUUAAAAAAAAAAAAAAAAAAAAAaaa50aaa.
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AGGCCCcGGGUUCGAUCCCCGGCAU----.........
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AAA--.-------------A.........
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AAAAAA.AAAAAAAAAAAAAAAAAAAAAAAAAA.........
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AAAAAA.AAAAAAAAAAAAAAAAAAAAAAAAAAaaa32aaa.
NotExtraLong&nbsp;&nbsp;AAAAAA.AAAAAAAAAAAAAAAAAAAAAAAAAAaa361aaa.

<P>
</PRE>

<P>
Alignment with the <TT>SW</TT> option set to 1 is much the same (again,
scoring will be improved), though the alignment procedure has managed
to better isolate the AAMediumAA sequence's tRNA core, modeled by
match states, from its prefix and postfix, modeling by internal
insertion nodes.

<P>
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........GGGGAUGUAGCUCAG-UGGU........A....GAGCGCA-UGCUUCGCAUG
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa11aaaGGGGAUGUAGCUCAG-UGGU........A....GAGCGCA-UGCUUCGCAUG
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........----------........-....--------CAUG
ShortReverse&nbsp;&nbsp;........UCGUAC-------........-....----------
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........----------........-....GAGCGCA-UGCUUCGCAUG
MediumReverse&nbsp;........--------UUGGG........CcccgGAGUAUGUACGCUUCGUACG
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;........AAAAAAAAAAAAAAAAAAAAaaa48aaaA....AAGAGCGCAUGCUUCGCAUG
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........-----GCUCAG-UGGU........A....GAGCGCA-UGCUUCGCAUG
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........----------........-....-------AAAAAA
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........--AAAAAAAAAAAAAAAA........A....AAAAAAAAAAAAAAAAAAAA
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........AAAAAAAAAAAAAAAAAAAA........A....AAAAAAAAAAAAAAAAAAAA
NotExtraLong&nbsp;&nbsp;........AAAAAAAAAAAAAAAAAAAA........A....AAAAAAAAAAAAAAAAAAAA

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;UAUGAGGCCCcGGGUUCGAUCCCCGGCAUCUCCA--.........
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;UAUGAGGCCCcGGGUUCGAUCCCCGGCAUCUCCAAAAaaaaaaaa.
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;CU----.-------------.........
ShortReverse&nbsp;&nbsp;-----.-------------.........
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;UAUGAGGCCCcGGGUU-----------.........
MediumReverse&nbsp;CGAG---.-------------.........
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;UAUGAGGCCCcGGGUUAAAAAAAAAAAAAAAAAAAAAaaa50aaa.
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;UAUGAGGCCCcGGGUUCGAUCCCCGGCAU----.........
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AAAAAAA--.-------------.........
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AAAAAAAAAA.AAAAAAAAAAAAAAAAAAAAAAAAAA.........
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AAAAAAAAAA.AAAAAAAAAAAAAAAAAAAAAAAAAAaaa32aaa.
NotExtraLong&nbsp;&nbsp;AAAAAAAAAA.AAAAAAAAAAAAAAAAAAAAAAAAAAaa361aaa.

<P>
</PRE>

<P>
When alignment is performed using the <TT>SW</TT> option set to
2, only the core tRNA segments are aligned:  the non-tRNAs, as well
as the prefix and postfix of AAMediumAA are aligned to the FIMs that
have been automatically added to the model.  The one problem is that
the Short sequence has made some use of the end FIM because it is not
long enough to make a really significant hit to the model's internal
nodes. 

<P>
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........GGGGAUGUAGCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa11aaaGGGGAUGUAGCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........-------------------CAUGC-----.-
ShortReverse&nbsp;&nbsp;u.......---CGUA---------------------.-
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........-----------GAGCGCA-UGCUUCGCAUGUAUGAGGCCC.C
MediumReverse&nbsp;uug20uac----------------GCUUCGUACGCGAG---.-
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;aaa68aaa----------AAAGAGCGCAUGCUUCGCAUGUAUGAGGCCC.C
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........-----GCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a.......--------------------------.-
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a.......--------------------------.-
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a.......--------------------------.-
NotExtraLong&nbsp;&nbsp;a.......--------------------------.-

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCCGGCAUCUCCA--.........
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCCGGCAUCUCCA--aaa11aaa.
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------------u........
ShortReverse&nbsp;&nbsp;-------------c........
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGGU-----------u........
MediumReverse&nbsp;-------------.........
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;GGGUUAAA---------aaa68aaa.
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCCGGC-----au.......
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;---AAA---------aaa9aaaa.
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;---AAA---------aaa69aaa.
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;---AAA---------aa105aaa.
NotExtraLong&nbsp;&nbsp;---AAA---------aa434aaa.

<P>
</PRE>

<P>
A different alignment is produced when <TT>SW</TT> is set to 2
(fully local) and <TT>adpstyle</TT> is set to 4 (posterior-decoded
alignment with transitions).  As can be seen in the example below,
this option does not neatly cut the sequences to their matching
positions, but may produce a better core alignment.

<P>
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........GGGGAUGUAGCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa11aaaGGGGAUGUAGCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ca......--------------------------.-
ShortReverse&nbsp;&nbsp;uc......---------GUA---------------.-
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........-----------GAGCGCA-UGCUUCGCAUGUAUGAGGCCC.C
MediumReverse&nbsp;uug.....GGCCCCGG---AGUAUGUA-----CGCUUCGUACGCG----.-
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;aaa64aaa--------AAAAAAAGAGCGCAUGCUUCGCAUGUAUGAGGCCCcG
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........-----GCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aa......-----------AAAAAAA------------.-
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa22aaa-------------AAAAAAAAAAAAAAAAAAAAAAAAAA.A
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa54aaa-------------------A------.-
NotExtraLong&nbsp;&nbsp;aa217aaa-------------------A------.-

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCC--GGCAUCUCCA.........
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCCGGCAUCUCCAAAAaaaaaaaa.
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-----------UGCU.........
ShortReverse&nbsp;&nbsp;-------------c........
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGGU-----------u........
MediumReverse&nbsp;-------------ag.......
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;GGUUAAAAAAAAA------aaa62aaa.
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCCGGCA-----u........
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------------aaaa.....
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------------aaa24aaa.
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------------aaa54aaa.
NotExtraLong&nbsp;&nbsp;-------------aa220aaa.

<P>
</PRE>

<P>
As well as when <TT>SW</TT> is set to 2
(fully local) and <TT>adpstyle</TT> is set to 5 (posterior-decoded
alignment solely on character emission).  

<P>
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;g.......-GGGAUGUAGCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa12aag-GGGAUGUAGCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ca......--------------------------.-
ShortReverse&nbsp;&nbsp;uc......--------------------------.-
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;g.......-----------AGCGCA-UGCUUCGCAUGUAUGAGGCCC.C
MediumReverse&nbsp;uug14agu--------------------------.-
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;aaa69aaa-----------AAGAGCGCAUGCUUCGCAUGUAUGAGGCCCcG
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........-----GCUCAG-UGGUAGAGCGCA-UGCUUCGCAUGUAUGAGGCCCcG
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaaaa...--------------------------.-
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa28aaa--------------------------.-
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa42aaa--------------------------.-
NotExtraLong&nbsp;&nbsp;aa167aaa--------------------------.-

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCCGGCAUCUCC--a........
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCCGGCAUCUCCA--aaa11aaa.
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------------ugcu.....
ShortReverse&nbsp;&nbsp;-------------guac.....
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GG------------guu......
MediumReverse&nbsp;-------------aug20gag.
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;GGUUAAAA---------aaa67aaa.
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGUUCGAUCCCCGG------cau......
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------------aaaaaaaa.
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------------aaa45aaa.
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------------aaa67aaa.
NotExtraLong&nbsp;&nbsp;-------------aa271aaa.

<P>
</PRE>

<P>
Here is the posterior-decoded (<TT>adpstyle</TT> 4) global alignment  (<TT>SW</TT> is 0).

<P>
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........GGGG........AUGUA........GCUC........AG-U........GGU.
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa11aaaGGGG........AUGUA........GCUC........AG-U........GGU.
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........--........---........--........--........--.
ShortReverse&nbsp;&nbsp;........UCG-........---........--........--........-U.
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........G--........---........--........--........--.
MediumReverse&nbsp;........UUGG........GC-C........CCGG........AGUA........UGU.
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;........AAAA........AAAAAaaa25aaaAAAA........AAAAaaa23aaaAAA.
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........G--........---........-CUC........AG-U........GGU.
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........--........---........--........--........--.
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........AAAA........AAAAA........AAAA........AAA-........--.
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........AAAA........AAAAA........AAAA........AAAA........AAAa
NotExtraLong&nbsp;&nbsp;........AAAAaaa43aaaAAAAA........AAAAaaa35aaaAAAA........AAAa

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......AGAGCGCA-UGCUU........CGCAUGUAU.......GAGGCCCc.......
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......AGAGCGCA-UGCUU........CGCAUGUAU.......GAGGCCCc.......
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......--------........-CAUGCU-.......----........
ShortReverse&nbsp;&nbsp;.......A-------........-----.......---C........
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......-AGCGCA-UGCUU........CGCAUGUAU.......GAGGCCC........
MediumReverse&nbsp;.......A-----CGCUU........CGUACGCGA.......G---........
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;.......AAAGAGCGCAUGCUU........CGCAUGUAU.......GAGGCCCc.......
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......AGAGCGCA-UGCUU........CGCAUGUAU.......GAGGCCCc.......
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......AAAAAA-----........-AAA--.......----........
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......AAAAAAAAAAAAAAA........AAAAAAAAA.......AAAAAAA........
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aa12aaaAAAAAAAAAAAAAAA........AAAAAAAAAaaaaaaaAAAAAAAaaa13aaa
NotExtraLong&nbsp;&nbsp;aa51aaaAAAAAAAAAAAAAAAaaa71aaaAAAAAAAAA.......AAAAAAA........

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGGUUCGAUCCCC........--GGCAUCUC........CA.........
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGGUUCGAUCCCC........GGCAUCUCCAAaaaaaaaaAA.........
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-------........------........-.........
ShortReverse&nbsp;&nbsp;-------........------........-.........
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;CGGGUU----........------........-.........
MediumReverse&nbsp;-------........------........-.........
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;GGGUUAAAAAAAAaaa24aaaAAAAAAAAAAA........AAaaa26aaa.
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;GGGUUCGAUCCCC........GGCAU---........-.........
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;---AAA--........------........-A.........
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AAAAAAAAAAAAA........AAAAAAAAAAA........AA.........
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AAAAAAAAAAAAA........AAAAAAAAAAA........AA.........
NotExtraLong&nbsp;&nbsp;AAAAAAAAAAAAA........AAAAAAAAAAA........AAaa161aaa.

<P>
</PRE>

<P>
Here is the posterior-decoded (<TT>adpstyle</TT> 5) global alignment
(<TT>SW</TT> is 0). 

<P>
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........G........G...G..G........AUGUA........G...CU.C.......
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aaa11aaaG........G...G..G........AUGUA........G...CU.C.......
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........C........-...-..-........---........-...-.-.......
ShortReverse&nbsp;&nbsp;........U........C...G..-........---........-...-.-.......
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........G........-...-..-........---........-...-.-.......
MediumReverse&nbsp;........U........U...G..G........GCC-........C...CG.G.......
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;........A........A...A..Aaaa17aaa---aaa10aaa-...-a-aaa12aa
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........G........-...-..-........---........-...CU.C.......
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........A........-...-..-........---........-...-.-.......
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........A........A...A..A........AAAAA........A...AA.A.......
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;........A........A...A..A........AAAAA........A...AA.A.......
NotExtraLong&nbsp;&nbsp;........Aaaa20aaa-aaa-aa-aaa19aaa---aaa19aaa-aaa-.-aaa25aa

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.AG-U.......G..G.U........AGAGCGCA-........-UGC........UU....
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.AG-U.......G..G.U........AGAGCGCA-........-UGC........UU....
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.--.......-..-.-........----A-........--........-....
ShortReverse&nbsp;&nbsp;.--.......-..-.U........----A-........--........-....
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.--.......-..-.-........-AGCGCA-........-UGC........UU....
MediumReverse&nbsp;.AGUA.......U..G.U........A----........-CGC........UU....
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;a--aaaaa..-a.-a-aaa17aaaAAAGAGCGC........AUGC........UU....
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.AG-U.......G..G.U........AGAGCGCA-........-UGC........UU....
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.--.......-..-.-........-AAAAA--........--........-....
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.AAAA.......A..A.A........AAAAAAAA-........--A........AA....
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.AAAA.......A..A.Aaaa11aaaAAAAAAAAAa.......AAAA........AA....
NotExtraLong&nbsp;&nbsp;a--aaaaaaa-aa-a-aaa44aaa-----aaa23aaa--aaa17aaa-aaa4

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....CGCAUGU........AU.......GAGGCCCc.......GGGU........UCGAU.
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....CGCAUGU........AU.......GAGGCCCc.......GGGU........UCGAU.
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....----........-.......----........--........---.
ShortReverse&nbsp;&nbsp;....----........-.......----........--........---.
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....CGCAUGU........AU.......GAGGCCC........CGGG........U--.
MediumReverse&nbsp;....CGUACGC........GA.......----........--........---.
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;....CGCAUGU........AU.......GAGGCCCc.......GGGU........UAAAAa
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....CGCAUGU........AU.......GAGGCCCc.......GGGU........UCGAU.
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....--AA-........-.......-A---........--........-AA-.
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....AAAAAAA........AA.......AAAAAAA........AAAA........AAAAA.
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....AAAAAAA........AAaaaaaaaAAAAAAAaaa13aaaAAAA........AAAAA.
NotExtraLong&nbsp;&nbsp;1aaa----aaa23aaa-.......----........--aaa24aaa---.

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......C.CC.C........-....-GGCAUCUC........CA.........
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......C.CC.C........GG....CAUCUCCAAaaaaaaaaAA.........
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......-.-.-........-....----UG........CU.........
ShortReverse&nbsp;&nbsp;.......-.-.-........-....-----........-C.........
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......-.-.-........-....-----........-U.........
MediumReverse&nbsp;.......-.-.-........-....-----........-G.........
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;aa15aaa-a-a-aaa15aaa-aaaa-----........-aaa31aaa.
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......C.CC.C........GG....CAU---........-.........
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......-.-.-........-....-----........AA.........
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......A.AA.A........AA....AAAAAAAAA........AA.........
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.......A.AA.A........AA....AAAAAAAAA........AA.........
NotExtraLong&nbsp;&nbsp;.......-.-.-........-....-----........-aa164aaa.

<P>
</PRE>

<P>

<H2><A NAME="SECTION000122000000000000000"></A><A NAME="sec:hmmscore"></A>
<BR>
10.2 <TT>hmmscore</TT>
</H2>

<P>
Any sequence can be compared to a model by calculating the probability
that the sequence was generated by that model.  Taking the negative
(natural) logarithm of this probability gives the NLL score.  For
sequences of equal length the NLL scores measures how `far' they are
from the model, and it can be used to select sequences that are from
the same family.  However, the NLL score has a strong dependence on
sequence length and model length.  <TT>Hmmscore</TT> provides several
less biased means of scoring by reporting NLL scores as the difference
between a null model and trained model NLL score (a log-odds score, as
used in HMMER).

<P>
Null model scoring is discussed in more detail in the Barrett,
Karplus, and Hughey paper mentioned in the introduction and available
from the SAM WWW page. <BR>(<A NAME="tex2html53"
  HREF="http://www.cse.ucsc.edu/research/compbio/papers/nullmod/nullmod.html"><TT>http://www.cse.ucsc.edu/research/compbio/papers/nullmod/nullmod.html</TT></A>).

<P>
The program <TT>hmmscore</TT> can find NLL and NLL<IMG
 WIDTH="18" HEIGHT="31" ALIGN="MIDDLE" BORDER="0"
 SRC="img13.png"
 ALT="$-$">NULL (log-odds)
scores.  E-values can be calculated for the reverse sequence null model.
The most common operation is to calculate NLL<IMG
 WIDTH="18" HEIGHT="31" ALIGN="MIDDLE" BORDER="0"
 SRC="img13.png"
 ALT="$-$">NULL scores for a large
number of sequences.  This can be done by supplying the name of the
model file and one or more sequence database files on the command line,
optionally followed by <TT>hmmscore</TT> parameter specifications.
For instance, for the example files described
earlier the NLL scores are found the following way
<PRE>

hmmscore test -insert test.mod -db trna10.seq -sw 2
</PRE>

<P>
Produces the file <TT>test.dist</TT> already displayed:
<PRE>

%&nbsp;&nbsp;SAM:&nbsp;hmmscore&nbsp;v3.3.1&nbsp;(December&nbsp;20,&nbsp;2001)&nbsp;compiled&nbsp;12/27/01_15:14:20
%&nbsp;&nbsp;(c)&nbsp;1992-2001&nbsp;Regents&nbsp;of&nbsp;the&nbsp;University&nbsp;of&nbsp;California,&nbsp;Santa&nbsp;Cruz
%
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence&nbsp;Alignment&nbsp;and&nbsp;Modeling&nbsp;Software&nbsp;System
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;http://www.cse.ucsc.edu/research/compbio/sam.html
%
%&nbsp;---------&nbsp;Citations&nbsp;(SAM,&nbsp;SAM-T99,&nbsp;HMMs)&nbsp;----------
%&nbsp;R.&nbsp;Hughey,&nbsp;A.&nbsp;Krogh,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;sequence&nbsp;analysis:
%&nbsp;&nbsp;Extension&nbsp;and&nbsp;analysis&nbsp;of&nbsp;the&nbsp;basic&nbsp;method,&nbsp;CABIOS&nbsp;12:95-107,&nbsp;1996.
%&nbsp;K.&nbsp;Karplus,&nbsp;C.&nbsp;Barrett,&nbsp;R.&nbsp;Hughey,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;detecting
%&nbsp;&nbsp;remote&nbsp;protein&nbsp;homologies,&nbsp;Bioinformatics&nbsp;14(10):846-856,&nbsp;1998.
%&nbsp;A.&nbsp;Krogh&nbsp;et&nbsp;al.,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;in&nbsp;computational&nbsp;biology:
%&nbsp;&nbsp;Applications&nbsp;to&nbsp;protein&nbsp;modeling,&nbsp;JMB&nbsp;235:1501-1531,&nbsp;Feb&nbsp;1994.
%&nbsp;-----------------------------------
%&nbsp;&nbsp;test&nbsp;&nbsp;&nbsp;Host:&nbsp;perch&nbsp;&nbsp;&nbsp;&nbsp;Thu&nbsp;Dec&nbsp;27&nbsp;16:02:54&nbsp;2001
%&nbsp;&nbsp;rph&nbsp;&nbsp;&nbsp;&nbsp;Dir:&nbsp;&nbsp;/projects/kestrel/rph/sam32/SAMBUILD/perch/demos
%&nbsp;-----------------------------------
%&nbsp;Inserted&nbsp;Files:&nbsp;&nbsp;test.mod
%&nbsp;Database&nbsp;Files:&nbsp;&nbsp;/projects/kestrel/rph/sam32/demos/trna10.seq
%
%&nbsp;Subsequence-submodel&nbsp;(local)&nbsp;(SW&nbsp;=&nbsp;2)
%&nbsp;Simple&nbsp;scores&nbsp;adjusted&nbsp;by&nbsp;ln(seq&nbsp;len)&nbsp;(adjust_score&nbsp;=&nbsp;2)
%&nbsp;Track&nbsp;0&nbsp;FIMs&nbsp;added&nbsp;(geometric&nbsp;mean&nbsp;of&nbsp;match&nbsp;probabilities&nbsp;(6))
%&nbsp;Single&nbsp;Track&nbsp;Model:&nbsp;&nbsp;test.mod
%&nbsp;DP&nbsp;Method:&nbsp;forward&nbsp;all-paths&nbsp;(dpstyle&nbsp;=&nbsp;0)
%&nbsp;10&nbsp;sequences,&nbsp;747&nbsp;residues,&nbsp;78&nbsp;nodes,&nbsp;0.03&nbsp;seconds
%
%&nbsp;Sequence&nbsp;scores&nbsp;selected:&nbsp;&nbsp;All&nbsp;&nbsp;(select_score=8)
%
%&nbsp;Simple:&nbsp;NLL-NULL&nbsp;using&nbsp;FIM&nbsp;probabilities
%&nbsp;Reverse:&nbsp;NLL-NULL&nbsp;for&nbsp;the&nbsp;reverse&nbsp;sequence&nbsp;NULL&nbsp;model
%&nbsp;&nbsp;&nbsp;&nbsp;Calculated&nbsp;when&nbsp;Simple&nbsp;&lt;&nbsp;simple_threshold&nbsp;(0.00)
%&nbsp;E-value&nbsp;on&nbsp;N&nbsp;(=10)&nbsp;sequences:
%&nbsp;&nbsp;&nbsp;&nbsp;N&nbsp;/&nbsp;(1&nbsp;+&nbsp;exp(-(lambda(=1.0000)&nbsp;*&nbsp;Reverse)**tau(=1.0000)))
%&nbsp;&nbsp;&nbsp;&nbsp;Calculated&nbsp;when&nbsp;Simple&nbsp;&lt;&nbsp;simple_threshold&nbsp;(0.00)
%&nbsp;&nbsp;&nbsp;&nbsp;Rescale&nbsp;E-values&nbsp;or&nbsp;use&nbsp;-dbsize&nbsp;for&nbsp;multiple&nbsp;scoring&nbsp;runs.
%&nbsp;&nbsp;&nbsp;&nbsp;WARNING:&nbsp;&nbsp;E-VALUES&nbsp;ARE&nbsp;NOT&nbsp;CALIBRATED!
%&nbsp;Scores&nbsp;sorted&nbsp;by&nbsp;E-value,&nbsp;best&nbsp;first
%
%&nbsp;Sequence&nbsp;ID&nbsp;&nbsp;&nbsp;Length&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Simple&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;E-value&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;X&nbsp;count
TRNA8&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;75&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-41.85&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-32.87&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;5.32e-14
TRNA10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-40.26&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-32.34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;9.02e-14
TRNA2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-38.68&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-31.45&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.20e-13
TRNA7&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-37.12&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.90&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.82e-12
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-35.84&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.51&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.14e-12
TRNA4&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;75&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-36.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.12&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.14e-12
TRNA9&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;77&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-33.86&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-26.57&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.88e-11
TRNA3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-34.30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-25.57&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;7.85e-11
TRNA5&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-34.73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-25.47&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;8.67e-11
TRNA6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;74&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-35.14&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-25.39&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;9.37e-11
</PRE>

<P>
As discussed in Section&nbsp;<A HREF="node4.html#sec:score">3.4</A>, the score file contains six
columns.  The first is the sequence identifier, followed by sequence
length, the `NLL-NULL' score using a simple null model.  The next
score column is either the raw NLL score if only the simple null model
is calculated (the default), or the complex or reverse sequence null
model's `NLL-NULL'  if one of the more time-consuming null models
is used, as discussed below.  If E-values are calculated, they are
listed after the two score columns.  Last, the number of all-character
wildcards in each sequence is listed for those that have any
wildcards.

<P>
By default, <TT>hmmscore</TT> uses the EM scoring method, just as is used
to train a model.  If desired, scores can be based on exact alignment
to the model, multiplying the probabilities along the best path rather
than all paths.  This method, which corresponds to the forward half of
<TT>align2model</TT>, can be turned on by setting <TT>viterbi</TT>
to 1.  Viterbi scoring is appropriate for finding out how good a
sequence's best alignment to a model is.

<P>
The E-value computation is based on a simple, but somewhat
unrealistic, assumption that the scores for the sequence and the
reversed-sequence are independent draws from an extreme-value (Gumbel)
distribution: 
<BR><P></P>
<DIV ALIGN="CENTER">
<!-- MATH
 \begin{displaymath}
P( \mbox{score} < a) \approx 1- e^(-k e^{\lambda a}) \ .
\end{displaymath}
 -->

<IMG
 WIDTH="389" HEIGHT="28" BORDER="0"
 SRC="img65.png"
 ALT="\begin{displaymath}P( \mbox{score} &lt; a) \approx 1- e^(-k e^{\lambda a}) \ .
\end{displaymath}">
</DIV>
<BR CLEAR="ALL">
<P></P>
Subtracting the two scores (derivation not published yet) gives 
<BR><P></P>
<DIV ALIGN="CENTER">
<!-- MATH
 \begin{displaymath}
P( \mbox{diff} < a) \approx {1 \over 1+ e^{-\lambda a}} \ .
\end{displaymath}
 -->

<IMG
 WIDTH="366" HEIGHT="40" BORDER="0"
 SRC="img66.png"
 ALT="\begin{displaymath}P( \mbox{diff} &lt; a) \approx {1 \over 1+ e^{-\lambda a}} \ .
\end{displaymath}">
</DIV>
<BR CLEAR="ALL">
<P></P>
The E-value is the expected number of sequences with that good a
score, so is simply the probability of seeing that negative a
difference, multiplied by the number of sequences scored (or
<TT>dbsize</TT>, if that is specified).

<P>
The only parameter that needs to be estimated is the natural scaling
<IMG
 WIDTH="15" HEIGHT="18" ALIGN="BOTTOM" BORDER="0"
 SRC="img67.png"
 ALT="$\lambda$">.  Since we use natural logs in computing our probabilities,
we set <IMG
 WIDTH="45" HEIGHT="18" ALIGN="BOTTOM" BORDER="0"
 SRC="img68.png"
 ALT="$\lambda=1$">, which seems to be correct experimentally.  Thus we
compute our E-values with no parameter fitting at all--it is based
purely on theoretical considerations.

<P>
The <TT>calibrate</TT> option of <TT>hmmscore</TT> can be used to calibrate
the lambda values for the E-value calculation.  Once a model library
has been created, it can be specified to <TT>hmmscore</TT> using the
<TT>model_library</TT> directive (or <TT>modlib</TT> alias). See Section&nbsp;<A HREF="node12.html#sec:calibrate">10.2.9</A>.

<P>
The <TT>hmmscore</TT> program can also be used to select sequences
according to various criteria.

<P>
Plots of the NLL-NULL scores  can be used to visually look
for a break between significant and insignificant matches.
See Section&nbsp;<A HREF="node12.html#sec:plottingprogs">10.10</A>. 

<P>

<H3><A NAME="SECTION000122100000000000000"></A><A NAME="sec:nll_null"></A>
<BR>
10.2.1 NLL-NULL scoring
</H3>

<P>
SAM includes several possibilities for NULL model scoring.  The null
model can be a simple probability distribution, effectively a model
with a single FIM.  The null model can be any model specified in SAM
format, with the key word `NULLMODEL' (rather than, for example,
`MODEL' or `REGULARIZER'), or the first model in a file specified with
the <TT>nullmodel_file</TT> parameter.  The null model can be the
reversed HMM, or equivalently, the score of the reversed sequence
through the original HMM.

<P>
To report differences between the model NLL score and the simple null
model score (possibly modified by <TT>FIM_method_score</TT>, see
below), set the <TT>subtract_null</TT> variable to 1.
To report differences between two models (for example, one trained
on positive family examples and one trained on negative examples of a
family), set <TT>subtract_null</TT> to 3.  To report differences
between the score of the sequence and the score of the reversed
sequence, which provides an automatic adjustment for compositional
bias and allows the simple calculation of E-values, set <TT>subtract_null</TT> to 4 (the default).  The complex null model
(previously  <TT>subtract_null</TT> 2) is no longer supported, and
reverts to  the simple null model.

<P>
The simple null model score (<TT>subtract_null</TT>=1) is actually
calculated in all cases.  If the other null model calculations were to
be calculated all the time, it would double the running time of <TT>hmmscore</TT>.  The <TT>simple_threshold</TT> variable can be used to
control when the more time consuming score should be calculated: the
it will only be calculated when the simple null model score is less
than <TT>simple_threshold</TT>.  Sequences for which the more time
consuming null model was not calculated will have the score 10000 in
the second score column of the distance file. The default value of
<TT>simple_threshold</TT> is 0; a setting of <IMG
 WIDTH="34" HEIGHT="31" ALIGN="MIDDLE" BORDER="0"
 SRC="img5.png"
 ALT="$-10$"> is a reasonable
setting to further reduce the amount of calculation performed without
losing too many E-values.

<P>
A related varable, <TT>viterbi_threshold</TT> can be used to perform
Viterbi simple null model score filtering before completing any
EM-style scoring.  Because EM-style scoring is 2-5 times slower than
Viterbi scoring, this can result in quite a runtime savings.
Typically, it should be set to the same or a less negative value than
<TT>simple_threshold</TT>.

<P>
Our current favorites (and the default) are the simple null model
(geometric average) mixed 
with the reverse sequence null model for well-scoring sequences.
Because the difference between the sequence and reverse sequence
scores automatically adjusts for model and sequence length, SAM is
able to add E-values to the score file in this case.  The
command
<PRE>

hmmscore testrev -i test.mod -db trna10.seq -subtract_null 4 -sw 2
</PRE>
produces a score file that includes both simple and reverse-sequence
null model scores:
<PRE>

%&nbsp;&nbsp;SAM:&nbsp;hmmscore&nbsp;v3.3.1&nbsp;(December&nbsp;20,&nbsp;2001)&nbsp;compiled&nbsp;12/27/01_15:14:20
%&nbsp;&nbsp;(c)&nbsp;1992-2001&nbsp;Regents&nbsp;of&nbsp;the&nbsp;University&nbsp;of&nbsp;California,&nbsp;Santa&nbsp;Cruz
%
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence&nbsp;Alignment&nbsp;and&nbsp;Modeling&nbsp;Software&nbsp;System
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;http://www.cse.ucsc.edu/research/compbio/sam.html
%
%&nbsp;---------&nbsp;Citations&nbsp;(SAM,&nbsp;SAM-T99,&nbsp;HMMs)&nbsp;----------
%&nbsp;R.&nbsp;Hughey,&nbsp;A.&nbsp;Krogh,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;sequence&nbsp;analysis:
%&nbsp;&nbsp;Extension&nbsp;and&nbsp;analysis&nbsp;of&nbsp;the&nbsp;basic&nbsp;method,&nbsp;CABIOS&nbsp;12:95-107,&nbsp;1996.
%&nbsp;K.&nbsp;Karplus,&nbsp;C.&nbsp;Barrett,&nbsp;R.&nbsp;Hughey,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;detecting
%&nbsp;&nbsp;remote&nbsp;protein&nbsp;homologies,&nbsp;Bioinformatics&nbsp;14(10):846-856,&nbsp;1998.
%&nbsp;A.&nbsp;Krogh&nbsp;et&nbsp;al.,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;in&nbsp;computational&nbsp;biology:
%&nbsp;&nbsp;Applications&nbsp;to&nbsp;protein&nbsp;modeling,&nbsp;JMB&nbsp;235:1501-1531,&nbsp;Feb&nbsp;1994.
%&nbsp;-----------------------------------
%&nbsp;&nbsp;testrev&nbsp;&nbsp;&nbsp;Host:&nbsp;perch&nbsp;&nbsp;&nbsp;&nbsp;Thu&nbsp;Dec&nbsp;27&nbsp;16:02:57&nbsp;2001
%&nbsp;&nbsp;rph&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Dir:&nbsp;&nbsp;/projects/kestrel/rph/sam32/SAMBUILD/perch/demos
%&nbsp;-----------------------------------
%&nbsp;Inserted&nbsp;Files:&nbsp;&nbsp;test.mod
%&nbsp;Database&nbsp;Files:&nbsp;&nbsp;/projects/kestrel/rph/sam32/demos/trna10.seq
%
%&nbsp;Subsequence-submodel&nbsp;(local)&nbsp;(SW&nbsp;=&nbsp;2)
%&nbsp;Simple&nbsp;scores&nbsp;adjusted&nbsp;by&nbsp;ln(seq&nbsp;len)&nbsp;(adjust_score&nbsp;=&nbsp;2)
%&nbsp;Track&nbsp;0&nbsp;FIMs&nbsp;added&nbsp;(geometric&nbsp;mean&nbsp;of&nbsp;match&nbsp;probabilities&nbsp;(6))
%&nbsp;Single&nbsp;Track&nbsp;Model:&nbsp;&nbsp;test.mod
%&nbsp;DP&nbsp;Method:&nbsp;forward&nbsp;all-paths&nbsp;(dpstyle&nbsp;=&nbsp;0)
%&nbsp;10&nbsp;sequences,&nbsp;747&nbsp;residues,&nbsp;78&nbsp;nodes,&nbsp;0.03&nbsp;seconds
%
%&nbsp;Sequence&nbsp;scores&nbsp;selected:&nbsp;&nbsp;All&nbsp;&nbsp;(select_score=8)
%
%&nbsp;Simple:&nbsp;NLL-NULL&nbsp;using&nbsp;FIM&nbsp;probabilities
%&nbsp;Reverse:&nbsp;NLL-NULL&nbsp;for&nbsp;the&nbsp;reverse&nbsp;sequence&nbsp;NULL&nbsp;model
%&nbsp;&nbsp;&nbsp;&nbsp;Calculated&nbsp;when&nbsp;Simple&nbsp;&lt;&nbsp;simple_threshold&nbsp;(0.00)
%&nbsp;E-value&nbsp;on&nbsp;N&nbsp;(=10)&nbsp;sequences:
%&nbsp;&nbsp;&nbsp;&nbsp;N&nbsp;/&nbsp;(1&nbsp;+&nbsp;exp(-(lambda(=1.0000)&nbsp;*&nbsp;Reverse)**tau(=1.0000)))
%&nbsp;&nbsp;&nbsp;&nbsp;Calculated&nbsp;when&nbsp;Simple&nbsp;&lt;&nbsp;simple_threshold&nbsp;(0.00)
%&nbsp;&nbsp;&nbsp;&nbsp;Rescale&nbsp;E-values&nbsp;or&nbsp;use&nbsp;-dbsize&nbsp;for&nbsp;multiple&nbsp;scoring&nbsp;runs.
%&nbsp;&nbsp;&nbsp;&nbsp;WARNING:&nbsp;&nbsp;E-VALUES&nbsp;ARE&nbsp;NOT&nbsp;CALIBRATED!
%&nbsp;Scores&nbsp;sorted&nbsp;by&nbsp;E-value,&nbsp;best&nbsp;first
%
%&nbsp;Sequence&nbsp;ID&nbsp;&nbsp;&nbsp;Length&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Simple&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;E-value&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;X&nbsp;count
TRNA8&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;75&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-41.85&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-32.87&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;5.32e-14
TRNA10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-40.26&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-32.34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;9.02e-14
TRNA2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-38.68&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-31.45&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.20e-13
TRNA7&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-37.12&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.90&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.82e-12
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-35.84&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.51&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.14e-12
TRNA4&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;75&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-36.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.12&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.14e-12
TRNA9&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;77&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-33.86&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-26.57&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.88e-11
TRNA3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-34.30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-25.57&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;7.85e-11
TRNA5&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-34.73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-25.47&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;8.67e-11
TRNA6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;74&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-35.14&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-25.39&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;9.37e-11
</PRE>

<P>
The NLL-NULL scores, especially for the simple null model, are most
useful when the model has had free insertion modules
(Section&nbsp;<A HREF="node10.html#sec:fims">8.5</A>) added to it.  Then, the null model and the
FIMs will cancel out, and the score will be based primarily on the
section of the sequence that matches the region that has been modeled.
By default, <TT>hmmscore</TT> automatically adds FIMs to any model that
does not already contain them when SW scoring is used.  To change
this, if for example you want to ensure that entire sequences are
modeled, rather than simply subregions, change <TT>auto_fim</TT> from
its default value of 1 to 0.  The <TT>auto_fim</TT> variable has no
effect when a user-specified null model is used.

<P>
Since NLL-NULL scores are negative logs, the lower the better.
In the case above, all of the tRNAs have been positively identified
as tRNAs. (Not surprising as they were all in the training set!)

<P>
New to Version 1.2 is the ability to adjust the null model scoring.
Since this determines the probability that a sequence was randomly generated
according to the residue insertion probabilities, these values should
reflect knowledge of the problem domain.  Five possibilities are
offered.  The flat distribution or the background distribution of
amino acids over all proteins can be used.  Both of these
distributions are invariant over all families, and are thus a
simplistic assumption.  The distribution can also be the distribution
of the residues in the training set or the average residue
distribution over all columns (match states) modeled by the HMM.  The
advantage of these two, especially the latter, is the ability to
partially correct for compositional bias in the sequences.  Lastly, the
insertion probabilities can reflect the residue distribution of the
sequence currently being scored.  This is the most pessimistic null
model, as it demands not that the HMM model the sequence better than
fixed background frequencies, but that is model the sequence
significantly better than frequencies exactly matching the sequence's
composition.

<P>
So the options available for <TT>FIM_method_score</TT> are
<DL COMPACT>
<DT>0</DT>
<DD>Use the tables present in the model.
</DD>
<DT>1</DT>
<DD>The relative frequencies of residues in the training
sequences (from the Lettercounts node or the training sequences).
</DD>
<DT>2</DT>
<DD>The relative frequencies of residues in model match
states (from the Frequency node).
</DD>
<DT>3</DT>
<DD>Uniform (flat) probability over all residues.
</DD>
<DT>5</DT>
<DD>Amino acid background frequencies over all proteins
(from the Generic node).
</DD>
<DT>6</DT>
<DD>Geometric average of the match state probabilities in the model.
</DD>
</DL>

<P>
The default setting, for experimental and statistical reasons, is the
geometric average of the model match states (6).  The insertion
tables can be similarly modified with <TT>Insert_method_score</TT>,
the default of which is no change (0).

<P>
As with the training methods, if the method value is negative instead,
the FIMs and insert tables will only be modified if there is no
initial model read in (an unlikely occurrence for <TT>hmmscore</TT>).

<P>
The FIM scoring methods are only applied to the primary track of
multitrack models.  Other tracks never have their insert or FIM tables
changed, and if FIMs are added to the model either the character table
of the generic node is used (when present) or the geometric match
state average (when there is no generic node).

<P>
The FIM scoring methods are not applied to the user's null model.

<P>
A more detailed discussion of these issues can be found in (Barrett,
Hughey &amp; Karplus 1996), mentioned in the Introduction.

<P>

<H3><A NAME="SECTION000122200000000000000"></A><A NAME="sec:select"></A>
<BR>
10.2.2 Selecting sequences, scores, and alignments
</H3>

<P>
Sequences can be selected by <TT>hmmscore</TT> and placed in a
<code>.sel</code> file.  

<P>
A selection mode is chosen by setting <TT>select_seq</TT>.  If 0, no
sequences are selected; if 1, sequences are selected according to
their simple null model scores and <TT>NLLNull</TT>; if 2, sequences are
selected according to their column 2 score (user or reverse
sequence null, or raw NLL score if <TT>subtract_null</TT> is 0 or 1)
and <TT>NLLcomplex</TT>; if 4, sequences are selected 
according to their E-value and <TT>Emax</TT>; if 8, all sequences are
selected.  Selection criteria can be combined: 3 requires sequences to
score better than <TT>NLLnull</TT> with the simple null model and <TT>NLLcomplex</TT> with the complex (user or reverse sequence) null model.  Negative numbers indicate
that sequences that do not pass the corresponding positive test should
be selected.

<P>
The following will place labeled copies of all sequences
scoring lower than -35 with the reverse sequence null model into <TT>test.sel</TT>. 
<PRE>

hmmscore tests -i test.mod -db trna10.seq -select_seq 2 -NLLcomplex -35 -sw 2
</PRE>

<P>
Selected sequences are written out in the same order they are
encountered in the database files, which may be different from the
order they are listed in the score file if scores are sorted. The <TT>sortseq</TT> program can be used to write out the sequences in the same
order they are listed in the score file. See Section&nbsp;<A HREF="node12.html#sec:sortseq">10.11.7</A>. The
<TT>sort</TT> variable controls whether sequence scores are unsorted
(0); sorted by E-value (4), column 2 of the distance file (2), or
column 1 of the distance file (1).  Sequences in the select file will
be selected by other than the sorting criteria unless <TT>sort</TT> and
<TT>select_seq</TT> are set to corresponding values.

<P>
The <TT>select_score</TT> variable can be set in the same manner as
<TT>select_seq</TT>, in which case only scores of those sequences that
match the specified criteria will be recorded in the distance
file. This is particularly useful for database searches in which only
sequence IDs are of interest.

<P>
The <TT>select_align</TT> variable can be used in a similar way to
cause selected sequence alignments to be placed in the <TT>runname.a2m</TT> file.  Note that all selection variables use the same
<TT>NLLnull</TT>, <TT>NLLcomplex</TT>, and <TT>Emax</TT> thresholds, though
different combinations of the thresholds can be used by the different
parameters.

<P>
If the binary expansion of the <TT>many_files</TT> variable includes a
 '2' (e.g., 2, 3, 6, 7), the score information is sent to standard
output.
If the binary expansion of the <TT>many_files</TT> variable includes a
'4' (e.g., 4, 5, 6, 7), the multiple domain score information is sent
to standard output. (See Section&nbsp;<A HREF="node12.html#sec:multdomain">10.2.4</A>.)
If the binary expansion of the <TT>many_files</TT> variable includes a
'1' (e.g., 1, 3, 5, 7), <TT>buildmodel</TT> will create multiple files.
See first two options enable UNIX pipe processing, such as:
<PRE>

<code>hmmscore tests -i test.mod -db trna10.seq | awk  '$1 !~ % { print $5 "\t" $1}'</code>
</PRE>
to print a file with an evalue and an identifier on each line after
removing the comment lines.  If both options are in force (e.g., 6,
7), both the multiple domain and the simple scoring values will be
sent to standard output in an undefined order.
In a terminal window, you will also see
<TT>hmmscore</TT>'s standard error output of various diagnostic messages.

<P>

<H3><A NAME="SECTION000122300000000000000"></A><A NAME="sec:fragments"></A>
<BR>
10.2.3 Scoring Fragments
</H3>

<P>
Consider a model of 100 nodes and a fragment of 25 that very closely
matches some contiguous section of the model.  Even though that
section would score very well, the overall score of the fragment could
be quite poor because of its need to use 75 delete states in the
model.  The problem here is that in addition to modeling conserved
regions, the model also models the length of the conserved region.

<P>
SAM has two scoring options for dealing with fragments.  The first
option allows a sequence to start matching the model at any location
(rather than only the begin node) and end at any location (rather than
only the end node).  This will improve scoring for short sequences
that match a segment of the model.  This option can be turned on by
setting <TT>hmmscore</TT>'s <TT>SW</TT> variable to one.

<P>
The second option is similar to Smith and Waterman method of sequence
comparison, which will find the best score for any pair of
subsequences within two sequences.  The same can be done with models,
allowing a submodel to match a subsequence.
In <TT>hmmscore</TT>, this type of scoring can be specified by setting
the <TT>SW</TT> variable to 2.  When this is done, sequences can
jump from the initial module (presumably a FIM) into the delete state
of any module in the model, and can also jump out of the delete state
of any module within the model to the delete state of the next-to-last
node.  The first and next-to-last module are assumed to be FIMs, hence
the rational is that a sequence will use the FIM for some period of
time to consume characters that do not match the model, then the
sequence will jump to the model node corresponding to the start of the
fragment, use several model nodes, and then jump to the ending FIM to
consume the rest of the sequence.  

<P>
The probability of these jumps is set by the variables <TT>jump_in_prob</TT> and <TT>jump_out_prob</TT>, both of which have a
default value of unity.  That is, as in the sequence-to-sequence Smith
and Waterman,  there is no cost associated with jumping in and out of
the model.

<P>
The file  <TT>trna1frag.seq</TT> contains several sequences that contain 
part or all of  TRNA1.  The sequence include TRNA1 (72 bases, TRNA1Long (the
complete tRNA with additional characters), Long (58 base segment)
TRNA), Medium (34 base segment), Short (6 base segment TRNA1), and
AAMediumA, an embedding of Medium within several segments of As to
bring it to 176 characters.  Additionally, the file contains several
(obviously) non-tRNAs of various lengths, all of whose IDs begin with
the word `Not'.

<P>
When this file is scored using <TT>hmmscore</TT> with <TT>SW</TT> set to 0
for global alignment, the scores of the full 
sequence and the long fragment place them clearly as tRNAs.  However,
the score of the short and medium fragments are greatly penalized by
the large number of delete states they must use.

<P>
<PRE>

TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-36.70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-51.91&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.43e-22	:the&nbsp;original&nbsp;sequence&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;94&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-17.83&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-40.84&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.21e-17	:TRNA1&nbsp;with&nbsp;flanking&nbsp;As&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;58&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-11.32&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-27.71&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.11e-11	:A&nbsp;long&nbsp;partial&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;19.38&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;medium-length&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;26.73&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;of&nbsp;equal&nbsp;length&nbsp;to&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
MediumReverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30.84&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:The&nbsp;reversal&nbsp;of&nbsp;medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;109&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;42.35&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;longer&nbsp;than&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;13&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;44.27&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;short&nbsp;segment&nbsp;that&nbsp;is&nbsp;not&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
ShortReverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;52.36&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;short&nbsp;reverse&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;52.95&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;short&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;176&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;63.45&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:The&nbsp;medium&nbsp;segment&nbsp;with&nbsp;long&nbsp;flanking&nbsp;A&nbsp;regions&nbsp;&nbsp;&nbsp;
NotExtraLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;438&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;238.28&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;much&nbsp;longer&nbsp;than&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</PRE>
(Several of the sequences had simple null model scores worse than
<TT>simple_threshold</TT>, so their reverse null model scores were not
calculated.)

<P>
When scoring is performed using the <TT>SW</TT> option set to 1,
the following score file is generated, which places even the short
fragment in the possible tRNA range.
<PRE>

TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-33.84&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-46.23&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.00e-19	:the&nbsp;original&nbsp;sequence&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;94&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-13.46&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-39.05&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.32e-16	:TRNA1&nbsp;with&nbsp;flanking&nbsp;As&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;58&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-20.85&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.16&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;7.11e-12	:A&nbsp;long&nbsp;partial&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-9.06&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-14.81&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.43e-06	:A&nbsp;medium-length&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
ShortReverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.32&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;short&nbsp;reverse&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.86&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;short&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;13&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.71&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;short&nbsp;segment&nbsp;that&nbsp;is&nbsp;not&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
MediumReverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;5.76&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:The&nbsp;reversal&nbsp;of&nbsp;medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;25.91&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;of&nbsp;equal&nbsp;length&nbsp;to&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;109&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;45.06&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;longer&nbsp;than&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;176&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;68.06&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:The&nbsp;medium&nbsp;segment&nbsp;with&nbsp;long&nbsp;flanking&nbsp;A&nbsp;regions&nbsp;&nbsp;&nbsp;
NotExtraLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;438&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;244.05&nbsp;&nbsp;1000000.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;much&nbsp;longer&nbsp;than&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</PRE>

<P>
When scoring is performed using the <TT>SW</TT> option set to 2, the
following score file is generated, which also picks up the sequence
AAMediumAA, which is a segment of a tRNA embedded within a longer sequence.

<P>
<PRE>

TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;94&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-35.95&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.53&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.87e-12	:TRNA1&nbsp;with&nbsp;flanking&nbsp;As&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-35.84&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.51&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.97e-12	:the&nbsp;original&nbsp;sequence&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;58&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-20.64&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.30e-08	:A&nbsp;long&nbsp;partial&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;176&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-18.11&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-10.56&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.11e-04	:The&nbsp;medium&nbsp;segment&nbsp;with&nbsp;long&nbsp;flanking&nbsp;A&nbsp;regions&nbsp;&nbsp;&nbsp;
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-15.48&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-8.62&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.16e-03	:A&nbsp;medium-length&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
ShortReverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-5.76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.09&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;5.74e+00	:A&nbsp;short&nbsp;reverse&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;13&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-6.70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.00e+00	:A&nbsp;short&nbsp;segment&nbsp;that&nbsp;is&nbsp;not&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;109&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.38&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.00e+00	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;longer&nbsp;than&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotExtraLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;438&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.43&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.00e+00	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;much&nbsp;longer&nbsp;than&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.00e+00	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;of&nbsp;equal&nbsp;length&nbsp;to&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-5.67&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.09&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.26e+00	:A&nbsp;short&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
MediumReverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-6.86&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;8.62&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:The&nbsp;reversal&nbsp;of&nbsp;medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</PRE>

<P>
The unclear separation in this file between the best non-tRNA and the
worst tRNA segment is because local alignment allows sequences to use
only the parts of the model that improve its score.  Thus, a long
random sequence will have a better score than a shorter random
sequence.  Better scores are gained by using the more expensive (to
calculate) reverse sequence null model.  Setting <TT>subtract_null</TT>
to 4 will differentiate the tRNAs, excepting the 6-nucleotide
sequence, from the non-tRNAs.  (In this contrived example, the long
sequences are composed of the single letter A, and thus the sequence
and the reverse sequence score the same going through the HMM.)

<P>
<PRE>

TRNA1Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;94&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-35.95&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.53&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.87e-12	:TRNA1&nbsp;with&nbsp;flanking&nbsp;As&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-35.84&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.51&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.97e-12	:the&nbsp;original&nbsp;sequence&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Long&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;58&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-20.64&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.30e-08	:A&nbsp;long&nbsp;partial&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
AAMediumAA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;176&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-18.11&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-10.56&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.11e-04	:The&nbsp;medium&nbsp;segment&nbsp;with&nbsp;long&nbsp;flanking&nbsp;A&nbsp;regions&nbsp;&nbsp;&nbsp;
Medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-15.48&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-8.62&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.16e-03	:A&nbsp;medium-length&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
ShortReverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-5.76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.09&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;5.74e+00	:A&nbsp;short&nbsp;reverse&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotShort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;13&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-6.70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.00e+00	:A&nbsp;short&nbsp;segment&nbsp;that&nbsp;is&nbsp;not&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;109&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.38&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.00e+00	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;longer&nbsp;than&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
NotExtraLong&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;438&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.43&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.00e+00	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;much&nbsp;longer&nbsp;than&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Not&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;73&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.00e+00	:A&nbsp;string&nbsp;of&nbsp;As&nbsp;of&nbsp;equal&nbsp;length&nbsp;to&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Short&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-5.67&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.09&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;6.26e+00	:A&nbsp;short&nbsp;segment&nbsp;from&nbsp;TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
MediumReverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;34&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-6.86&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;8.62&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.20e+01	:The&nbsp;reversal&nbsp;of&nbsp;medium&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</PRE>

<P>
Significance levels for the simple null model, especially for the
second option, change greatly.  In the first option, because sequences
can start at any location (e.g., the initial FIM) and jump out of any
location (e.g., also the initial FIM), no sequence will have scores
worse than zero -- even non-family members will have a negative
NLL<IMG
 WIDTH="18" HEIGHT="31" ALIGN="MIDDLE" BORDER="0"
 SRC="img13.png"
 ALT="$-$">NULL score.  However, the significance level will be similar to
that of standard scoring.

<P>
In the second option, the number of placements of a sequence to the
model is essentially the number of starting points of the sequence
plus the number of exit points once the sequence has started using the
core of the model.  That is, sequences start in the
initial FIM, may at any time jump anywhere into the model, and then
jump out again latter.  The effect seen in the significance level
depends on both the sequence length and the model length, and for
comparison between different models, must be done to the scores
themselves as they are being generated.  If the <TT>adjust_score</TT>
variable is set to 1 and <TT>SW</TT>=2, all simple null model scores will 
have added to them the log of the sum of sequence and model length.  If
<TT>adjust_score</TT> is set to 2 (the default) and <TT>SW</TT>=1 or 2, 
then all scores will have added to them the log of the sequence length.  
In the future, the <TT>adjust_score</TT> parameter may be  refined as we 
further explore the length dependence of scores.  The adjustment is
not sufficient for models that have internal FIMs.
See Section&nbsp;<A HREF="node12.html#sec:nll_null">10.2.1</A>.

<P>

<H3><A NAME="SECTION000122400000000000000"></A><A NAME="sec:multdomain"></A>
<BR>
10.2.4 Selecting multiple domain alignments
</H3>

<P>
The <TT>hmmscore</TT> program also can create multiple-domain alignments
and score files from selected sequences.  Prior to Version 2.1, this
feature was called the <TT>multdomain</TT> program.  To enable this
option, the <TT>select_mdalign</TT> parameter is set in a manner
similar to other selection parameters.  See Section&nbsp;<A HREF="node12.html#sec:select">10.2.2</A>.

<P>
For each selected sequence, the multiple domain search procedure will
locate copies of a single motif within each selected sequence.  A user
specified <TT>select_md</TT> selection variable, along with thresholds
<TT>mdNLLnull</TT>, <TT>mdNLLcomplex</TT>, and <TT>mdEmax</TT> is the
criterion by which a subsequence is judged to be a match to the model.
If <TT>select_md</TT> is 1, whenever an <TT>mdNLLnull</TT> simple null
model score or better is achieved, a match to the model has been
found, and so forth.  The default <TT>select_md</TT> value of 4 uses
<TT>mdEmax</TT> (default of 0.01) to select subsequence matches.
The <TT>adpstyle</TT> paramater determines whether Viterbi (1, default),
posterior-decoded with transitions and emissions (4), or
posterior-decoded with only character emissions (5) alignment is
used. 

<P>
Once this match is found, it is cut from the sequence and another
match is looked for.  The process terminates when no matches scoring
better than the selection criteria are found.  Note that the multiple
domain scoring procedure always uses Viterbi scoring.  Thus, it is
theoretically possible for a sequence to be selected by <TT>hmmscore</TT>
for multiple domain search but for no domain to be found even if the
selection criteria are the same.

<P>
The output is similar to that of <TT>align2model</TT>, except that for
each match the sequence ID is modified to indicate where in the
sequence the match occurred.  Additionally, all letters in the
sequence that are part of the match are capitalized.  Unlike <TT>align2model</TT>, multiple domain search sequence output does not include
periods (`.') as spacers: <TT>prettyalign</TT> must be used to correctly
space the multiple alignment.

<P>
As an example, the file <TT>multtrna.seq</TT> contains two sequences, each
of which  contains two tRNA motifs.  By invoking <TT>multdomain</TT> in
the following way  (selecting all sequences for a multiple domain)
<PRE>

hmmscore multtrna -i testf.mod -db multtrna.seq -select_mdalign 8 -sw 2
prettyalign multtrna.mult -l90 &gt; multtrna.pretty
</PRE>

<P>
the file <TT>multtrna.pretty</TT> generated is

<P>
<PRE>

;&nbsp;&nbsp;SAM:&nbsp;prettyalign&nbsp;v3.3.1&nbsp;(December&nbsp;20,&nbsp;2001)&nbsp;compiled&nbsp;12/27/01_15:15:58
;&nbsp;&nbsp;(c)&nbsp;1992-2001&nbsp;Regents&nbsp;of&nbsp;the&nbsp;University&nbsp;of&nbsp;California,&nbsp;Santa&nbsp;Cruz
;
;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence&nbsp;Alignment&nbsp;and&nbsp;Modeling&nbsp;Software&nbsp;System
;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;http://www.cse.ucsc.edu/research/compbio/sam.html
;
;&nbsp;---------&nbsp;Citations&nbsp;(SAM,&nbsp;SAM-T99,&nbsp;HMMs)&nbsp;----------
;&nbsp;R.&nbsp;Hughey,&nbsp;A.&nbsp;Krogh,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;sequence&nbsp;analysis:
;&nbsp;&nbsp;Extension&nbsp;and&nbsp;analysis&nbsp;of&nbsp;the&nbsp;basic&nbsp;method,&nbsp;CABIOS&nbsp;12:95-107,&nbsp;1996.
;&nbsp;K.&nbsp;Karplus,&nbsp;C.&nbsp;Barrett,&nbsp;R.&nbsp;Hughey,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;detecting
;&nbsp;&nbsp;remote&nbsp;protein&nbsp;homologies,&nbsp;Bioinformatics&nbsp;14(10):846-856,&nbsp;1998.
;&nbsp;A.&nbsp;Krogh&nbsp;et&nbsp;al.,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;in&nbsp;computational&nbsp;biology:
;&nbsp;&nbsp;Applications&nbsp;to&nbsp;protein&nbsp;modeling,&nbsp;JMB&nbsp;235:1501-1531,&nbsp;Feb&nbsp;1994.
;&nbsp;-----------------------------------
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA12_88:163&nbsp;ugcuaggggauguagcucagugguagagcgcaugcuucgcauguaugaggccccgg
TRNA12_6:77&nbsp;&nbsp;&nbsp;ugcua...................................................
TRNA34_98:172&nbsp;gcuagcguaggcccuguggcuagcuggucaaagcgccugucuaguaaacaggagau
TRNA34_10:85&nbsp;&nbsp;gcuagcgua...............................................

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA12_88:163&nbsp;guucgauccccggcaucuccaguacugcguu..........GCGGCCGUCGUCUAG
TRNA12_6:77&nbsp;&nbsp;&nbsp;.........................................GGGGAUGUAGCUCAG
TRNA34_98:172&nbsp;ccuggguucgaaucccagcggggccuccagcauaguuugacGGGCGAAUAGUGUCA
TRNA34_10:85&nbsp;&nbsp;.........................................GGCCCUGUGGC-UAG

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;7
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA12_88:163&nbsp;UCUGGAuUAGGACGCUGGCCUCCCAAGCCAGCAAUCCcGGGUUCGAAUCCCGGCGG
TRNA12_6:77&nbsp;&nbsp;&nbsp;-UGGUA.GAGCGCA-UGCUUCGCAUGUAUGAGGCCCcGGGUUCGAUCCCCGGCAU
TRNA34_98:172&nbsp;GCGGG-.-AGCACACCAGACUUGCAAUCUGGUAGGGA.GGGUUCGAGUCCCUCUUU
TRNA34_10:85&nbsp;&nbsp;CUGGU-.CAAAGCGCCUGUCUAGUAAACAGGAGAUCCuGGGUUCGAAUCCCAGCGG

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA12_88:163&nbsp;CCGCA--ucgcuu.........................................
TRNA12_6:77&nbsp;&nbsp;&nbsp;CUCCA--guacugcguugcggccgucgucuagucuggauuaggacgcuggccuc
TRNA34_98:172&nbsp;GUCCACCA-guacguagauccgcggc..............................
TRNA34_10:85&nbsp;&nbsp;GGCCUCCA-gcauaguuugacgggcgaauagugucagcgggagcacaccagacuug

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA12_88:163&nbsp;........................................................
TRNA12_6:77&nbsp;&nbsp;&nbsp;ccaagccagcaaucccggguucgaaucccggcggccgcaucgcuu...........
TRNA34_98:172&nbsp;........................................................
TRNA34_10:85&nbsp;&nbsp;caaucugguagggaggguucgagucccucuuuguccaccaguacguagauccgcgg

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA12_88:163&nbsp;..
TRNA12_6:77&nbsp;&nbsp;&nbsp;..
TRNA34_98:172&nbsp;..
TRNA34_10:85&nbsp;&nbsp;c.

<P>
</PRE>

<P>
This file shows the matching area of the sequence within several
copies of the sequence.  Even though there is an extra FIM state in
the model,  the required ending delete state is automatically removed
from the alignment because <TT>auto_fim</TT> is at its default value of 1.

<P>
If the variable <TT>alignshort</TT> is set to zero or higher, matching
segments of the sequence are clipped, with <TT>alignshort</TT> positions
shown on either side.  The IDs are the same as if complete sequences
are printed, corresponding to the starting and ending points of the
motif within the original sequence.  Depending on how large <TT>alignshort</TT> is, the subsequences may overlap.  For example, the
commands
<PRE>

multdomain multtrnas -i testf.mod -db multtrna.seq -alignshort 3
prettyalign multtrnas.mult -l90 &gt; multtrnas.pretty
</PRE>
produce the alignment
<PRE>

;&nbsp;&nbsp;SAM:&nbsp;prettyalign&nbsp;v3.3.1&nbsp;(December&nbsp;20,&nbsp;2001)&nbsp;compiled&nbsp;12/27/01_15:15:58
;&nbsp;&nbsp;(c)&nbsp;1992-2001&nbsp;Regents&nbsp;of&nbsp;the&nbsp;University&nbsp;of&nbsp;California,&nbsp;Santa&nbsp;Cruz
;
;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence&nbsp;Alignment&nbsp;and&nbsp;Modeling&nbsp;Software&nbsp;System
;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;http://www.cse.ucsc.edu/research/compbio/sam.html
;
;&nbsp;---------&nbsp;Citations&nbsp;(SAM,&nbsp;SAM-T99,&nbsp;HMMs)&nbsp;----------
;&nbsp;R.&nbsp;Hughey,&nbsp;A.&nbsp;Krogh,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;sequence&nbsp;analysis:
;&nbsp;&nbsp;Extension&nbsp;and&nbsp;analysis&nbsp;of&nbsp;the&nbsp;basic&nbsp;method,&nbsp;CABIOS&nbsp;12:95-107,&nbsp;1996.
;&nbsp;K.&nbsp;Karplus,&nbsp;C.&nbsp;Barrett,&nbsp;R.&nbsp;Hughey,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;detecting
;&nbsp;&nbsp;remote&nbsp;protein&nbsp;homologies,&nbsp;Bioinformatics&nbsp;14(10):846-856,&nbsp;1998.
;&nbsp;A.&nbsp;Krogh&nbsp;et&nbsp;al.,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;in&nbsp;computational&nbsp;biology:
;&nbsp;&nbsp;Applications&nbsp;to&nbsp;protein&nbsp;modeling,&nbsp;JMB&nbsp;235:1501-1531,&nbsp;Feb&nbsp;1994.
;&nbsp;-----------------------------------
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;
TRNA12_88:163&nbsp;guuGCGGCCGUCGUCUAGUCUGGAuUAGGACGCUGGCCUCCCAAGCCAGCAAUCCc
TRNA12_6:77&nbsp;&nbsp;&nbsp;cuaGGGGAUGUAGCUCAG-UGGUA.GAGCGCA-UGCUUCGCAUGUAUGAGGCCCc
TRNA34_98:172&nbsp;gacGGGCGAAUAGUGUCAGCGGG-.-AGCACACCAGACUUGCAAUCUGGUAGGGA.
TRNA34_10:85&nbsp;&nbsp;guaGGCCCUGUGGC-UAGCUGGU-.CAAAGCGCCUGUCUAGUAAACAGGAGAUCCu

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA12_88:163&nbsp;GGGUUCGAAUCCCGGCGGCCGCA--ucg.
TRNA12_6:77&nbsp;&nbsp;&nbsp;GGGUUCGAUCCCCGGCAUCUCCA--gua.
TRNA34_98:172&nbsp;GGGUUCGAGUCCCUCUUUGUCCACCA-gua.
TRNA34_10:85&nbsp;&nbsp;GGGUUCGAAUCCCAGCGGGGCCUCCA-gca.

<P>
</PRE>

<P>
In addition to <TT>multtrna.mult</TT>, the file <TT>multtrna.mstat</TT> is
produced.  It reports the NLL-NULL score for each of the motifs listed
in <TT>multtrna.mult</TT>.  In this case, <TT>multtrna.mstat</TT> (or <TT>multtrnas.mstat</TT>) looks like

<P>
<PRE>

%&nbsp;&nbsp;SAM:&nbsp;hmmscore&nbsp;v3.3.1&nbsp;(December&nbsp;20,&nbsp;2001)&nbsp;compiled&nbsp;12/27/01_15:14:20
%&nbsp;&nbsp;(c)&nbsp;1992-2001&nbsp;Regents&nbsp;of&nbsp;the&nbsp;University&nbsp;of&nbsp;California,&nbsp;Santa&nbsp;Cruz
%
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence&nbsp;Alignment&nbsp;and&nbsp;Modeling&nbsp;Software&nbsp;System
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;http://www.cse.ucsc.edu/research/compbio/sam.html
%
%&nbsp;---------&nbsp;Citations&nbsp;(SAM,&nbsp;SAM-T99,&nbsp;HMMs)&nbsp;----------
%&nbsp;R.&nbsp;Hughey,&nbsp;A.&nbsp;Krogh,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;sequence&nbsp;analysis:
%&nbsp;&nbsp;Extension&nbsp;and&nbsp;analysis&nbsp;of&nbsp;the&nbsp;basic&nbsp;method,&nbsp;CABIOS&nbsp;12:95-107,&nbsp;1996.
%&nbsp;K.&nbsp;Karplus,&nbsp;C.&nbsp;Barrett,&nbsp;R.&nbsp;Hughey,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;detecting
%&nbsp;&nbsp;remote&nbsp;protein&nbsp;homologies,&nbsp;Bioinformatics&nbsp;14(10):846-856,&nbsp;1998.
%&nbsp;A.&nbsp;Krogh&nbsp;et&nbsp;al.,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;in&nbsp;computational&nbsp;biology:
%&nbsp;&nbsp;Applications&nbsp;to&nbsp;protein&nbsp;modeling,&nbsp;JMB&nbsp;235:1501-1531,&nbsp;Feb&nbsp;1994.
%&nbsp;-----------------------------------
%&nbsp;&nbsp;multtrna&nbsp;&nbsp;&nbsp;Host:&nbsp;perch&nbsp;&nbsp;&nbsp;&nbsp;Thu&nbsp;Dec&nbsp;27&nbsp;16:03:16&nbsp;2001
%&nbsp;&nbsp;rph&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Dir:&nbsp;&nbsp;/projects/kestrel/rph/sam32/SAMBUILD/perch/demos
%&nbsp;-----------------------------------
%&nbsp;Inserted&nbsp;Files:&nbsp;&nbsp;testf.mod
%&nbsp;Database&nbsp;Files:&nbsp;&nbsp;/projects/kestrel/rph/sam32/demos/multtrna.seq
%
%&nbsp;Motifcutoff&nbsp;0.500000
%&nbsp;Motifs&nbsp;selected&nbsp;from&nbsp;sequences&nbsp;into&nbsp;file&nbsp;multtrna.mult&nbsp;if&nbsp;better&nbsp;than:
%	&nbsp;E-value&nbsp;(mdEmax&nbsp;1.0e-02)&nbsp;(select_md=4)
%&nbsp;See&nbsp;related&nbsp;information&nbsp;in&nbsp;multtrna.dist
%
%&nbsp;Simple:&nbsp;NLL-NULL&nbsp;using&nbsp;FIM&nbsp;probabilities
%&nbsp;Reverse:&nbsp;NLL-NULL&nbsp;for&nbsp;the&nbsp;reverse&nbsp;sequence&nbsp;NULL&nbsp;model
%&nbsp;&nbsp;&nbsp;&nbsp;Calculated&nbsp;when&nbsp;Simple&nbsp;&lt;&nbsp;simple_threshold&nbsp;(0.00)
%&nbsp;E-value&nbsp;on&nbsp;N&nbsp;(=2)&nbsp;sequences:
%&nbsp;&nbsp;&nbsp;&nbsp;N&nbsp;/&nbsp;(1&nbsp;+&nbsp;exp(-(lambda(=1.0000)&nbsp;*&nbsp;Reverse)**tau(=1.0000)))
%&nbsp;&nbsp;&nbsp;&nbsp;Calculated&nbsp;when&nbsp;Simple&nbsp;&lt;&nbsp;simple_threshold&nbsp;(0.00)
%&nbsp;&nbsp;&nbsp;&nbsp;Rescale&nbsp;E-values&nbsp;or&nbsp;use&nbsp;-dbsize&nbsp;for&nbsp;multiple&nbsp;scoring&nbsp;runs.
%&nbsp;&nbsp;&nbsp;&nbsp;WARNING:&nbsp;&nbsp;E-VALUES&nbsp;ARE&nbsp;NOT&nbsp;CALIBRATED!
%&nbsp;Scores&nbsp;sorted&nbsp;by&nbsp;E-value,&nbsp;best&nbsp;first
%
%&nbsp;Sequence&nbsp;ID&nbsp;&nbsp;&nbsp;&nbsp;Length&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Simple&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;E-value&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;X&nbsp;count
TRNA12_88:163&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-33.28&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-31.41&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.58e-14
TRNA12_6:77&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-31.46&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-29.06&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.81e-13
TRNA34_98:172&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;75&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-31.21&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-27.94&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.46e-12
TRNA34_10:85&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-28.94&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-25.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.77e-11
</PRE>

<P>
Because reliable results are only obtained if FIMs are added to the
model, multiple domain searchers are best performed when <TT>auto_fim</TT> is set to 1 (the default).

<P>
If the binary expansion of the <TT>many_files</TT> variable includes a
'4' (e.g., 4, 5, 6, 7), the multiple domain score information (.mstat)
is sent to standard output instead. See Section&nbsp;<A HREF="node12.html#sec:select">10.2.2</A>.

<P>

<H3><A NAME="SECTION000122500000000000000"></A><A NAME="sec:multitrack"></A>
<BR>
10.2.5 Multi-track HMM scoring
</H3>

<P>
Suppose one wanted to model a group of protein sequences with
associated secondary structure information.  The secondary structure
labels are important information that could lead to better modelling.
To enable such analysis (without the need, for example, of a 60
character alphabet of each amino acid with each secondary structure
type), SAM can process multi-track sequences and multi-track HMMs.

<P>
A multi-track sequence is specified by a pair or set of sequence
files.  The first sequence file contains the first track of data
(amino acid sequences in this case), while the second sequence file
contains the second track of data (secondary structure) with each
sequence corresponding exactly in length, identifier, and position to
a sequence in the first file.  The sequences are specified as a
comma-separated list with the <TT>db</TT> paramater.

<P>
Multi-track HMMs are similarly paired or sets of SAM HMM files.  The
models are specified as a comma-separated list for the <TT>-trackmod</TT> filename list variable.  During the dynamic programming
calculation, all transition probabilities are taken from the first or
primary track, while character emission probabilities are taken from
the joint (product) probability of the first track character being
emitted from the given first track HMM state, and the second track
character being emitted from the given second track HMM state, and so
on.  All track models must, of course, be of the same length.

<P>
By default, the tracks are all given unity weight, so that, ignoring
transition probabilities, a 2-track model will produce scores of about
twice the magnitude of a 1-track model.  To compensate for this, each
track can also be given a character emission coefficient with <TT>trackcoeff</TT>.  The scores (log-probabilities) of the match and insert
states are scaled by this (<IMG
 WIDTH="43" HEIGHT="31" ALIGN="MIDDLE" BORDER="0"
 SRC="img69.png"
 ALT="$\geq 0.0$">) value, so that a 2-track model with
identical tracks and coefficients of 0.5 will produce the same scores
(approximately, due to rounding) as a one-track model, such as with
the following two commands:
<PRE>

hmmscore t2  -trackmod test.mod,test.mod -db trna10.seq,trna10.seq
             -a RNA,RNA -trackcoeff 0.5,0.5
hmmscore t1 -i test.mod -db trna10.seq
</PRE>

<P>
Multi-track models are easy to create with external software scripts.

<P>
The additional tracks have the same SW mode preprocessing (i.e.,
possibly adding FIMs), but never have their insert or FIM character
tables changed (i.e., <TT>Insert_method_score</TT> and <TT>FIM_method_score</TT> are ignored for tracks other than the first).  If
FIMs are added to the model either the character table of the generic
node is used (when present) or the geometric match state average (when
there is no generic node).

<P>

<H3><A NAME="SECTION000122600000000000000"></A><A NAME="sec:segment"></A>
<BR>
10.2.6 Distributed scoring
</H3>

<P>
Scoring a large database with <TT>hmmscore</TT> can take several hours on
even the fastest workstation.  The scoring program includes primitive
support for distributed scoring.  If the <TT>segments</TT> variable is
set to an integer larger than 1, <TT>hmmscore</TT> assumes that that many
runs of <TT>hmmscore</TT> are being used to score a complete database.
The <TT>segment_number</TT> variable is used to label each segment.

<P>
These parameters might be used as follows:
<PRE>

hmmscore  test1 -i test.mod -db bigdatabase -segments 2 -segment_number 1 -sw 2&amp;
rsh othermachine hmmscore test2 -i test.mod -db bigdatabase -segments 2 -segment_number 2 -sw 2
wait cat test1.dist test2.dist &gt; test.dist
</PRE>

<P>
The associated parameter, <TT>segment_size</TT>, specifies that number
of sequences that are read in at a time.  At a value of 100,
two segments would produce the effect that the first 100 sequences are
processed by segment 1, the next 100 be segment 2, the next 100 by
segment 1, and so on.  Note that workload is partitioned according to
the number of sequences rather than the number of residues, some
segments may take longer to complete than other segments.

<P>
The <TT>segment_size</TT> parameter is important even if distributed
scoring is not being performed to reduce memory consumption by <TT>hmmscore</TT>.  Unfortunately, selection based on E-values requires
knowledge of the database size.  For this reason, if selection is
being performed, segmented file reading will be performed twice; the
first time to calculate the size of the database, in the second time
to perform the scoring. The <TT>dbsize</TT> variable can be set to
externally indicate the size of the database for calculating E-values.
When set, the first reading of the database is not performed.

<P>

<H3><A NAME="SECTION000122700000000000000"></A><A NAME="sec:smithwaterman"></A>
<BR>
10.2.7 Smith &amp; Waterman Scoring and
Alignment
</H3>

<P>
The <TT>hmmscore</TT> program can be used to perform Smith and Waterman
scoring, alignment, and multiple domain alignment.  This option is
achieved by internally creating a SAM model based on a single
query sequence, gap and continue penalties, and a scoring matrix.
These models cannot be stored, and can only be used with Viterbi
scoring.  Local or global alignment can be performed on this model by
appropriately setting <TT>sw</TT>.

<P>
The internal model zeros the appropriate transitions and uses an
apporpriately simplified dynamic programming calculation.  Its dynamic
programming speed is about the same as a dedicated Smith and Waterman
program, though the default reverse-sequence e-value calculation
doubles execution time.  

<P>
Smith and Waterman mode is specified by including a <TT>query</TT> file
on the command line.  The first sequence in this file is taken to be
the query sequence.  Additional parameters include <TT>gap</TT> and
<TT>continue</TT>, which are by default set to 12 and 1, respectively,
and <TT>matrix</TT>, set to ``blosum62.''  Matrices are read in BLAST
format. They are first looked for in the current working directory,
then in the directory pointed to by <TT>BLASTMAT</TT>, if set.  If not,
the subdirectories <TT>aa</TT> and <TT>nt</TT>, depending on alphabet, of
the environment variable <code>PRIOR_PATH</code> are checked, and finally,
the same subdirectories of the default prior path compiled into the
code. See Section&nbsp;<A HREF="node10.html#sec:priors">8.1.1</A>.

<P>
As with other forms of scoring, SAM is able to calculate e-values for
Smith and Waterman queries based on the reverse sequence in all model.
For this to be effective, the system must know the scaling factor for
comping E-values based on how the scoring matrix has been formed.  The
value <TT>swlambda</TT> should be set to <IMG
 WIDTH="56" HEIGHT="34" ALIGN="MIDDLE" BORDER="0"
 SRC="img70.png"
 ALT="$\ln(b)/u$">, where <IMG
 WIDTH="13" HEIGHT="18" ALIGN="BOTTOM" BORDER="0"
 SRC="img59.png"
 ALT="$b$"> is the
base of the matrix and <IMG
 WIDTH="31" HEIGHT="34" ALIGN="MIDDLE" BORDER="0"
 SRC="img71.png"
 ALT="$1/u$"> is the unit weight.  For example, the
Hennikoff &amp; Hennifkoff ``blosum62'' matrix distributed with BLAST and
included with SAM is in units of 1/2 bit, so <IMG
 WIDTH="42" HEIGHT="18" ALIGN="BOTTOM" BORDER="0"
 SRC="img72.png"
 ALT="$b=2$"> (bits are base 2)
and <IMG
 WIDTH="44" HEIGHT="17" ALIGN="BOTTOM" BORDER="0"
 SRC="img73.png"
 ALT="$u=2$">, since a `2' in the matrix corresponds one unit (a bit, in
this case).  The default value of <TT>lambda</TT> is thus
<!-- MATH
 $\ln(2)/2=0.34657$
 -->
<IMG
 WIDTH="129" HEIGHT="34" ALIGN="MIDDLE" BORDER="0"
 SRC="img74.png"
 ALT="$\ln(2)/2=0.34657$">.

<P>
<PRE>

hmmscore sw -query globins50 -db globins50              -matrix blosum62 -gap 12 -continue 1 
</PRE>
will  produce a score file using the first sequence of <TT>globins50</TT>
as a query and the first 10 sequence of <TT>globins50</TT> as a database.
<PRE>

%&nbsp;&nbsp;SAM:&nbsp;hmmscore&nbsp;v3.3.1&nbsp;(December&nbsp;20,&nbsp;2001)&nbsp;compiled&nbsp;12/27/01_15:14:20
%&nbsp;&nbsp;(c)&nbsp;1992-2001&nbsp;Regents&nbsp;of&nbsp;the&nbsp;University&nbsp;of&nbsp;California,&nbsp;Santa&nbsp;Cruz
%
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence&nbsp;Alignment&nbsp;and&nbsp;Modeling&nbsp;Software&nbsp;System
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;http://www.cse.ucsc.edu/research/compbio/sam.html
%
%&nbsp;---------&nbsp;Citations&nbsp;(SAM,&nbsp;SAM-T99,&nbsp;HMMs)&nbsp;----------
%&nbsp;R.&nbsp;Hughey,&nbsp;A.&nbsp;Krogh,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;sequence&nbsp;analysis:
%&nbsp;&nbsp;Extension&nbsp;and&nbsp;analysis&nbsp;of&nbsp;the&nbsp;basic&nbsp;method,&nbsp;CABIOS&nbsp;12:95-107,&nbsp;1996.
%&nbsp;K.&nbsp;Karplus,&nbsp;C.&nbsp;Barrett,&nbsp;R.&nbsp;Hughey,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;detecting
%&nbsp;&nbsp;remote&nbsp;protein&nbsp;homologies,&nbsp;Bioinformatics&nbsp;14(10):846-856,&nbsp;1998.
%&nbsp;A.&nbsp;Krogh&nbsp;et&nbsp;al.,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;in&nbsp;computational&nbsp;biology:
%&nbsp;&nbsp;Applications&nbsp;to&nbsp;protein&nbsp;modeling,&nbsp;JMB&nbsp;235:1501-1531,&nbsp;Feb&nbsp;1994.
%&nbsp;-----------------------------------
%&nbsp;&nbsp;sw&nbsp;&nbsp;&nbsp;Host:&nbsp;perch&nbsp;&nbsp;&nbsp;&nbsp;Thu&nbsp;Dec&nbsp;27&nbsp;16:03:16&nbsp;2001
%&nbsp;&nbsp;rph&nbsp;&nbsp;&nbsp;Dir:&nbsp;&nbsp;/projects/kestrel/rph/sam32/SAMBUILD/perch/demos
%&nbsp;-----------------------------------
%&nbsp;Inserted&nbsp;Files:&nbsp;
%&nbsp;Database&nbsp;Files:&nbsp;&nbsp;/projects/kestrel/rph/sam32/demos/globins50.seq
%
%&nbsp;Smith&nbsp;&amp;&nbsp;Waterman&nbsp;Query&nbsp;Sequence:&nbsp;&nbsp;BAHG$VITSP
%&nbsp;Cost&nbsp;matrix:&nbsp;&nbsp;/projects/kestrel/rph/sam32/lib/matrix/aa/blosum62
%&nbsp;Gap&nbsp;cost&nbsp;12&nbsp;&nbsp;&nbsp;&nbsp;Continue&nbsp;cost&nbsp;1
%&nbsp;DP&nbsp;Method:&nbsp;viterbi&nbsp;single-path&nbsp;(dpstyle&nbsp;=&nbsp;1)
%&nbsp;50&nbsp;sequences,&nbsp;7369&nbsp;residues,&nbsp;0&nbsp;nodes,&nbsp;0.41&nbsp;seconds
%
%&nbsp;Sequence&nbsp;scores&nbsp;selected:&nbsp;&nbsp;All&nbsp;&nbsp;(select_score=8)
%
%&nbsp;Simple:&nbsp;NLL-NULL&nbsp;using&nbsp;FIM&nbsp;probabilities
%&nbsp;Reverse:&nbsp;NLL-NULL&nbsp;for&nbsp;the&nbsp;reverse&nbsp;sequence&nbsp;NULL&nbsp;model
%&nbsp;&nbsp;&nbsp;&nbsp;Calculated&nbsp;when&nbsp;Simple&nbsp;&lt;&nbsp;simple_threshold&nbsp;(0.00)
%&nbsp;E-value&nbsp;on&nbsp;N&nbsp;(=50)&nbsp;sequences:
%&nbsp;&nbsp;&nbsp;&nbsp;N&nbsp;/&nbsp;(1&nbsp;+&nbsp;exp(-(swlambda(=0.3466)&nbsp;*&nbsp;Reverse)**tau(=1.0000)))
%&nbsp;&nbsp;&nbsp;&nbsp;Rescale&nbsp;E-values&nbsp;or&nbsp;use&nbsp;-dbsize&nbsp;for&nbsp;multiple&nbsp;scoring&nbsp;runs.
%&nbsp;&nbsp;&nbsp;&nbsp;WARNING:&nbsp;&nbsp;E-VALUES&nbsp;ARE&nbsp;NOT&nbsp;CALIBRATED!
%&nbsp;Scores&nbsp;sorted&nbsp;by&nbsp;E-value,&nbsp;best&nbsp;first
%
%&nbsp;Sequence&nbsp;ID&nbsp;&nbsp;&nbsp;Length&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Simple&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reverse&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;E-value&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;X&nbsp;count
BAHG$VITSP&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;146&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;734.00&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;755.00&nbsp;&nbsp;&nbsp;&nbsp;1.15e-112
</PRE>
The various <TT>select</TT> options (e.g., <TT>selectalign</TT>, <TT>selectmdalign</TT>, etc.) can also be used, according to Smith and
Waterman score (<TT>NLLnull</TT> and <TT>mdNLLnull</TT>), reverse sequence
(<TT>NLLcomplex</TT> and <TT>mdNLLcomplex</TT>, or e-value (<TT>Evalue</TT>
and <TT>mdEvalue</TT>) score.

<P>

<H3><A NAME="SECTION000122800000000000000"></A><A NAME="sec:hmmscore:kestrel"></A>
<BR>
10.2.8 Scoring using the Kestrel parallel processor
</H3>

<P>
The <TT>hmmscore</TT> program can use the UCSC  Kestrel parallel processor
to perform high-speed scoring of sequence databases.  

<P>
Currently <TT>hmmscore</TT> only supports global or local EM scoring on Kestrel
and models that have a final length, with added FIMs, of no more than 512
nodes. If any of these restrictions are not met, <TT>hmmscore</TT> reverts to
the sequential algorithm, unless <TT>kestrel_fallback</TT> is
set to <TT>0</TT>.

<P>
Kestrel scoring is enabled using the <TT>use_kestrel</TT> parameter.  Since
a request queue mechanism is not currently available as part of the 
Kestrel runtime environment, a sleep and retry mechanism is implemented
as part of <TT>hmmscore</TT>.  Retrying is controlled using the
<TT>kestrel_retry_cnt</TT> and <TT>kestrel_retry_time</TT> parameters.
The <TT>kestrel</TT> runtime environment program is executed to access
the Kestrel server and is found using the <TT>PATH</TT>.

<P>
Small models will generally score more quickly using the sequential
algorithm than with Kestrel.  A minimum model size for using Kestrel
may be specified with the <TT>kestrel_min_model_len</TT>
parameter. Models smaller than this will use sequential scoring.

<P>
Kestrel-format databases, created using _kestrel_db, are used by
<TT>hmmscore</TT>.  The original sequence databases are specified as the value of
the <TT>-db</TT> arguments to <TT>hmmscore</TT>.  The file names of the
Kestrel-specific database will be derived from the <TT>db</TT> file names by
appending the appropriate suffix.  If <TT>subtract_null</TT> is <TT>4</TT>, then
<TT>.krseq</TT> is used, otherwise <TT>.kseq</TT>.  Additionally, the <TT>.kids</TT>
file must exist in the same directory as the <TT>-db</TT> file.  The <TT>.kseq</TT>
or <TT>.krseq</TT> are found either in the <TT>-db</TT> file directory or in the
directory on the Kestrel server specified by kestrel_remote_db_dir.
See Section&nbsp;<A HREF="node12.html#sec:mkkestreldb">10.11.9</A>.

<P>

<H3><A NAME="SECTION000122900000000000000"></A><A NAME="sec:calibrate"></A>
<BR>
10.2.9 Calibration and Model libraries
</H3>

<P>
A SAM model library is a SAM parameter file that has been divided into
several sections, one for each item in the model library.  For a
model, each section includes either a <TT>modelfile</TT> or <TT>trackmod</TT>, though it is acceptable to actually include the model
definition in the library file. The typical parameters completely
specify the method of scoring, including dynamic programming mode
(<TT>dpstyle</TT>, <TT>sw</TT>, <TT>autoaddfims</TT>) and scoring parameters
(<TT>subtractnull</TT>, and <TT>swlambda</TT> or  <TT>emlambda</TT>).  Also,
each entry includes a name and a comment.  

<P>
Single-entry model libraries are created when the calibrate option of
<TT>hmmscore</TT> is used on a <TT>modelfile</TT> or a <TT>trackmod</TT>.  If a 
<TT>modellibrary</TT> is passed to <TT>hmmscore</TT>, the resulting <TT>runname.mlib</TT> file has the same number of models as the original model
library.  To create a multi-model library, individual libraries can be
concatenated, though you may wish to remove the comments (lines
beginning with  `<code>%</code>') for a more compact form.

<P>
SAM model library output will use absolute path names if the <TT>modlib_absolute</TT> variable is set to 1.  This should be done if model
libraries created in different directories are to be combined, and
absolute path names were not used on the command line.

<P>
All models in a model library must have the same number of sequence
tracks to facilitate comparison against a sequence database.

<P>
Suppose that model library <TT>test.mlib</TT> contains:
<PRE>

MODLIBMOD first Test.mod with sw 1
modelfile test.mod
dpstyle 0
sw 1
ENDMODLIBMOD

<P>
MODLIBMOD second Test.mod with sw 2
modelfile test.mod
dpstyle 0
sw 2
ENDMODLIBMOD
</PRE>

<P>
This <TT>test.mlib</TT> file can then be calibrated using <TT>hmmscore</TT>: 
<PRE>

hmmscore caltest -modellibrary test.mlib -calibrate 1
</PRE>

<P>
To produce the file
<PRE>

%&nbsp;&nbsp;SAM:&nbsp;hmmscore&nbsp;v3.3.1&nbsp;(December&nbsp;20,&nbsp;2001)&nbsp;compiled&nbsp;12/27/01_15:14:20
%&nbsp;&nbsp;(c)&nbsp;1992-2001&nbsp;Regents&nbsp;of&nbsp;the&nbsp;University&nbsp;of&nbsp;California,&nbsp;Santa&nbsp;Cruz
%
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence&nbsp;Alignment&nbsp;and&nbsp;Modeling&nbsp;Software&nbsp;System
%&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;http://www.cse.ucsc.edu/research/compbio/sam.html
%
%&nbsp;---------&nbsp;Citations&nbsp;(SAM,&nbsp;SAM-T99,&nbsp;HMMs)&nbsp;----------
%&nbsp;R.&nbsp;Hughey,&nbsp;A.&nbsp;Krogh,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;sequence&nbsp;analysis:
%&nbsp;&nbsp;Extension&nbsp;and&nbsp;analysis&nbsp;of&nbsp;the&nbsp;basic&nbsp;method,&nbsp;CABIOS&nbsp;12:95-107,&nbsp;1996.
%&nbsp;K.&nbsp;Karplus,&nbsp;C.&nbsp;Barrett,&nbsp;R.&nbsp;Hughey,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;detecting
%&nbsp;&nbsp;remote&nbsp;protein&nbsp;homologies,&nbsp;Bioinformatics&nbsp;14(10):846-856,&nbsp;1998.
%&nbsp;A.&nbsp;Krogh&nbsp;et&nbsp;al.,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;in&nbsp;computational&nbsp;biology:
%&nbsp;&nbsp;Applications&nbsp;to&nbsp;protein&nbsp;modeling,&nbsp;JMB&nbsp;235:1501-1531,&nbsp;Feb&nbsp;1994.
%&nbsp;-----------------------------------
%&nbsp;&nbsp;caltest&nbsp;&nbsp;&nbsp;Host:&nbsp;perch&nbsp;&nbsp;&nbsp;&nbsp;Thu&nbsp;Dec&nbsp;27&nbsp;16:05:16&nbsp;2001
%&nbsp;&nbsp;rph&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Dir:&nbsp;&nbsp;/projects/kestrel/rph/sam32/SAMBUILD/perch/demos
%&nbsp;-----------------------------------
%&nbsp;Inserted&nbsp;Files:&nbsp;
%&nbsp;Database&nbsp;Files:&nbsp;
%

<P>
MODLIBMOD&nbsp;first&nbsp;Test.mod&nbsp;with&nbsp;sw&nbsp;1
%&nbsp;Calibration&nbsp;data&nbsp;for&nbsp;1000&nbsp;random&nbsp;sequences
%&nbsp;1000&nbsp;of&nbsp;1000&nbsp;scores&nbsp;&gt;&nbsp;calibrate_threshold&nbsp;(&nbsp;-1e+20)&nbsp;used&nbsp;(default).
model_file&nbsp;test.mod
lambda&nbsp;0.574107
tau&nbsp;1.060932
sw&nbsp;1
dpstyle&nbsp;0
jumpinprob&nbsp;1.000000
jumpoutprob&nbsp;1.000000
subtractnull&nbsp;4
FIMmethodscore&nbsp;-6
insertmethodscore&nbsp;0
adjust_score&nbsp;2
ENDMODLIBMOD

<P>
MODLIBMOD&nbsp;second&nbsp;Test.mod&nbsp;with&nbsp;sw&nbsp;2
%&nbsp;Calibration&nbsp;data&nbsp;for&nbsp;1000&nbsp;random&nbsp;sequences
%&nbsp;1000&nbsp;of&nbsp;1000&nbsp;scores&nbsp;&gt;&nbsp;calibrate_threshold&nbsp;(&nbsp;-1e+20)&nbsp;used&nbsp;(default).
model_file&nbsp;test.mod
lambda&nbsp;1.909900
tau&nbsp;0.835846
sw&nbsp;2
dpstyle&nbsp;0
jumpinprob&nbsp;1.000000
jumpoutprob&nbsp;1.000000
subtractnull&nbsp;4
FIMmethodscore&nbsp;-6
insertmethodscore&nbsp;0
adjust_score&nbsp;2
ENDMODLIBMOD
</PRE>

<P>
As can be seen, model calibration has added additional information to
the model library.  The most important is the setting of the <TT>lambda</TT> parameter for E-value calculation.

<P>
If any of the parameters shown in the resulting model library are
changed, callibration must be performed again.  It is important to
note that models must be calibrated separately for all differences in
scoring method.

<P>
The process of model calibration involves generating random sequences
(for protein sequences, a dirichlet mixture distribution is used) and
scoring them against the HMM.  This can be a time-consuming process,
but once calculated <TT>hmmscore</TT> will produce much more accurate
E-values.  

<P>
The only required  calibration parameter include <TT>calibrate</TT>, which
specifies the number of random sequences to score (if <TT>calibrate</TT>
is set to `1', an internal default is used).  Optionally,
<TT>trackprior</TT> specifies a list (one per track) of Dirichlet
mixtures over sequence composition, <TT>genprot_prior</TT> indicates the
default Dirichlet mixture prior for protein sequences,
<TT>genehl2_prior</TT> indicates the
default Dirichlet mixture prior for the EHL2 alphabet, <TT>gs_mean_log_len</TT> is the natural logarithm of the mean sequence length to
generate, and <TT>gs_sd_log_len</TT> is the natural logarithm of the
standard deviation of the synthetic sequence length distribution.
If a Dirichlet mixture is not available for one or more of the
sequence tracks, the characters are drawn according to SAM's internal
default background frequencies with no variation in distribution
between sequences.

<P>
We are reasonably happy with our single-track model calibration
method.  Our random sequence generators for multi-track model
calibration are not appropriately linked, and thus do not effectively
calibrate multi-track models, such as those with amino acid and
secondary structure sequences.  This is currently an active research
area, and we hope to improve the calibration method for these models
in the next release.

<P>

<H2><A NAME="SECTION000123000000000000000"></A> <A NAME="sec:addfims"></A>
<BR>
10.3 <TT>addfims</TT>
</H2>

<P>
The <TT>addfims</TT> program can be used to add Free Insertion Modules
to the beginning and end of a model.
The program <TT>modifymodel</TT> can also be used for this purpose, and
`O' characters in <TT>modelfromalign</TT> also produce FIMs.  See Section&nbsp;<A HREF="node10.html#sec:fims">8.5</A>.
This is particularly useful if a model has been trained on a clipped
sequence motif, and is to be used in analyzing full sequences.

<P>
For example, the file <TT>trna10f.seq</TT> is the same as <TT>trna10.seq</TT>, except that sequences 2-5 have had extraneous characters
appended to one or both ends.  Using the <TT>test.mod</TT> file
previously generated, an alignment of the first 5 sequences looks like
this:
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
TRNA1&nbsp;&nbsp;...GGGGAUGUAGCUCAG-.................UGGUA.............GAGCGCA-
TRNA2X&nbsp;...GCGGCCGUCGUCUAGUgcggccgucgucuagucUGGAU.............UAGGACGCU
TRNA3X&nbsp;cccGGCCCUGUGGC-UAGC.................UGGU-.............CAAAGCGCC
TRNA4X&nbsp;...GGGCGAAUAGUGUCAG.................GGCGAauagugucagcggGAGCACACC
TRNA5X&nbsp;...GCCGGGAUAGCUCAGU.................UGGUA.............GAGCAGA-

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;UGCUUCGCAUGUAU.GAGGCCCcGGGUUCGAUCCC.........C--GGCAUCUC.......
TRNA2X&nbsp;GGCCUCCCAAGCCA.GCAAUCCcGGGUUCGAAUCC.........CGGCGGCCGCACggcggcc
TRNA3X&nbsp;UGUCUAGUAAACAG.GAGAUCCuGGGUUCGAAUCC.........CAGCGGGGCCUC.......
TRNA4X&nbsp;AGACUUGCAAUCUG.GUAGGGA.GGGUUCGAGUCC.........CUCUUUGUCCAC.......
TRNA5X&nbsp;GGACUGAAAAUCCUcGUGUCAC.CAGUUCAAAUCUgguuccuggCAUGGUUCCUGG.......

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;.CA.....
TRNA2X&nbsp;gCA.....
TRNA3X&nbsp;.CAgggg.
TRNA4X&nbsp;.CA.....
TRNA5X&nbsp;.CA.....

<P>
</PRE>
This alignment is incorrect:  the extra end characters do not all use
end insert states.  Rather, internal insert states are found to
minimize the alignment cost for sequences 2X, 4X, and 5X.  

<P>
The <TT>addfims</TT> program has an interface identical to that of <TT>buildmodel</TT>, though only the runname, model file (and its model, or if
not present, its regularizer), and the alphabet are used:
<PRE>

addfims testf -insert test.mod
align2model testf -i testf.mod -db trna10f.seq
prettyalign testf.a2m -l90 &gt; testf.align
</PRE>
will produce the following alignment, which correctly places the
extra characters at the ends.
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;................GGGGAUGUAGCUCAG-UGGUA.GAGCGCA-UGCUUCGCAUGUAU.G
TRNA2X&nbsp;gcggccgucgucuaguGCGGCCGUCGUCUAGUCUGGAuUAGGACGCUGGCCUCCCAAGCCA.G
TRNA3X&nbsp;ccc.............GGCCCUGUGGC-UAGCUGGU-.CAAAGCGCCUGUCUAGUAAACAG.G
TRNA4X&nbsp;gggcgaauaguguca.GGGCGAAUAGUGUCAGCGGG-.-AGCACACCAGACUUGCAAUCUG.G
TRNA5X&nbsp;................GCCGGGAUAGCUCAGUUGGUA.GAGCAGA-GGACUGAAAAUCCUcG

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;AGGCCCcGGGUUCGAUCCCC--GGCAUCUCCA-.............
TRNA2X&nbsp;CAAUCCcGGGUUCGAAUCCCGGCGGCCGCACGG-cggccgca.....
TRNA3X&nbsp;AGAUCCuGGGUUCGAAUCCCAGCGGGGCCUCCA-gggg.........
TRNA4X&nbsp;UAGGGA.GGGUUCGAGUCCCUCUUUGUCCACCA-.............
TRNA5X&nbsp;UGUCAC.CAGUUCAAAUC--UGGUUCCUGGCA-ugguuccuggca.

<P>
</PRE>

<P>

<H3><A NAME="SECTION000123100000000000000"></A><A NAME="sec:fimtrain"></A>
<BR>
10.3.1 Training with FIMs
</H3>

<P>
A similar effect can be achieved by training with free insertion
modules.  Suppose the file <TT>ftrain.init</TT> contains
<PRE>

alphabet RNA
train trna10f.seq
randseed 0
REGULARIZER
alphabet RNA
GENERIC 1.88 0.25 0.37 1.81 15.52 3.76 0.22 0.26 4.00 0.25 0.25 0.25 0.25 0.25 0.25 0.25 0.25
TYPE 0 FIM
TYPE -1 FIM
ENDMODEL
</PRE>
This file specifies a default regularizer and that both the BEGIN node
and the last node are free insertion modules.  Training and aligning
on this file with the commands
<PRE>

 buildmodel ftrain -i ftrain.init
 align2model ftrain -i ftrain.mod -db trna10f.seq
 prettyalign ftrain.a2m -l90 &gt; ftrain.align
</PRE>
will produce the alignment
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;................GGGGAUGUAGCUCAG-.UGG...U.AGAGCGCAUGCUUCGCAUGUAU
TRNA2X&nbsp;gcggccgucgucuaguGCGGCCGUCGUCUAGU.CUGgauU.AGGACGCUGGCCUCCCAAGCCA
TRNA3X&nbsp;ccc.............GGCCCUGUGGC-UAGC.UGG...UcAAAGCGCCUGUCUAGUAAACAG
TRNA4X&nbsp;gggcgaauaguguca.GGGCGAAUAGUGUCAG.CGG...G.AGCACACCAGACUUGCAAUCUG
TRNA5X&nbsp;................GCCGGGAUAGCUCAGU.UGG...U.AGAGCAGAGGACUGAAAAUCCU
TRNA6&nbsp;&nbsp;................GGGGCCUUAGCUCAGC.UGG...G.AGAGCGCCUGCUUUGCACGCAG
TRNA7&nbsp;&nbsp;................GGGCACAUGGCGCAGU.UGG...U.AGCGCGCUUCCCUUGCAAGGAA
TRNA8&nbsp;&nbsp;................GGGCCCGUGGCCUAGU.CUGga.U.ACGGCACCGGCCUUCUAAGCCG
TRNA9&nbsp;&nbsp;c...............-GGCACGUAGCGCAGCcUGG...U.AGCGCACCGUCCUGGGGUUGCG
TRNA10&nbsp;................UCCGUCGUAGUCUAGG.UGGu..U.AGGAUACUCGGCUUUCACCCGA

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;GAGGCCCCGGGUUCGAUCCCCGGCAUCUCCA-............
TRNA2X&nbsp;GCAAUCCCGGGUUCGAAUCCCGGCGGCCGCAC-ggcggccgca..
TRNA3X&nbsp;GAGAUCCUGGGUUCGAAUCCCAGCGGGGCCUC-cagggg......
TRNA4X&nbsp;GUAG-GGAGGGUUCGAGUCCCUCUUUGUCCAC-ca..........
TRNA5X&nbsp;CGUGUCACCAGUUCAAAUCUGGUUCCUGGCAU-gguuccuggca.
TRNA6&nbsp;&nbsp;GAGGUCAGCGGU-CGA-CCCGCUAGGCUCCAC-ca..........
TRNA7&nbsp;&nbsp;GAGGUCAUCGGUUCGAUUCCGGUUGCGUCCA-............
TRNA8&nbsp;&nbsp;GGGAUCGGGGGUUCAAAUCCCUCCGGGUCCG-............
TRNA9&nbsp;&nbsp;GGGGUCGGAGGUUCAAAUCCUCUCGUGCCGAC-ca..........
TRNA10&nbsp;GAGA-CCCGGGUUCAAGUCCCGGCGACGGAAC-ca..........

<P>
</PRE>

<P>
The initial <TT>buildmodel</TT> run in the above case will produce an
warning message along the lines of
<PRE>

Some FIMs have non-zero match states (e.g., node 0),
a natural result of turning an existing or GENERIC node into a FIM.
These nodes are being renormalized as discussed in the SAM manual.
To avoid this message, use addfims or hmmedit when adding FIMs.
</PRE>
The purpose of this message is to apprise the user of a frequent
occurrence when using FIMs.  Effectively, the FIM has no match state,
and therefor transitions into the FIM nodes match state, as well as
the character generating probability table itself, should have zero
probability.  If they don't, SAM assumes that the node has not been
completely turned into a FIM, and renormalizes the node for you, in
the same manner in which <TT>addfims</TT> works.  SAM will do the
following:

<OL>
<LI>Zero the probabilities in the match table.
</LI>
<LI>Assign the insert to insert and the delete to insert
transitions within the node, as well as the transitions out of the
insert and delete state,  unity probability (in a FIM, probabilities
exiting the insert state and exiting the delete state will sum to two
rather than one).
</LI>
</OL>

<P>
In all cases, a FIM is never trained.

<P>
A similar affect can be achieved internally by running <TT>buildmodel</TT> with <TT>SW</TT> set to 1 (semi-local: part of the sequence
matching the entire model), 2 (local: part of the sequence matching
part of the model), or 3 (domain: part of the sequence matching all of
the model). See Section&nbsp;<A HREF="node4.html#sec:score">3.4</A>..

<P>
For example,

<P>
<PRE>

 buildmodel ftrainb -train trna10f.seq -randseed 1 -trainseed 1 -sw 3
 align2model ftrainb -i ftrainb.mod -db trna10f.seq -sw 3
 prettyalign ftrainb.a2m -l90 &gt; ftrainb.align
</PRE>
will produce the alignment
<PRE>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;.................GGGGAUGUAGCUCAG-..UGG..UAGAGCGCAUGCUUCGCAUGUAU
TRNA2X&nbsp;gcggccgucgucuagugCGGCCGUCGUCUAGUC..UGGauUAGGACGCUGGCCUCCCAAGCCA
TRNA3X&nbsp;ccc..............GGCCCUGUGGCUAGCU..GGU..CAAAGCGCCUGUCUAGUAAACAG
TRNA4X&nbsp;gggcgaauaguguca..GGGCGAAUAGUGUCAG..CGG..GAGCACACCAGACUUGCAAUCUG
TRNA5X&nbsp;.................GCCGGGAUAGCUCAGU..UGG..UAGAGCAGAGGACUGAAAAUCCU
TRNA6&nbsp;&nbsp;.................GGGGCCUUAGCUCAGC..UGG..GAGAGCGCCUGCUUUGCACGCAG
TRNA7&nbsp;&nbsp;.................GGGCACAUGGCGCAGU..UGG..UAGCGCGCUUCCCUUGCAAGGAA
TRNA8&nbsp;&nbsp;.................GGGCCCGUGGCCUAGUcuGGA..UACGGCACCGGCCUUCUAAGCCG
TRNA9&nbsp;&nbsp;.................CGGCACGUAGCGCAGCc.UGG..UAGCGCACCGUCCUGGGGUUGCG
TRNA10&nbsp;u................CCGUCGUAGUCUAGGU..GGU..UAGGAUACUCGGCUUUCACCCGA

<P>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
TRNA1&nbsp;&nbsp;GAGGCCCCGGGUUCGAUCCCCGGCAUCUCCA.............
TRNA2X&nbsp;GCAAUCCCGGGUUCGAAUCCCGGCGGCCGCAcggcggccgca..
TRNA3X&nbsp;GAGAUCCUGGGUUCGAAUCCCAGCGGGGCCUccagggg......
TRNA4X&nbsp;GUAGGGA-GGGUUCGAGUCCCUCUUUGUCCAcca..........
TRNA5X&nbsp;CGUGUCACCAGUUCAAAUCUGGUUCCUGGCAugguuccuggca.
TRNA6&nbsp;&nbsp;GAGGUCAGCGGU-CGA-CCCGCUAGGCUCCAcca..........
TRNA7&nbsp;&nbsp;GAGGUCAUCGGUUCGAUUCCGGUUGCGUCCA.............
TRNA8&nbsp;&nbsp;GGGAUCGGGGGUUCAAAUCCCUCCGGGUCCG.............
TRNA9&nbsp;&nbsp;GGGGUCGGAGGUUCAAAUCCUCUCGUGCCGAcca..........
TRNA10&nbsp;GAGA-CCCGGGUUCAAGUCCCGGCGACGGAAcca..........

<P>
</PRE>

<P>
Note that in this case, the model output by <TT>buildmodel</TT> does not
include the FIMs that SAM internally added to the model, thus we must
specify domain alignment (essentially, global alignment using a model
that has FIMs on both ends) with <TT>align2model</TT>.

<P>

<H2><A NAME="SECTION000124000000000000000"></A><A NAME="sec:grabdp"></A>
<BR>
10.4 <TT>grabdp</TT>
</H2>

<P>
The <TT>grabdp</TT> program can be used to examine the dynamic
programming table of the forward-backward dynamic programming
calculation.  The resulting <TT>runname.pdoc</TT> file is very large, as
it includes the posteriors for the three transitions into match and
the character posteriors for each for the index points in the dynamic
programming matrix (i.e., approximately sequence length times model
length records).

<P>
Because of the size of these files, the <TT>grabdp</TT> program will only
output the values for a single sequence and model, choosing the first
sequence in a database.  The program requires a model file, a database
file, and optionally a single sequence identifier.

<P>
The <TT>view_pdoc</TT> program is a viewer for the posterior decoding
files.  It is available in the associated tools directory, available
for download from the SAM WWW site.  The program requires
the <TT>wish</TT> simple windowing shell, part of the Tk suite.  It may
be necessary to edit the <TT>view_pdoc</TT> script to change path names
for <TT>xwd</TT>, the X window dump program, and <TT>ppmtogif</TT>, the
image conversion program.

<P>
For diagnostic purposes, it can sometimes be useful to view the amino acid
priors and posteriors.  When the <TT>dump_match_probs</TT> flag is set, 
<TT>grabdp</TT> will generate a RDB file listing the posteriors for each node,
and listing the priors under the label ``FREQAVE''.

<P>

<H2><A NAME="SECTION000125000000000000000"></A><A NAME="sec:get_fisher_scores"></A>
<BR>
10.5 <TT>get_fisher_scores</TT>
</H2>

<P>
The <TT>get_fisher_scores</TT> program scores sequences with a model and
outputs <EM>Fisher score vectors</EM> for input an external discriminative learning
program as described in:

<P>
<P>
<BLOCKQUOTE>T. Jaakkola and M. Diekhans and D. Haussler
<A NAME="tex2html54"
  HREF="http://www.soe.ucsc.edu/research/compbio/discriminative/Jaakola2-1998.ps">A discriminative framework for detecting remote protein homologies</A>,
<EM>Journal of Computational Biology</EM> 7(1):2000

</BLOCKQUOTE>

<P>
The <TT>hmmscore</TT> arguments specifying the model, sequences, and
scoring parameters (e.g. -dpstyle, -sw) are valid.
The  <TT>-fisher_feature</TT> parameter specifies which features
are used to calculate the Fisher score vectors.
Use the <TT>-null_score_weight_scale</TT> parameter to weight sequences by their 
NLL-NULL score.

<P>
As the Fisher score vectors can be large, they are 
written as a stream to <TT>stdout</TT>, allowing piping directly to
another program without creating an intermediate file.  The output
stream maybe in ASCII or binary format, selected by
the <TT>-binary_output</TT><A HREF="node10.html#sec:binmods">8.4.4</A> option.
The first line of each record indicates the record format and
sequence id.  ASCII records start with a line in the form
<PRE>

&gt;A <EM>seqid1</EM>
</PRE>
and binary start with
<PRE>

&gt;B <EM>seqid1</EM>
</PRE>

<P>
In both formats, the record header is terminated by a newline.

<P>
The first number in a record indicates the total number of features
in the record. It is followed by that many floating point numbers.  In
an ASCII stream, the numbers area whitespace separated (both spaces
and newlines are used) designed to be read by <TT>fscanf</TT>.
In the binary format, the feature count is an <TT>int</TT> and the
features are <TT>float</TT>s.

<P>
<PRE>

&gt;A seqid1
3
0.1 0.2
&gt;A seqid2
3
1.2 -0.3 1.0
</PRE>

<P>
The features may also be written to a file in RDB format, specified
with the <TT>-rdb</TT>, creating a file names 
<TT><EM>runname</EM>.fisher-rdb</TT>. 
The <TT>-write_dist</TT> parameter can be used to
generate a <TT><EM>runname</EM>.dist-rdb</TT> score file, which maybe
useful in analyzing the results.

<P>
For example, the following command will derive <EM>Fisher scores vectors</EM>
from the match states and Dirichlet mixture used to train the model: 
<PRE>

get_fisher_scores unused -fisher_feature match_prior -i test.mod -db prots.seq -sw 2
</PRE>

<P>

<H2><A NAME="SECTION000126000000000000000"></A><A NAME="sec:modfa"></A><A NAME="sec:modelfromalign"></A>
<BR>
10.6 <TT>modelfromalign</TT>
</H2>

<P>
<TT>Modelfromalign</TT> takes a multiple alignment and converts it to a
model.  As a base model, the program starts with the default
regularizer (or a specified regularizer, as for <TT>buildmodel</TT>), and
then calculates node frequencies according to the given multiple
alignment.  Sequences in the alignment can be weighted according to
the <TT>alignment_weights</TT> file, discussed in Section&nbsp;<A HREF="node11.html#sec:weights">9.4</A>.

<P>
If a trustworthy hand alignment is available, this is often the best
way to build a model:  create one from an alignment, and then refine
it using <TT>buildmodel</TT>.  If some sections of the alignment
are particularly important, it may be desirable to make them fixed
nodes, as described in Section&nbsp;<A HREF="node10.html#sec:spec-nodes">8.4.2</A>.

<P>
The <TT>alignfile</TT> parameter can also be used with <TT>buildmodel</TT>
to specify a seed alignment. 
See Section&nbsp;<A HREF="node10.html#sec:alignfile">8.3</A>.

<P>
The <TT>modelfromalign</TT> program will read any <TT>readseq</TT> format,
but has a few special interpretations.  It follows the <TT>align2model</TT> convention that lowercase letters are insertions, hyphens
are deletions, and dots are simply filler for insertions in other
sequences.  Additionally, the letter `O' is converted into a FIM,
following a convention used in some multiple alignment formats (the
other SAM programs will convert `O' to the `X' wildcard).
If
all sequences do not have the same number of uppercase letters and
hyphens, then <TT>modelfromalign</TT> will try treating all characters as
uppercase and all periods as hyphens (i.e., it will try modeling each
character as a match column).    

<P>
The program has one required parameter that must be set on the command
line or in an inserted or <TT>.samrc</TT> file:  the <TT>alignfile</TT> is
the name of the file with the alignment.  For example,

<P>
<PRE>

modelfromalign trna2 -alignfile trna2.align 
</PRE>

<P>
will produce a model from the <TT>trna2.align</TT> alignment.  All
<TT>buildmodel</TT> parameters dealing with regularization and prior
libraries are used in the conversion from column frequencies to a
model.  Prior libraries are particularly helpful when converting a
small protein alignment to a model.

<P>
If the alignment is of a motif, setting the <TT>align_fim</TT> variable
to 1 will cause FIMs to be added to the model before printing it out. 

<P>

<H2><A NAME="SECTION000127000000000000000"></A> <A NAME="sec:pathprobs"></A>
<BR>
10.7 <TT>pathprobs</TT>
</H2>

<P>
The <TT>pathprobs</TT> program can be used to generate the posterior
probabilities of each of the characters in an alignment.  This is a
small subset of the information generated by the <TT>grabdp</TT>
program.  In effect, <TT>pathprobs</TT> uses the alignment to trace
through the posterior matrix and decide which character emission
probabilities are desired.

<P>
The program can be run using, for example,
<PRE>

pathprobs 1babA-50p -modelfile 1babA-t99-fw0.7.mod -alignfile 1babA-50.a2m
</PRE>
which will produce tab-separated RDB output alignment probability file
<TT>1babA-50p.apr</TT> along the lines of:
<PRE>

SequenceID	SeqIndex	ModIndex	ModState	NegLnPosterior	T0:protein
10s	8n	8n	8s	9n	10s
1babA	0	4	I	7.764000	X
1babA	1	4	I	7.754000	M
1babA	2	5	M	14.888000	E
1babA	3	6	M	15.737000	L
1babA	4	7	M	16.393000	S
1babA	5	8	M	16.734000	P
1babA	6	9	M	15.943000	A
</PRE>
Here, the first line identifies the columns.  In this case,
a single-track model and sequences were used.  The second line our
field witdth and type specifiers.  The remaining lines show, for each
character in the labeled sequence, the state that it was generated by
according to the alignment (a model index and an `M' for match or `I'
for insert; `D' never occurs because it does not generate any
characters), the negative natural log posterior probability according to the
forward-backward algorithm of the character appearing in that state,
and the (possibly multi-track) character itself.

<P>
The program will also generate the <TT>1babA-50p.pa2m</TT> file that interleaves
sequences and digits corresponding to log posterior probabilities
multiplied by <TT>pplambda</TT> and added to 10. The values are truncated
to a single digit, so 10.0 (unity probability) is reduced to the digit
9.  With the default <TT>pplambda</TT> of <!-- MATH
 $\ln(2)/2=0.34657$
 -->
<IMG
 WIDTH="129" HEIGHT="34" ALIGN="MIDDLE" BORDER="0"
 SRC="img74.png"
 ALT="$\ln(2)/2=0.34657$">,  10 (if not reduced to
9)  indicates a  completely determined position or 0 bits of
information, 8 indicates 1 bit of information, 6 indicates two bits, 4
indicates 3 bits, 2 indicates 4 bits, and 0 indicates 5 or more bits.
Insertions are indicated by the letters inserted.    For example, here
is a prettyaligned excerpt of a <TT>.pa2m</TT> file for several globins.
<PRE>

;&nbsp;&nbsp;SAM:&nbsp;prettyalign&nbsp;v3.3.1&nbsp;(December&nbsp;20,&nbsp;2001)&nbsp;compiled&nbsp;12/27/01_15:15:58
;&nbsp;&nbsp;(c)&nbsp;1992-2001&nbsp;Regents&nbsp;of&nbsp;the&nbsp;University&nbsp;of&nbsp;California,&nbsp;Santa&nbsp;Cruz
;
;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence&nbsp;Alignment&nbsp;and&nbsp;Modeling&nbsp;Software&nbsp;System
;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;http://www.cse.ucsc.edu/research/compbio/sam.html
;
;&nbsp;---------&nbsp;Citations&nbsp;(SAM,&nbsp;SAM-T99,&nbsp;HMMs)&nbsp;----------
;&nbsp;R.&nbsp;Hughey,&nbsp;A.&nbsp;Krogh,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;sequence&nbsp;analysis:
;&nbsp;&nbsp;Extension&nbsp;and&nbsp;analysis&nbsp;of&nbsp;the&nbsp;basic&nbsp;method,&nbsp;CABIOS&nbsp;12:95-107,&nbsp;1996.
;&nbsp;K.&nbsp;Karplus,&nbsp;C.&nbsp;Barrett,&nbsp;R.&nbsp;Hughey,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;for&nbsp;detecting
;&nbsp;&nbsp;remote&nbsp;protein&nbsp;homologies,&nbsp;Bioinformatics&nbsp;14(10):846-856,&nbsp;1998.
;&nbsp;A.&nbsp;Krogh&nbsp;et&nbsp;al.,&nbsp;Hidden&nbsp;Markov&nbsp;models&nbsp;in&nbsp;computational&nbsp;biology:
;&nbsp;&nbsp;Applications&nbsp;to&nbsp;protein&nbsp;modeling,&nbsp;JMB&nbsp;235:1501-1531,&nbsp;Feb&nbsp;1994.
;&nbsp;-----------------------------------
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
1babA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;xm......ELSPADKTNVKAAWGKVGA.....H......AGEYGA.EALERMFLSFPTTKTYFPHFDLS.H-.-.-G
1babA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;xm......5555555554434445687.....7......776555.44433445444454344577666.6-.-.-2
BAHG$VITSP&nbsp;........MLDQQTINIIKATVPVLKE.....H......GVTITT.TFYKNLFAKHPEVRPLFDMGRQE.S-.-.--
BAHG$VITSP&nbsp;........7777777554556666788.....8......888887.77754445455677677899999.9-.-.--
GLB$APLJU&nbsp;&nbsp;a.......-LSAADAGLLAQSWAPVFA.....N......SDANGA.SFLVALFTQFPESANFFNDFKGK.SL.AD.IQA
GLB$APLJU&nbsp;&nbsp;a.......-665565555534456677.....7......666555.45554555444555456677555.65.54.444
GLB$APLKU&nbsp;&nbsp;s.......-LSAAEADLVGKSWAPVYA.....N......KDADGA.NFLLSLFEKFPNNANYFADFKGK.SI.AD.IKA
GLB$APLKU&nbsp;&nbsp;s.......-666665454534456677.....6......666555.55554455445665456777555.55.54.444
</PRE>

<P>
It is of course critical that the alignment and the model are of the
same length.  A typical use of the program would be to generate a
Viterbi or Posterior-Decoded alignment and then use <TT>pathprobs</TT> to
determine how strongly that alignment is reflected in the posterior
matrix.

<P>

<H2><A NAME="SECTION000128000000000000000"></A> <A NAME="sec:predicttrack"></A>
<BR>
10.8 <TT>predict_track</TT>
</H2>

<P>
The predict track program can be used to predict the secondary
structure of a sequence based on a protein model and a set of
sequences of known secondary structure.  The program can either be
provided with models for all tracks or (in the 2-track case) create a
model for the second track based on a database of 2-track sequences.
The program thus requires a base model or set of base models, a
prediction sequence, a multi-track alphabet definition, and (when it
generates its own second-track model) a 2-track database.
See Section&nbsp;<A HREF="node12.html#sec:multitrack">10.2.5</A>.

<P>
For output, the program will produce a <TT>.ptrack</TT> file of
per-seuqence structure predictions or if <TT>rdb</TT> is set, a <TT>.rdb</TT> file.  If the program created a model for the second track, it
will be written to a <TT>.mod</TT> file, and can be used, for example,
with multi-track HMM scoring and alignment.

<P>
For example, the command:
<PRE>

predict_track db1.test1 -sw 2 -subtractnull 4  -i db1.test1.mod
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-a protein,ELH -db predict.db1.seq,predict.db1.2d 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-predictseq db1.test1.seq
</PRE>
will produce as output <TT>db1.test1.ptrack</TT>, a predicted secondary
structure sequence for the first sequence in the <TT>db1.test1.seq</TT>
file based on the protein model and the primary and secondary
structure sequences provided to <TT>predict_track</TT>, and <TT>db1.test1.mod</TT>, a secondary structure model.  The model is  created by
aligning the <TT>predict.db1.seq</TT> sequences to <TT>db1.test1.mod</TT>,
weighting the sequences, and then tabulating the frequencies of each
letter in each model node according to the generated sequence
weights.  The frequencies are regularized with the ELH default
regularizer, and transitions are copied from <TT>db1.test1.mod</TT>.
See Section&nbsp;<A HREF="node11.html#sec:alignweight">9.4.3</A>.

<P>

<P></P>
<DIV ALIGN="CENTER"><A NAME="fig:ptracklogo"></A><A NAME="6211"></A>
<TABLE>
<CAPTION ALIGN="BOTTOM"><STRONG>Figure 16:</STRONG>
A model and its predicted secondary struction (<TT>makelogo
db1.test1.eps -modelfile db1.test1.mod -logo_captionf
db1.test1.ptrack -logo_sections 1-32</TT>).</CAPTION>
<TR><TD><IMG
 WIDTH="446" HEIGHT="94" BORDER="0"
 SRC="img75.png"
 ALT="\begin{figure}\begin{center}
\psfig{figure=db1.test1.eps,width=0.8\textwidth}\end{center}\end{figure}"></TD></TR>
</TABLE>
</DIV><P></P>

<P>
The model and the prediction can be visualized with <TT>makelogo</TT>
using the <TT>logo_captionf</TT> option (Figure&nbsp;<A HREF="node12.html#fig:ptracklogo">16</A>).

<P>
In creating a secondary track model, <TT>FIM_method_train</TT> and
<TT>Insert_method_train</TT> are used to set the FIM and Insert
tables.  It is highly recommended to set these to 0 to use the
regularizer (equivalent to the generic node) rather than the <TT>buildmodel</TT> letter-count defaults.  

<P>

<H2><A NAME="SECTION000129000000000000000">
10.9 Model manipulation</A>
</H2>

<P>

<H3><A NAME="SECTION000129100000000000000"></A> <A NAME="sec:drawmodel"></A>
<BR>
10.9.1 <TT>drawmodel</TT>
</H3>

<P>
The <code>drawmodel</code> program is a means of generating a postscript
drawing from a model, regularizer, or frequency count data (created
with the <TT>print_frequencies</TT> option).  The program is run with two
arguments, the first being a model file, and the second the output
file.
<PRE>

drawmodel model.mod drawing.ps
</PRE>

<P>
The program will scan the file, looking for models, regularizers, and
frequency counts, and query whether or not each one should be printed,
after presenting a line from the file.

<P>
There are two drawing options: overall and local.  Overall is the
correct option for frequency counts -- the outgoing transitions of
each state are drawn in different styles depending on the fraction of
all sequences that use that transition.  The circular delete states
show the node number, while the diamond insert states show average
number of characters, rounded up, inserted by each sequence that used
the insert state.

<P>
In the local option, suitable for models and regularizers, transitions
from a given state are drawn according to what percentage of sequences
in that given state take each transition.  Also, the diamond insert
states have the percent of sequences which, once within the insert
state, remain in the insert state. (See Figure&nbsp;<A HREF="node4.html#fig:drawmodel">4</A> on
page&nbsp;<A HREF="node4.html#fig:drawmodel"><IMG  ALIGN="BOTTOM" BORDER="1" ALT="[*]"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/crossref.png"></A>.) 

<P>
The drawmodel program has several command-line options: <code>-landscape</code>
to draw models in landscape format, and <code>-scale num</code>, to change
the scale of the drawing to an arbitrary floating-point number.  The
default is portrait mode with a scale 0.235, which fits six row of 19
protein model nodes (plus one ghost node, the first model of the next
row) on each page.  Larger scale settings increase the size of the
model nodes, and cause fewer to be placed on each line.  For
additional customization, the postscript file, which is readable, can
be modified (e.g., to change print on a different size of paper).

<P>
If <code>-mod n</code>, <code>-freq n</code>, or <code>-reg n</code> is specified on the
command line, the <IMG
 WIDTH="16" HEIGHT="17" ALIGN="BOTTOM" BORDER="0"
 SRC="img42.png"
 ALT="$n$">th model, frequency count, or regularizer will be
selected for printing, and the interactive queries on which model to
print will not occur.

<P>
Europeans and other people who like the A4 paper size can change
`<TT>/A4 false def</TT>' to `<TT>/A4 true def</TT>' near the bottom
of the header commands in the output of <TT>drawmodel</TT>.

<P>

<H3><A NAME="SECTION000129200000000000000"></A><A NAME="sec:hmmcon"></A>
<BR>
10.9.2 <TT>hmmconvert</TT>
</H3>

<P>
The <TT>hmmconvert</TT> program takes a model file as input and outputs
a new model file in the `opposite' format (binary or text) from the
original.

<P>
The command syntax is as follows:
<PRE>

hmmconvert runname -model_file test.mod
</PRE>

<P>
If <TT>test.mod</TT> is in text format, <TT>runname.mod</TT> will be in binary 
format and vice versa.  If you wish to destroy your original model file, use a
runname of the file's name without the <TT>.mod</TT> extension.

<P>
<PRE>

hmmconvert test -model_file test.mod
</PRE>

<P>
If <TT>test.mod</TT> is in text format, this command will create a new file
called <TT>test.mod</TT> in binary format.  If <TT>test.mod</TT> is binary, the new
<TT>test.mod</TT> will be text.

<P>
To preserve your original modelfile, enter a new runname.

<P>
<PRE>

hmmconvert new -model_file test.mod
</PRE>

<P>
This will create a file called <TT>new.mod</TT> in the opposite format of
<TT>test.mod</TT>.  <TT>test.mod</TT> will be left alone.

<P>

<H3><A NAME="SECTION000129300000000000000"></A> <A NAME="sec:makelogo"></A>
<BR>
10.9.3 <TT>makelogo</TT>
</H3>

<P>
Sequence logos (Schneider and Stephens, NAR 18:6097-6199, 1990) are
an excellent means of viewing model match state distributions
(Figures&nbsp;<A HREF="node4.html#fig:testlogo">3</A> and&nbsp;<A HREF="node12.html#fig:ptracklogo">16</A>).  Each
HMM match state is represented by a bar.  The bar height corresponds
to the negative entropy in bits of that match state distribution.
Each bar is composed of letters in the target alphabet and the letter
`X'.  The size of the letters correpond to the relative freqeuncies of
the letters; letters of low frequency are lumped together into the `X'
character.  FIMs are represented with a large letter `O'.   The <TT>makelogo</TT> program has default internal colors for 
protein, nucleotide, and secondary structure alphabets.

<P>
The simplest way to run the program is with just a model file to
create the postscript file <TT>test.eps</TT>,  as:
<PRE>

makelogo test -modelfile test.mod
</PRE>

<P>

<P></P>
<DIV ALIGN="CENTER"><A NAME="fig:logofig1"></A><A NAME="6179"></A>
<TABLE>
<CAPTION ALIGN="BOTTOM"><STRONG>Figure 17:</STRONG>
Primary structure model of 3chy with sequence and secondary
structure</CAPTION>
<TR><TD><IMG
 WIDTH="557" HEIGHT="434" BORDER="0"
 SRC="img76.png"
 ALT="\begin{figure}\centerline{\psfig{figure=advancedlogo1.eps,width=\textwidth}}\end{figure}"></TD></TR>
</TABLE>
</DIV><P></P>

<P>

<P></P>
<DIV ALIGN="CENTER"><A NAME="fig:logofig2"></A><A NAME="6184"></A>
<TABLE>
<CAPTION ALIGN="BOTTOM"><STRONG>Figure 18:</STRONG>
Predicted secondary structure  model of 3chy with sequence
and true secondary structure</CAPTION>
<TR><TD><IMG
 WIDTH="556" HEIGHT="404" BORDER="0"
 SRC="img77.png"
 ALT="\begin{figure}\centerline{\psfig{figure=advancedlogo.eps,width=\textwidth}}\end{figure}"></TD></TR>
</TABLE>
</DIV><P></P>

<P>
The logos are colored according to the alphabet of the model; new
color schemes can be specified with the <TT>logo_color</TT> file name.
The internal defaults are in the example files <TT>protein.colors</TT>,
<TT>nucleotide.colors</TT>, and <TT>stride.colors</TT> included in the SAM
distribution, and can be modified to create new color schemes.  Each
line of the file is either a comment (preceeded by an asterix) or
includes a single letter and a triple of RGB values between 0.0 and
1.0.  The <TT>makelogo</TT> file will look for color files in the current
directory as well as the <TT>PRIOR_PATH</TT>. See Section&nbsp;<A HREF="node13.html#sec:env">11.1</A>.

<P>
The logo diagram can be annotated with a title (<TT>logo_title</TT>),
and also with sequences in two ways.  First, a single sequence, such
as a protein that the model is based on, can displayed under the graph
using <TT>logo_under_file</TT> (and the color file <TT>logo_under_color</TT> if black is not desired).  If a secondary
structure sequence is specified with <TT>logo_captionf</TT> (and the
corresponding <TT>logo_captionf_color</TT>), repeated occurrences of
the secondary structure character will be indicated as a range of
positions with that the character.

<P>
For example, Figure&nbsp;<A HREF="node12.html#fig:logofig1">17</A> shows a SAM-T2K (the
pre-release sucessor to SAM-T99) model for the PDB sequect 2chy.  The
logo diagram includes the 2chy sequence under the sequence, as well as
the secondary structure, both input from separate FASTA-format files.
The following Figure&nbsp;<A HREF="node12.html#fig:logofig2">18</A> shows a SAM-T2K secondary
structure prediction HMM calculated without the aid of the 
secondary structure sequence for 2chy.  The logo again includes the
2chy sequence as a <TT>logo_under_file</TT>, and the true secondary
structure as a <TT>logo_captionf</TT>.  

<P>
The program has a number of other options (all beginning with
<TT>logo_</TT>).  See Section&nbsp;<A HREF="node14.html#sec:parameters">12</A>.

<P>

<H3><A NAME="SECTION000129400000000000000"></A> <A NAME="sec:modifymodel"></A>
<BR>
10.9.4 <TT>modifymodel</TT>
</H3>

<P>
<TT>modifymodel</TT> is a utility program for modifying
SAM models.  It has an interactive type in
interface as well as command line arguments.
This program will grow over time to meet the needs
of users.
It contains extensive built in help with examples and a
question mark (<TT>?</TT>) will list the commands:
<PRE>

AddNodes&nbsp;nodeNumber&nbsp;count&nbsp;&nbsp;#adds&nbsp;generic&nbsp;nodes&nbsp;after&nbsp;specified
Change&nbsp;subCommand...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;#has&nbsp;sub&nbsp;commands,&nbsp;change&nbsp;?&nbsp;to&nbsp;see
DeleteNodes&nbsp;nodeSpec&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;#deletes&nbsp;the&nbsp;specified&nbsp;nodes
DuplicateNode&nbsp;nodeNumber&nbsp;count&nbsp;&nbsp;#copies&nbsp;the&nbsp;specified&nbsp;node
NewModel&nbsp;modelLength&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;#make&nbsp;new&nbsp;model&nbsp;from&nbsp;default&nbsp;generic
PrependToCurrentModel&nbsp;fileName&nbsp;&nbsp;#reads&nbsp;model&nbsp;from&nbsp;file,&nbsp;pastes&nbsp;on&nbsp;front
PostpendToCurrentModel&nbsp;fileName&nbsp;#reads&nbsp;model&nbsp;from&nbsp;file,&nbsp;puts&nbsp;on&nbsp;end
ReadModel&nbsp;fileName
Show&nbsp;subCommand...
WriteModel&nbsp;[fileName]&nbsp;&nbsp;&nbsp;#defaults&nbsp;to&nbsp;name&nbsp;of&nbsp;file&nbsp;last&nbsp;read&nbsp;in
Quit
Help
</PRE>

<P>
Most (at the moment all) operations act on the 
current model: which is the last read in model.
The basic usage is to read a model in, do what you want to it
(add nodes, change types, paste models together, etc.)
and write it back out.

<P>
The currently supported functions are

<UL>
<LI>Changing node types.
</LI>
<LI>Adding/duplicating/deleting nodes.
</LI>
<LI>Cutting models into smaller parts.
</LI>
<LI>Pasting models together.
</LI>
</UL>

<P>
To print all the types of the nodes (use this to see
how long the model is):
<PRE>

show type all
</PRE>

<P>
To change all nodes that do not have a type into KEEP nodes,
<PRE>

change type _ K ALL
</PRE>

<P>
To change node 9 into a FIM:
<PRE>

change type + F 9
</PRE>

<P>
To shorten a model by deleting nodes 4 through the end:
<PRE>

delete 4,end
</PRE>

<P>
Below is an example of reading a model, deleting the first 20 nodes,
adding a new node to the front and changing it to a FIM, then pasting
it on to the end of a second model, and finally writing the modified
model out.

<P>
<PRE>

read one.mod
delete 1,20
add 0 1
change type + F 1
prepend two.mod
writemodel both.mod
quit
</PRE>

<P>

<H3><A NAME="SECTION000129500000000000000"></A><A NAME="sec:convert"></A>
<BR>
10.9.5 Conversion between SAM and HMMer
</H3>

<P>
SAM and HMMer are the two most widely used HMM sequences alignment and
modeling systems.  Even though they both employ hidden Markov models,
their file formats are vastly different.  Provided with this release
are the conversion programs <TT>sam2hmmer</TT> and <TT>hmmer2sam</TT> that
make almost complete conversion between SAM and HMMer v1.7
possible.

<P>
Martin Madera and Julian Gough have written a perl converter between
SAM and HMMer 2.0 formats, available from
<A NAME="tex2html58"
  HREF="http://www.mrc-lmb.cam.ac.uk/genomes/julian/convert/convert.html">http://www.mrc-lmb.cam.ac.uk/genomes/julian/convert/convert.html</A>
and the <A NAME="tex2html59"
  HREF="http://www.cse.ucsc.edu/research/compbio/sam.html">SAM WWW site</A>.

<P>
The conversion is ``almost'' because there is some information
loss, and this information loss is due to the structural 
differences between the two systems.  For those users who
wish to use the conversion utilities, we discuss these issues 
below.

<P>

<H4><A NAME="SECTION000129510000000000000">
10.9.5.1 Internal HMM Structure</A>
</H4>

<P>
Both SAM and HMMer use the same linear model of nodes that contain 3
states - match, delete, insert.  The number of transitions per node
is the same, 9.  The internal difference is in how transitions
between nodes are viewed.  SAM views transitions INTO a node, thus
each node contains the transitions FROM the previous node.  HMMer is
the opposite, transitions are TO the next node.  HMMer's convention
makes more sense because each triple of transitions stored in a
node sums to one. SAM's convention, slightly helpful in the internal
dynamic programming loop, is maintained  for historical
reasons.  This becomes a problem at the beginning and end of the
model.

<P>
The begin node of HMMer has a valid transition probability
distribution out of the MATCH state (in that the transition
probabilities sum to one), but has the null distribution for matching
characters.  This means HMMer always starts in the insert state, and
the begin node match state is not used.  The begin node of SAM can
start in either non-character-generating match or an insert.  So the
SAM to HMMer conversion is fine, but HMMer to SAM conversion leaves
SAM's begin node match state character table zero.  The net result is
that SAM is forced to start in the insert state.

<P>
The end node of HMMer also has a zero match state character
table.

<P>
The two programs are used as follows:  
<PRE>

hmmer2sam hmmerfile.hmmer  newsamfile.mod
sam2hmmer newhmmerfile -i samfile.mod
</PRE>
Above, the output of <TT>sam2hmmer</TT> is stored in the file <TT>newhmmerfile.hmmer</TT>. 

<P>

<H4><A NAME="SECTION000129520000000000000">
10.9.5.2 Extra Information in HMMer</A>
</H4>

<P>
HMMer allows other information on a per node basis.  This information
seems to be solvent access and consensus information.  This
information cannot be used in SAM at this
time, and is silently dropped.  A note of
the global presence of this extra information is noted as a comment in
the SAM model, solely for documentation purposes.

<P>

<H4><A NAME="SECTION000129530000000000000">
10.9.5.3 Extra Information in SAM</A>
</H4>

<P>
SAM has constructs in the model that are not supported in HMMer, and
these constitute the greatest difference in the two systems.  SAM has
per node ``types'' that control learning and evaluation parameters.
The problematic type is the Free Insertion Module (or FIM).  This
special type allows SAM to have a single node give equal cost for all
characters by effectively matching one or more characters.  These FIMs
can be positioned anywhere in a model, but are most commonly
positioned at the ends (the begin node and the node before the end
node).  HMMer has no such concept, and handling FIMs in a conversion
is a problem.  FIMs at the begin and end are are not too critical,
because HMMer's system always treats the begin and end just like they
are free inserts (also, HMMer re-normalizes all nodes so any strange
distributions are ``fixed'').  FIMs in the middle of a model can NOT be
converted in an exact manner.  Due to the structure in SAM, some
probabilities in a FIM node do not sum to one and the match state
transition probabilities are all zero.  During conversion, these
nodes have to be ``fixed'' so that valid (sum to one) probability
distributions are given to HMMer.

<P>
One way to fix up a FIM is to remove it: it maps to nothing in a HMMer
model.  This method is not too bad if the FIM does not usually match
many characters, but is very bad if the FIM is used a lot.  A second
way is fabricate some normal cheap insert states and hope that they
match the average length of the FIM.  This method is hard, since there
is not enough information to guess the number of insert nodes to add.
(If the frequencies are present in the SAM model, then this can be
done.  However this information is not always present.).  <TT>sam2hmmer</TT> utilizes the former option.  The transition probabilities
for the match state are computed from the exit probabilities of the
FIM (a chain of FIMs in a row are removed as they were one).  A SAM
FIM has exits from the insert and delete states, and the previous
non-FIM node has a transition from match to insert.  The new
transitions are computed from these numbers by breaking the match to
FIM transition into two parts that have the same ratio as the exit
probabilities of the FIM.  This same computation is done to compute
the new transitions from delete and insert.

<P>
It is not possible to record any extra information from SAM into a
HMMer model as HMMer does not support comments in the file.

<P>

<H2><A NAME="SECTION0001210000000000000000"></A><A NAME="sec:plottingprogs"></A>
<BR>
10.10 Plotting Programs
</H2>

<P>
SAM has
several plotting programs to assist in data analysis. The programs
require <TT>gnuplot</TT>
<A NAME="tex2html60"
  HREF="http://www.cs.dartmouth.edu/gnuplot_info.html">http://www.cs.dartmouth.edu/gnuplot_info.html</A>
to be in the user's path.
 In general, the programs create one or more
<TT>.data</TT> files, a <TT>.plt</TT> file of <TT>gnuplot</TT> commands, and a
postscript file.

<P>
Options common to the programs include

<P>
<TT>plotcolumn</TT> --
Column of score file to use in calculating plots.  Length (0), simple
null model (1), complex or reverse null model (2), or Evalue (3, the
default).   If Evalues are requested but not present, then column 1 is
used. 

<P>
<TT>plotps</TT> --
Creates a postscript file <TT>runname.ps</TT> using <TT>gnuplot</TT> if set
to 1, 2, or 3.  When set to 0, only a <TT>.plt</TT> file and one or two
<TT>.data</TT> 
files are generated.  A setting of 1 generates a plot in <TT>gnuplot</TT>'s default rectangular shape, while a setting of 2 generates a
square plot.  For options 1 and 2, the <TT>.data</TT> and <TT>.plt</TT> files used to
create the postscript file are deleted.  When set to 3, the postscript
file is generated and the <TT>.data</TT> and <TT>.plt</TT> files are retained.  The
default setting is 1.

<P>
<TT>plotleft</TT> --
Lowest X axis value on a graph generated by <TT>gnuplot</TT>.  The X axis
is calculated internally if  <TT>plotleft</TT>=<TT>plotright</TT>.  The
default setting is 0.0.

<P>
<TT>plotright</TT> --  
Highest X axis value on a graph generated by <TT>gnuplot</TT>.  The X axis
is calculated internally if  <TT>plotleft</TT>=<TT>plotright</TT>. 
The default setting is 0.0.

<P>
<TT>plotmax</TT> -- 
Highest Y axis value on a graph generated by <TT>gnuplot</TT>. The Y axis
is calculated internally if <TT>plotmax</TT>=<TT>plotmin</TT>.  The default 
setting is 0.0

<P>
<TT>plotmin</TT> -- 
Lowest Y axis value on a graph generated by <TT>gnuplot</TT>.  The Y axis
is calculated internally if <TT>plotmax</TT>=<TT>plotmin</TT>. 
The default setting is 0.0.

<P>
<TT>plotline</TT> -- 
Creates a vertical line at this value in a graph generated by <TT>gnuplot</TT>
if plotline is nonzero.  The default setting is 0.0.

<P>
<TT>plotnegate</TT> -- 
Negates the scores on a graph generated by <TT>gnuplot</TT> if set to 1.
The default setting is 0 (off).

<P>

<H3><A NAME="SECTION0001210100000000000000"></A><A NAME="sec:makehist"></A>
<BR>
10.10.1 <TT>makehist</TT>
</H3>

<P>
The <TT>makehist</TT> program will generate two or three  files from one
or two <TT>.dist</TT>
distance files created by <TT>hmmscore</TT> or a <TT>.mstat</TT> distance
file created by <TT>multdomain</TT>.  The files can be used with the
popular <TT>gnuplot</TT> plotting package.  Histograms are most useful to
discover and  examine the separation between family members and
non-family members during a database search.  The <TT>makehist</TT>
program requires a <TT>distfile</TT> as an argument, in addition to the
runname.  For example, to make a histogram from <TT>test.dist</TT>, the
following command would be used:
<PRE>

makehist test -distfile test.dist
</PRE>
Then, to view the histogram, run the <TT>gnuplot</TT> program and enter
the command <code>load "test.plt"</code> to view the results.  The <TT>test.plt</TT> file may be modified to change the graph. If the two marked
lines in the file are uncommented, a postscript file will be generated.

<P>

<P></P>
<DIV ALIGN="CENTER"><A NAME="histfig"></A><A NAME="6570"></A>
<TABLE>
<CAPTION ALIGN="BOTTOM"><STRONG>Figure 19:</STRONG>
A histogram generated by <TT>makehist</TT> and <TT>gnuplot</TT>.</CAPTION>
<TR><TD><IMG
 WIDTH="554" HEIGHT="392" BORDER="0"
 SRC="img78.png"
 ALT="\begin{figure}\centerline{\psfig{figure=testhist.ps}}\par\end{figure}"></TD></TR>
</TABLE>
</DIV><P></P>

<P>
If a second file is specified, using the <TT>distfile2</TT> option, a
second histogram is placed above the first one.
The <TT>makehist</TT> program has an optional argument, <TT>histbins</TT>,
which can be used to set the number of bins between which scores
should be divided.  The histogram in Figure&nbsp;<A HREF="node12.html#histfig">19</A> was
generated using five bins.

<P>

<H3><A NAME="SECTION0001210200000000000000"></A><A NAME="sec:makeroc"></A>
<BR>
10.10.2 <TT>makeroc</TT>
</H3>

<P>
The <TT>makeroc</TT> program generates a postscript graph file using the
<TT>gnuplot</TT> plotting package.  It takes two <TT>.dist</TT> distance
files created by <TT>hmmscore</TT> as input and plots false negatives/
false positives on the axis Score vs. Counts where Score is the
NLL-NULL score of a sequence and Counts is the number of sequences
in the files with a particular score.

<P>
To plot false negatives vs.&nbsp;false positives in two distance files
<TT>globins.dist</TT> and <TT>nonglobins.dist</TT>, the following command
would be used:

<P>
<PRE>

makeroc test -distfile globins.dist -distfile2 nonglobins.dist
</PRE>

<P>
The program will output a file <TT>test.ps</TT> which you can view with
<TT>ghostview</TT> or other postscript viewer.

<P>

<P></P>
<DIV ALIGN="CENTER"><A NAME="makerocfig"></A><A NAME="6573"></A>
<TABLE>
<CAPTION ALIGN="BOTTOM"><STRONG>Figure 20:</STRONG>
A plot generated by <TT>makeroc</TT> and <TT>gnuplot</TT>.</CAPTION>
<TR><TD><IMG
 WIDTH="384" HEIGHT="404" BORDER="0"
 SRC="img79.png"
 ALT="\begin{figure}\centerline{\psfig{figure=makeroc.ps}}\par\end{figure}"></TD></TR>
</TABLE>
</DIV><P></P>

<P>

<H3><A NAME="SECTION0001210300000000000000"></A><A NAME="sec:makeroc2"></A>
<BR>
10.10.3 <TT>makeroc2</TT>
</H3>

<P>
The <TT>makeroc2</TT> program generates a postscript graph file using
the <TT>gnuplot</TT> plotting package.  It takes two <TT>.dist</TT> distance
files created by <TT>hmmscore</TT> as input and plots false negative
vs. false positive scores, each of which are a function of the setting
of threshold score.  At each integral point on, for example, the false
positive axis, the value of the false negative coordinate is a linear
interpolation between the two nearest integral false negative values
based on the scores of the two false negative points and the false
positive point in question.  This means, for example, that if two
negative examples score 10 and 20, and two positive examples score
19 and 21, the point (1,1.1) will be plotted for the score of 19,
indicating that if the threshold is set at 19, there will be one
false negative (the sequence that scores 21) and a little over one
false positive (the sequence scoring 19 is 10two negative examples).

<P>
To plot false negatives vs.&nbsp;false positives in two distance files
<TT>globins.dist</TT> and <TT>nonglobins.dist</TT>, the following command
would be used:

<P>
<PRE>

makeroc2 test -distfile globins.dist -distfile2 nonglobins.dist
</PRE>

<P>
The program will output a file <TT>test.ps</TT> which you can view with
<TT>ghostview</TT> or other postscript viewer.

<P>
To plot multiple pairs of distance files on the same axis, use the
perl script <TT>multi_roc2.pl</TT>.
<PRE>

multi_roc2.pl true1.dist false1.dist title1 true2.dist false2.dist title2
true3.dist false3.dist title3 . . .
</PRE>

<P>
The titles appear in a legend in the upper right hand corner of the plot
and identify the individual curves.

<P>

<P></P>
<DIV ALIGN="CENTER"><A NAME="makeroc2fig"></A><A NAME="6576"></A>
<TABLE>
<CAPTION ALIGN="BOTTOM"><STRONG>Figure 21:</STRONG>
A plot generated by <TT>multi_roc2.pl</TT>, <TT>makeroc2</TT> and <TT>gnuplot</TT>.</CAPTION>
<TR><TD><IMG
 WIDTH="556" HEIGHT="403" BORDER="0"
 SRC="img80.png"
 ALT="\begin{figure}\centerline{\psfig{figure=makeroc2.ps}}\par\end{figure}"></TD></TR>
</TABLE>
</DIV><P></P>

<P>

<H3><A NAME="SECTION0001210400000000000000">
10.10.4 <TT>makeroc3</TT></A>
</H3>

<P>
The third of the series is <TT>makeroc3</TT>.  This program is given two
data files, as with <TT>makeroc2</TT> and prints numerical information
about a positive distance file and a negative distance file.  With the
command
<PRE>

makeroc2 test -distfile true.dist -distfile2 false.dist
</PRE>
the program will output a <TT>test.dat</TT> file of the form
<PRE>

flips between x:2.000000, y:2.280622 score:-36.376999 and x:3.000000, y:2.263981 score:-36.743000
3 sequences at FP=FN
7 FN at FP=0
2 FP at FN=0
</PRE>

<P>

<H2><A NAME="SECTION0001211000000000000000">
10.11 Sequence manipulation</A>
</H2>

<P>
The  programs described in this section can be used to perform
many helpful tasks.

<P>

<H3><A NAME="SECTION0001211100000000000000"></A><A NAME="sec:checkseq"></A>
<BR>
10.11.1 <TT>checkseq</TT>
</H3>

<P>
The <TT>checkseq</TT> program will read a sequence file and list various
pieces of information about it. For example,
<PRE>

checkseq unused_runname -db trna10.seq
</PRE>
will produce the output
<PRE>

Alphabet:	RNA
Input&nbsp;Format:	ig
Alignment:	no
AlignType:	unaligned
AlignColumns:	0
Num&nbsp;Sequences:	1000000
Average&nbsp;Length:	74.70
Max&nbsp;Length:	77
Min&nbsp;Length:	0
Total&nbsp;Length:	747
</PRE>

<P>
SAM internally recognizes several types of alignments.  The check
proceeds as follows:  if the file is an HSSP file, it is considered an
alignment.  Otherwise, if the file is an a2m format file with the same
number of upper case (match) and hyphen (delete) characters in each
line, it is an a2m format alignment.  Otherwise, if all characters
have the same total number of upper case, lower case, hyphen, and
period characters, it is an ``all positions alignment'' -- if all
positions are regarded as match columns, the file can be viewed as an
alignment.

<P>
If the <TT>sort</TT> variable is set on the command line, <TT>checkseq</TT>
will indicate whether or not all IDs are unique and whether or not all
sequences are unique.  This can be time consuming for large data
files, which is why <TT>checkseq</TT> will ignore the default setting of
<TT>sort</TT> if it is non-zero.

<P>

<H3><A NAME="SECTION0001211200000000000000"></A><A NAME="sec:genseq"></A>
<BR>
10.11.2 <TT>genseq</TT>
</H3>

<P>
The <TT>genseq</TT> program generates random sequences from either a
standard 1-component regularizer, such as SAM's default per-alphabet
regularizers, or a Dirichlet mixture regularizer.  In this case, the
Dirichlet mixture must be a mixture of sequence compositions, not a
mixture of alignment column distributions, as is used in <TT>buildmodel</TT>.

<P>
<PRE>

genseq rseqs -nseq 20 -a protein
</PRE>

<P>
The SAM system includes default multi-component regularizers for the
protein alphabet (<TT>rsdb-comp2.32comp</TT>) and the EHL2 secondary
structure alphabet (<TT>t99-t90comp.9comp</TT>), which are the default
settings of the <TT>genprot_prior</TT> and <TT>genrhl2_prior</TT>
variables.  If a <TT>track_prior</TT> is not specified as a single
prior file (<TT>genseq</TT> cannot generate multi-track sequences) and
alphabet is protein or EHL2, the <TT>genprot_prior</TT> or <TT>genrhl2_prior</TT> is used.  Otherwise, the <TT>regularizer_file</TT> is
used, or the default internal regularizer.

<P>
The length distribution of the generated sequences is based on the
normal distribution.  The mean and standard deviation are presented in
(natural) log format as <TT>gs_mean_log_len</TT> (default 5.4151) and
<TT>gs_sd_log_len</TT> (default 1.03632564).  The default values are
intended for protein sequences.

<P>
The sophisticated distribution sampling source code (designed and
written by Kevin Karplus) at the core of these routines is availble at:
<BR><A NAME="tex2html64"
  HREF="http://www.cse.ucsc.edu/research/compbio/dirichlets/index.html"><TT>http://www.cse.ucsc.edu/research/compbio/dirichlets/index.html</TT></A>
<P>

<H3><A NAME="SECTION0001211300000000000000"></A><A NAME="sec:permuteseq"></A>
<BR>
10.11.3 <TT>permuteseq</TT>
</H3>

<P>
The <TT>permuteseq</TT> program requires a run name and a database file.
Its output will be a file of sequences that are permuted copies of the
sequences in the database file.  If <TT>Nseq</TT> is specified, the
sequence file will be looped through multiple times until the
requested total number of sequences are created.  That is, in a file
of 10 sequences, if <TT>Nseq</TT> is 15, the output <TT>.seq</TT> file will
contain 10 random sequences, one for each original sequence, followed
by 5 random sequences, one for each of the first 5 sequences in the file.

<P>
The <TT>permuteseq</TT> program can be used to verify scoring results.
For example, the command

<P>
<PRE>

permuteseq permuted -db trna10.seq -Nseq 8 -id TRNA1 -id TRNA2 -id TRNA3
hmmscore permuted -i test.mod -db permuted.seq -sw 2
</PRE>

<P>
produces the following distance file, which shows all the permuted
tRNA sequences as having poor scores.

<P>
<PRE>

Rand3-TRNA3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-17.47&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-8.71&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.33e-03
Rand6-TRNA3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-8.84&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.55e+00
Rand1-TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.47&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.25&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.50e+00
Rand5-TRNA2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.28&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-0.19&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.63e+00
Rand7-TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.22&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.09&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.18e+00
Rand2-TRNA2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.65&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;5.25e+00
Rand4-TRNA1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;72&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-6.99&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.75&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;5.43e+00
Rand8-TRNA2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;76&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-7.20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.07&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;5.96e+00
</PRE>

<P>

<H3><A NAME="SECTION0001211400000000000000"></A><A NAME="sec:randseq"></A>
<BR>
10.11.4 <TT>randseq</TT>
</H3>

<P>
The <TT>randseq</TT> program can be used to randomly sample a database
file.  It requires a database file (<TT>db</TT>), and a number of
sequences (<TT>Nseq</TT>), and can optionally be provided a random
number seed (<TT>trainseed</TT>).  For example,
<PRE>

randseq rseqs -db nrp -nseq 1000
</PRE>
would place 1000 sequences randomly selected from <TT>nrp</TT> into the
file <TT>rseqs.seq</TT>.

<P>
The <TT>randseq</TT> program performs the random selection in
memory.

<P>

<H3><A NAME="SECTION0001211500000000000000"></A><A NAME="sec:splitseq"></A>
<BR>
10.11.5 <TT>splitseq</TT>
</H3>

<P>
The <TT>splitseq</TT> program can be used to split a database according
to a sequence length threshold.  It requires a database file (<TT>db</TT>), and a maximum sequence length (<TT>max_seq_length</TT>).
For example,
<PRE>

splitseq split -db nrp -max_seq_length 1000
</PRE>
would place sequences of length less than or equal to 1000
in <TT>splitA.seq</TT> and sequences of length greater than 1000 in <TT>splitB.seq</TT>. 

<P>
The <TT>splitseq</TT> program performs the selection in
memory.  It is particularly useful with posterior-decoded alignment
styles because they are not yet implemented with a reduced space
algorithm. 

<P>

<H3><A NAME="SECTION0001211600000000000000"></A> <A NAME="sec:sample-seqs"></A>
<BR>
10.11.6 <TT>sampleseqs</TT>
</H3>

<P>
The <TT>sampleseqs</TT> program will, given a model, randomly generate
sequences according to the model's probabilities.  For example,
<PRE>

sampleseqs sample -i test.mod -Nseq 5
</PRE>
will produce the following sample.seq file of synthetic tRNAs: 
<PRE>

&gt;SampleSeq0
GGGAUGGAGCGUAAUCGCCAUGGCAGAGGGGACCGGACUUGCAAUGAGUC
ACACCGGUUCAAACCCCCGCCACGCCACCC
&gt;SampleSeq1
GGGGACACGGCUCAGUCUGGAGAUCGGGCGCGCCUGGCAUACCGGAGGUC
UGAGCUGAAGUCGCGCAGGCUCCUCCA
&gt;SampleSeq2
GCGCGAUCGCUCAGCUGGGAUAGCACACUCCUUGCGACGGGAAGCGCCCG
ACGCCCGGUGGUUCCUCCA
&gt;SampleSeq3
GCGAGAAUAGCGCAGACCUUGGGACACGGUAUUGGCUUGGAAAACAUGGG
GGGCGAGUUAAAAUCCGGGUGUGGCUGCCA
&gt;SampleSeq4
GUGGACGUAGUUCAAUCUGUACUUGGGACACUAGCCUUAGAAUCCUGUGU
GCACAUCUUCGAGUGCCCCCCGUGCCGUUCCA
&gt;SampleSeq5
GCGGCCAUAGCCGAGCCUGGAUAGCGGACUGGGUUUCUUCGAUGACCGAG
UUCAAUUCCCGGCCGCCCCUCCA
&gt;SampleSeq6
GCGCACGUAGCUCGCCAGUUUGUAUAGCGGACCGGCCUCUUGACCAUGCG
AUGAGGAGCGUGGAAUCCCAGCGGGCCCACCA
&gt;SampleSeq7
GCGGCCGUGCUUCAGGUUCGAGAGCACGCGUCGCUUACGAGGCAGGGCAA
CCAGAGUUCCGAUCCCGCGGGUUCAUCCA
&gt;SampleSeq8
GGCGCCGUAGCGCCCGUCGGGCGCGGACGUCGGACUCUUGCCGGACGUCA
CUGUUCGGGCCCCCUGGGCCUGGGCA
&gt;SampleSeq9
UUCCCAAAGUCCAGGUGGGGCACGACGCCUGACUCGGAGUCGGUGUUCGA
GUAGGUUCAAUUCUCGUCUUACAGCA
</PRE>

<P>
Run with no arguments for a usage message.

<P>

<H3><A NAME="SECTION0001211700000000000000"></A> <A NAME="sec:sortseq"></A>
<BR>
10.11.7 <TT>sortseq</TT>
</H3>

<P>
The <TT>sortseq</TT> program takes as input a database file(s) of sequences
(<TT>*.sel</TT> or <TT>*.seq</TT>) or an input <TT>-alignfile</TT> (<TT>*.a2m</TT>
or <TT>*.mult</TT>) and a distance file (<TT>*.dist</TT>) of 
scores generated by <TT>hmmscore</TT>.  It outputs a file of sequences or
alignment (depending on input)
in the same order as they occur in the distance file.  <TT>Sortseq</TT>
provides an enhancement to <TT>hmmscore</TT>.  It outputs a <TT>.seq</TT> or
<TT>.a2m</TT> file that contains the actual sequences in sorted order
while <TT>hmmscore</TT> 
outputs a file only containing sorted sequence id's and scores.

<P>
For example,
<PRE>

sortseq sort -db trna10.seq -distfile test.dist
</PRE>
will produce the following sort.seq file of sequences:
<PRE>

&gt;TRNA8
GGGCCCGUGGCCUAGUCUGGAUACGGCACCGGCCUUCUAAGCCGGGGAUC
GGGGGUUCAAAUCCCUCCGGGUCCG
&gt;TRNA10
UCCGUCGUAGUCUAGGUGGUUAGGAUACUCGGCUUUCACCCGAGAGACCC
GGGUUCAAGUCCCGGCGACGGAACCA
&gt;TRNA2
GCGGCCGUCGUCUAGUCUGGAUUAGGACGCUGGCCUCCCAAGCCAGCAAU
CCCGGGUUCGAAUCCCGGCGGCCGCA
&gt;TRNA7
GGGCACAUGGCGCAGUUGGUAGCGCGCUUCCCUUGCAAGGAAGAGGUCAU
CGGUUCGAUUCCGGUUGCGUCCA
&gt;TRNA1
GGGGAUGUAGCUCAGUGGUAGAGCGCAUGCUUCGCAUGUAUGAGGCCCCG
GGUUCGAUCCCCGGCAUCUCCA
&gt;TRNA4
GGGCGAAUAGUGUCAGCGGGAGCACACCAGACUUGCAAUCUGGUAGGGAG
GGUUCGAGUCCCUCUUUGUCCACCA
&gt;TRNA9
CGGCACGUAGCGCAGCCUGGUAGCGCACCGUCCUGGGGUUGCGGGGGUCG
GAGGUUCAAAUCCUCUCGUGCCGACCA
&gt;TRNA3
GGCCCUGUGGCUAGCUGGUCAAAGCGCCUGUCUAGUAAACAGGAGAUCCU
GGGUUCGAAUCCCAGCGGGGCCUCCA
&gt;TRNA5
GCCGGGAUAGCUCAGUUGGUAGAGCAGAGGACUGAAAAUCCUCGUGUCAC
CAGUUCAAAUCUGGUUCCUGGCA
&gt;TRNA6
GGGGCCUUAGCUCAGCUGGGAGAGCGCCUGCUUUGCACGCAGGAGGUCAG
CGGUCGACCCGCUAGGCUCCACCA
</PRE>

<P>

<H3><A NAME="SECTION0001211800000000000000"></A> <A NAME="sec:uniqueseq"></A>
<BR>
10.11.8 <TT>uniqueseq</TT>
</H3>

<P>
The <TT>uniqueseq</TT> program takes as input a database file(s) of
sequences (<TT>*.sel</TT> or <TT>*.seq</TT>) or a single alignment file.  It
sorts the sequences and outputs a file containing every sequence with
a unique ID (i.e. no dupes are copied to the output file).  It also
outputs messages to the user informing of the following conditions in
the database file: same id for different sequences; same sequence for
different ids; and duplicate id found.

<P>
If an alignment file is specified, a sequence is considered the same
if <TT>percent_id</TT> of the sequence's match characters are present in
another sequence.  If two sequences mutually satisfy this threshold,
the one that is more identical to the other is removed, or the one
that has the fewest total number of match columns if their relative
percentages are the same.  The order in which this is performed is
undefined, so that when <TT>percent_id</TT> is less than the default
1.0, it is possible to get to different results with different
orderings of an alignment due to transitivity considerations.  If <TT>percent_id</TT> is negative, a message is printed about each sequence
that is dropped.

<P>
If the optional <TT>-train
mytrainfile.seq</TT> is used, the program checks the sequences in the <TT>trainfile</TT> and writes any sequence not in the database file to the
output file.  In this case, the following user information messages
are output: training sequence (ID number) is in database with
different ID (ID number); training sequence (ID number) is in database
with same ID; training sequence (ID number) not in database (note: the
program currently checks only one training file).

<P>
For example,
<PRE>

uniqueseq unique -db testme.seq -train trna10.seq
uniqueseq unique1 -alignfile testme.a2m
</PRE>
will produce <TT>unique.seq</TT> and  <TT>unique1.a2m</TT>, respectively. 

<P>
Messages about which sequences are dropped are printed to standard
error. 

<P>

<H3><A NAME="SECTION0001211900000000000000"></A> <A NAME="sec:mkkestreldb"></A>
<BR>
10.11.9 <TT>mk_kestrel_db</TT>
</H3>

<P>
The <TT>mk_kestrel_db</TT> command creates databases for use with
Kestrel <TT>hmmscore</TT>.  These databases are in a format optimized for
scoring with the Kestrel processor and parallel the standard sequence
databases.
For each <TT>db</TT> parameter specified, <TT>mk_kestrel_db</TT> will
create a <TT>.kids</TT> file and either a <TT>.kseq</TT> file or a
<TT>.krseq</TT> if <TT>subtract_null</TT> is set to <TT>4</TT>. 
These files will be in the same directory as the <TT>db</TT> file
and will have the same name, with the extension appended.  The
<TT>runname</TT> argument is ignored.

<P>
<HR>
<!--Navigation Panel-->
<A NAME="tex2html457"
  HREF="node13.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<A NAME="tex2html453"
  HREF="sam_doc.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up.png"></A> 
<A NAME="tex2html447"
  HREF="node11.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev.png"></A> 
<A NAME="tex2html455"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html458"
  HREF="node13.html">11 System installation</A>
<B> Up:</B> <A NAME="tex2html454"
  HREF="sam_doc.html">SAM (Sequence Alignment and</A>
<B> Previous:</B> <A NAME="tex2html448"
  HREF="node11.html">9 The buildmodel estimation</A>
 &nbsp <B>  <A NAME="tex2html456"
  HREF="node1.html">Contents</A></B> 
<!--End of Navigation Panel-->
<ADDRESS>
<a href=http://www.cse.ucsc.edu/research/compbio/sam.html>SAM</a><BR><a href=mailto:sam-info@cse.ucsc.edu>sam-info@cse.ucsc.edu</a><BR><A HREF=http://www.cse.ucsc.edu/research/compbio>UCSC Computational Biology Group</A>
</ADDRESS>
</BODY>
</HTML>
