<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">

<!--Converted with LaTeX2HTML 2K.1beta (1.47)
original version by:  Nikos Drakos, CBLU, University of Leeds
* revised and updated by:  Marcus Hennecke, Ross Moore, Herb Swan
* with significant contributions from:
  Jens Lippmann, Marek Rouchal, Martin Wilck and others -->
<HTML>
<HEAD>
<TITLE>Bibliography</TITLE>
<META NAME="description" CONTENT="Bibliography">
<META NAME="keywords" CONTENT="sam_doc">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">

<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META NAME="Generator" CONTENT="LaTeX2HTML v2K.1beta">
<META HTTP-EQUIV="Content-Style-Type" CONTENT="text/css">

<LINK REL="STYLESHEET" HREF="sam_doc.css">

<LINK REL="next" HREF="node7.html">
<LINK REL="previous" HREF="node5.html">
<LINK REL="up" HREF="sam_doc.html">
<LINK REL="next" HREF="node7.html">
</HEAD>

<BODY >
<!--Navigation Panel-->
<A NAME="tex2html355"
  HREF="node7.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<A NAME="tex2html351"
  HREF="sam_doc.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up.png"></A> 
<A NAME="tex2html345"
  HREF="node5.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev.png"></A> 
<A NAME="tex2html353"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html356"
  HREF="node7.html">5 File types</A>
<B> Up:</B> <A NAME="tex2html352"
  HREF="sam_doc.html">SAM (Sequence Alignment and</A>
<B> Previous:</B> <A NAME="tex2html346"
  HREF="node5.html">4 SAM-T99</A>
 &nbsp <B>  <A NAME="tex2html354"
  HREF="node1.html">Contents</A></B> 
<BR><BR>
<!--End of Navigation Panel-->

<H2><A NAME="SECTION00060000000000000000">
Bibliography</A>
</H2><DL COMPACT><DD>
S.&nbsp;Altschul, T.&nbsp;Madden, A.&nbsp;Schaffer, J.&nbsp;Zhang, Z.&nbsp;Zhang, W.&nbsp;Miller, and
  D.&nbsp;Lipman.
Gapped BLAST and PSI-BLAST: A new generation of protein database
  search programs.
<EM>Nucleic Acids Research</EM>, 25:3899-3402, 1997.

Stephen&nbsp;F. Altschul.
Amino acid substitution matrices from an information theoretic
  perspective.
<EM>Journal of Molecular Biology</EM>, 219:555-565, 1991.

D.&nbsp;Haussler, A.&nbsp;Krogh, I.&nbsp;S. Mian, and K.&nbsp;Sj&#246;lander.
Protein modeling using hidden Markov models: Analysis of globins.
In <EM>Proceedings of the Hawaii International Conference on System
  Sciences</EM>, volume&nbsp;1, pages 792-802, Los Alamitos, CA, 1993. IEEE Computer
  Society Press.

Steven Henikoff and Jorja&nbsp;G. Henikoff.
Position-based sequence weights.
<EM>Journal of Molecular Biology</EM>, 243(4):574-578, November 1994.

Liisa Holm and Chris Sander.
Mapping the protein universe.
<EM>Science</EM>, 273(5275):595-603, Aug 2 1996.

R.&nbsp;Karchin and R.&nbsp;Hughey.
Weighting hidden Markov models for maximum discrimination.
<EM>Bioinformatics</EM>, 14(9):772-782, 1998.

A.&nbsp;G. Murzin, S.&nbsp;E. Brenner, T.&nbsp;Hubbard, and C.&nbsp;Chothia.
SCOP: a structural classification of proteins database for the
  investigation of sequences and structures.
<EM>Journal of Molecular Biology</EM>, 247:536-540, 1995.
Information on SCOP is available at <TT>  http://scop.mrc-lmb.cam.ac.uk/scop</TT>.

J.&nbsp;Park, K.&nbsp;Karplus, C.&nbsp;Barrett, R.&nbsp;Hughey, D.&nbsp;Haussler, T.&nbsp;Hubbard, and
  C.&nbsp;Chothia.
Sequence comparisons using multiple sequences detect three times as
  many remote homologues as pairwise methods.
<EM>Journal of Molecular Biology</EM>, 284(4):1201-1210, 1998.

J.&nbsp;Park, S.&nbsp;Teichmann, T.&nbsp;Hubbard, and C.&nbsp;Chothia.
Intermediate sequences increase the detection of homology between
  sequences.
<EM>Journal of Molecular Biology</EM>, 273:349-354, 1997.

William Pearson.
Comparison of methods for searching protein sequence databases.
<EM>Protein Science</EM>, 4:1145-1160, 1995.

C.&nbsp;Sander and R.&nbsp;Schneider.
Database of homology-derived protein structures and the structural
  meaning of sequence alignment.
<EM>Proteins: Structure, Function, and Genetics</EM>, 9(1):56-68, 1991.

E.L.L Sonnhammer, S.R. Eddy, and R.&nbsp;Durbin.
Pfam: A comprehensive database of protein families based on seed
  alignments.
<EM>Proteins: Structure, Function, and Genetics</EM>, 28:405-420, 1997.

Julie&nbsp;D. Thompson, Desmond&nbsp;G. Higgins, and Toby&nbsp;J. Gibson.
CLUSTAL W: Improving the sensitivity of progressive multiple
  sequence alignment through sequence weighting, position-specific gap
  penalties, and weight matrix choice.
<EM>Nucleic Acids Research</EM>, 22(22):4673-4680, 1994.
</DL>

<P>
<BR><HR>
<ADDRESS>
<a href=http://www.cse.ucsc.edu/research/compbio/sam.html>SAM</a><BR><a href=mailto:sam-info@cse.ucsc.edu>sam-info@cse.ucsc.edu</a><BR><A HREF=http://www.cse.ucsc.edu/research/compbio>UCSC Computational Biology Group</A>
</ADDRESS>
</BODY>
</HTML>
