<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">

<!--Converted with LaTeX2HTML 2K.1beta (1.47)
original version by:  Nikos Drakos, CBLU, University of Leeds
* revised and updated by:  Marcus Hennecke, Ross Moore, Herb Swan
* with significant contributions from:
  Jens Lippmann, Marek Rouchal, Martin Wilck and others -->
<HTML>
<HEAD>
<TITLE>SAM (Sequence Alignment and Modeling) documentation</TITLE>
<META NAME="description" CONTENT="SAM (Sequence Alignment and Modeling) documentation">
<META NAME="keywords" CONTENT="sam_doc">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">

<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META NAME="Generator" CONTENT="LaTeX2HTML v2K.1beta">
<META HTTP-EQUIV="Content-Style-Type" CONTENT="text/css">

<LINK REL="STYLESHEET" HREF="sam_doc.css">

<LINK REL="next" HREF="node1.html">
</HEAD>

<BODY >
<!--Navigation Panel-->
<A NAME="tex2html69"
  HREF="node1.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up_g.png"> 
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev_g.png"> 
<A NAME="tex2html67"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html70"
  HREF="node1.html">Contents</A>
 &nbsp <B>  <A NAME="tex2html68"
  HREF="node1.html">Contents</A></B> 
<BR>
<BR>
<!--End of Navigation Panel-->
<H1 ALIGN="CENTER"><FONT SIZE="+4"><B>SAM
<BR></B></FONT>
Sequence Alignment and Modeling
<BR>
Software System</H1>
<P ALIGN="CENTER"><STRONG><A NAME="tex2html1"
  HREF="http://www.cse.ucsc.edu/~rph/">Richard Hughey</A>
 - 
<A NAME="tex2html2"
  HREF="http://www.cse.ucsc.edu/~karplus/">Kevin Karplus</A>
 - 
<A NAME="tex2html3"
  HREF="http://www.sanger.ac.uk/~krogh/">Anders Krogh</A>
</STRONG></P>
<P ALIGN="CENTER"><STRONG>
<A NAME="tex2html4"
  HREF="mailto:sam-info@cse.ucsc.edu">sam-info@cse.ucsc.edu</A>
<BR><A NAME="tex2html5"
  HREF="http://www.cse.ucsc.edu/">Baskin Center for Computer Engineering and Science</A>
<BR><A NAME="tex2html6"
  HREF="http://www.ucsc.edu/">University of California</A>
<BR>
Santa Cruz, CA 95064
<BR>
Technical Report UCSC-CRL-99-11
<BR>
Updated for SAM Version 3.3.1
<BR>
December 20, 2001
<BR>&nbsp;
<BR>
Contributors:
<BR>
Mark Diekhans, Leslie Grate
<BR>
Christian Barrett, Michael Brown, Jonathan Casper, Melissa Cline, 
<BR>
Terry Figel, Rachel Karchin, Kimmen Sj&#246;lander, Christopher Tarnas
&nbsp;
<BR><FONT SIZE="-1">supersedes UCSC-CRL-95-07 and UCSC-CRL-96-22</FONT>
</STRONG></P>

<P>

<H3>Abstract:</H3>
<DIV>

<P>
The Sequence Alignment and Modeling system (SAM) is a collection of
flexible software tools for creating, refining, and using linear
hidden Markov models for biological sequence analysis.  The model
states can be viewed as representing the sequence of columns in a
multiple sequence alignment, with provisions for arbitrary
position-dependent insertions and deletions in each sequence.  The
models are trained on a family of protein or nucleic acid sequences
using an expectation-maximization algorithm and a variety of
algorithmic heuristics.  A trained model can then be used to both
generate multiple alignments and search databases for new members of
the family.

<P>
SAM includes programs and scripts for the SAM-T99 method of remote
homology detection. SAM-T99 is an iterative HMM search method for
creating an HMM from a single protein sequence or seed alignment using
iterative search of a protein database.  The method is currently the
most sensitive purely-sequence-based remote homology detection
algorithm.  SAM-T99 is based on successful methods created for the
CASP2 and CASP3 protein structure prediction experiments.

<P>
The algorithms and methods used by SAM have been described in several
pioneering papers from the University of California, Santa Cruz.
These papers, as well as the SAM software suite, several servers, and
links to  related sites are available 
on the World-Wide Web to
<DIV ALIGN="CENTER">
<A NAME="tex2html7"
  HREF="http://www.cse.ucsc.edu/research/compbio/sam.html"><TT>http://www.cse.ucsc.edu/research/compbio/sam.html</TT></A>

</DIV>

<P>
</DIV>
<P>

<P>
<BR><HR>
<!--Table of Child-Links-->
<A NAME="CHILD_LINKS"></A>

<UL>
<LI><A NAME="tex2html71"
  HREF="node1.html">Contents</A>
<LI><A NAME="tex2html72"
  HREF="node2.html">1 Introduction</A>
<UL>
<LI><A NAME="tex2html73"
  HREF="node2.html#SECTION00021000000000000000">1.1 Acknowledgments</A>
</UL>
<BR>
<LI><A NAME="tex2html74"
  HREF="node3.html">2 New to this version</A>
<UL>
<LI><A NAME="tex2html75"
  HREF="node3.html#SECTION00031000000000000000">2.1 Version 3.3.1</A>
<LI><A NAME="tex2html76"
  HREF="node3.html#SECTION00032000000000000000">2.2 Version 3.2.1</A>
<LI><A NAME="tex2html77"
  HREF="node3.html#SECTION00033000000000000000">2.3 Version 3.2</A>
<LI><A NAME="tex2html78"
  HREF="node3.html#SECTION00034000000000000000">2.4 Version 3.1</A>
<LI><A NAME="tex2html79"
  HREF="node3.html#SECTION00035000000000000000">2.5 Version 3.0</A>
</UL>
<BR>
<LI><A NAME="tex2html80"
  HREF="node4.html">3 Quick overview</A>
<UL>
<LI><A NAME="tex2html81"
  HREF="node4.html#SECTION00041000000000000000">3.1 Building a model</A>
<LI><A NAME="tex2html82"
  HREF="node4.html#SECTION00042000000000000000">3.2 Aligning sequences</A>
<LI><A NAME="tex2html83"
  HREF="node4.html#SECTION00043000000000000000">3.3 Examining models</A>
<LI><A NAME="tex2html84"
  HREF="node4.html#SECTION00044000000000000000">3.4 Scoring sequences</A>
<LI><A NAME="tex2html85"
  HREF="node4.html#SECTION00045000000000000000">3.5 Parameter selection</A>
</UL>
<BR>
<LI><A NAME="tex2html86"
  HREF="node5.html">4 SAM-T99</A>
<UL>
<LI><A NAME="tex2html87"
  HREF="node5.html#SECTION00051000000000000000">4.1 SAM-T99 for superfamily modeling</A>
<LI><A NAME="tex2html88"
  HREF="node5.html#SECTION00052000000000000000">4.2 Improved verification of homology</A>
<LI><A NAME="tex2html89"
  HREF="node5.html#SECTION00053000000000000000">4.3 Family-level multiple alignments</A>
<LI><A NAME="tex2html90"
  HREF="node5.html#SECTION00054000000000000000">4.4 Modeling non-contiguous domains</A>
<LI><A NAME="tex2html91"
  HREF="node5.html#SECTION00055000000000000000">4.5 Building an HMM from a structural alignment</A>
<LI><A NAME="tex2html92"
  HREF="node5.html#SECTION00056000000000000000">4.6 Improving existing multiple alignments</A>
<LI><A NAME="tex2html93"
  HREF="node5.html#SECTION00057000000000000000">4.7 Creating a multiple alignment from unaligned sequences</A>
<LI><A NAME="tex2html94"
  HREF="node5.html#SECTION00058000000000000000">4.8 Parameters for <TT>target99</TT> perl script</A>
<LI><A NAME="tex2html95"
  HREF="node5.html#SECTION00059000000000000000">4.9 The model building scripts</A>
<LI><A NAME="tex2html96"
  HREF="node5.html#SECTION000510000000000000000">4.10 Installing SAM-T99</A>
<LI><A NAME="tex2html97"
  HREF="node5.html#SECTION000511000000000000000">4.11 SAM-T99 references</A>
</UL>
<BR>
<LI><A NAME="tex2html98"
  HREF="node6.html">Bibliography</A>
<LI><A NAME="tex2html99"
  HREF="node7.html">5 File types</A>
<LI><A NAME="tex2html100"
  HREF="node8.html">6 Parameter specification</A>
<LI><A NAME="tex2html101"
  HREF="node9.html">7 Sequence formats</A>
<UL>
<LI><A NAME="tex2html102"
  HREF="node9.html#SECTION00091000000000000000">7.1 Alphabets</A>
<UL>
<LI><A NAME="tex2html103"
  HREF="node9.html#SECTION00091100000000000000">7.1.1 User-defined alphabets</A>
</UL>
<LI><A NAME="tex2html104"
  HREF="node9.html#SECTION00092000000000000000">7.2 Sequences</A>
<LI><A NAME="tex2html105"
  HREF="node9.html#SECTION00093000000000000000">7.3 Training sets, test sets, and databases</A>
</UL>
<BR>
<LI><A NAME="tex2html106"
  HREF="node10.html">8 Regularizers and models</A>
<UL>
<LI><A NAME="tex2html107"
  HREF="node10.html#SECTION000101000000000000000">8.1 Regularizers</A>
<UL>
<LI><A NAME="tex2html108"
  HREF="node10.html#SECTION000101100000000000000">8.1.1 Regularizer alternatives</A>
<LI><A NAME="tex2html109"
  HREF="node10.html#SECTION000101200000000000000">8.1.2 Transition regularizers with
structural information</A>
</UL>
<LI><A NAME="tex2html110"
  HREF="node10.html#SECTION000102000000000000000">8.2 Initial model</A>
<LI><A NAME="tex2html111"
  HREF="node10.html#SECTION000103000000000000000">8.3 Initial alignment</A>
<LI><A NAME="tex2html112"
  HREF="node10.html#SECTION000104000000000000000">8.4 Model format</A>
<UL>
<LI><A NAME="tex2html113"
  HREF="node10.html#SECTION000104100000000000000">8.4.1 Model length</A>
<LI><A NAME="tex2html114"
  HREF="node10.html#SECTION000104200000000000000">8.4.2 Special nodes</A>
<LI><A NAME="tex2html115"
  HREF="node10.html#SECTION000104300000000000000">8.4.3 Node labels</A>
<LI><A NAME="tex2html116"
  HREF="node10.html#SECTION000104400000000000000">8.4.4 Binary models</A>
</UL>
<LI><A NAME="tex2html117"
  HREF="node10.html#SECTION000105000000000000000">8.5 Free-insertion modules</A>
<LI><A NAME="tex2html118"
  HREF="node10.html#SECTION000106000000000000000">8.6 FIM, insert and match tables</A>
</UL>
<BR>
<LI><A NAME="tex2html119"
  HREF="node11.html">9 The <TT>buildmodel</TT> estimation process</A>
<UL>
<LI><A NAME="tex2html120"
  HREF="node11.html#SECTION000111000000000000000">9.1 Noise and annealing</A>
<LI><A NAME="tex2html121"
  HREF="node11.html#SECTION000112000000000000000">9.2 Surgery</A>
<LI><A NAME="tex2html122"
  HREF="node11.html#SECTION000113000000000000000">9.3 Training statistics</A>
<LI><A NAME="tex2html123"
  HREF="node11.html#SECTION000114000000000000000">9.4 Weighted training</A>
<UL>
<LI><A NAME="tex2html124"
  HREF="node11.html#SECTION000114100000000000000">9.4.1 External weighting</A>
<UL>
<LI><A NAME="tex2html125"
  HREF="node11.html#SECTION000114110000000000000">9.4.1.1 Multi-subfamily weighting</A>
</UL>
<LI><A NAME="tex2html126"
  HREF="node11.html#SECTION000114200000000000000">9.4.2 Annealing with Weights</A>
<LI><A NAME="tex2html127"
  HREF="node11.html#SECTION000114300000000000000">9.4.3 Internal weighting of alignments</A>
<LI><A NAME="tex2html128"
  HREF="node11.html#SECTION000114400000000000000">9.4.4 Internal weighting during training</A>
</UL>
<LI><A NAME="tex2html129"
  HREF="node11.html#SECTION000115000000000000000">9.5 Viterbi training</A>
<LI><A NAME="tex2html130"
  HREF="node11.html#SECTION000116000000000000000">9.6 Building models with constraints</A>
<UL>
<LI><A NAME="tex2html131"
  HREF="node11.html#SECTION000116100000000000000">9.6.1 Model Node Labels</A>
<LI><A NAME="tex2html132"
  HREF="node11.html#SECTION000116200000000000000">9.6.2 Constraints Definition File</A>
<LI><A NAME="tex2html133"
  HREF="node11.html#SECTION000116300000000000000">9.6.3 Using Constraints</A>
</UL>
</UL>
<BR>
<LI><A NAME="tex2html134"
  HREF="node12.html">10 Related programs</A>
<UL>
<LI><A NAME="tex2html135"
  HREF="node12.html#SECTION000121000000000000000">10.1  <TT>align2model</TT> and <TT>prettyalign</TT></A>
<UL>
<LI><A NAME="tex2html136"
  HREF="node12.html#SECTION000121100000000000000">10.1.1 <TT>prettyalign</TT></A>
<LI><A NAME="tex2html137"
  HREF="node12.html#SECTION000121200000000000000">10.1.2 Aligning fragments</A>
</UL>
<LI><A NAME="tex2html138"
  HREF="node12.html#SECTION000122000000000000000">10.2 <TT>hmmscore</TT></A>
<UL>
<LI><A NAME="tex2html139"
  HREF="node12.html#SECTION000122100000000000000">10.2.1 NLL-NULL scoring</A>
<LI><A NAME="tex2html140"
  HREF="node12.html#SECTION000122200000000000000">10.2.2 Selecting sequences, scores, and alignments</A>
<LI><A NAME="tex2html141"
  HREF="node12.html#SECTION000122300000000000000">10.2.3 Scoring Fragments</A>
<LI><A NAME="tex2html142"
  HREF="node12.html#SECTION000122400000000000000">10.2.4 Selecting multiple domain alignments</A>
<LI><A NAME="tex2html143"
  HREF="node12.html#SECTION000122500000000000000">10.2.5 Multi-track HMM scoring</A>
<LI><A NAME="tex2html144"
  HREF="node12.html#SECTION000122600000000000000">10.2.6 Distributed scoring</A>
<LI><A NAME="tex2html145"
  HREF="node12.html#SECTION000122700000000000000">10.2.7 Smith &amp; Waterman Scoring and
Alignment</A>
<LI><A NAME="tex2html146"
  HREF="node12.html#SECTION000122800000000000000">10.2.8 Scoring using the Kestrel parallel processor</A>
<LI><A NAME="tex2html147"
  HREF="node12.html#SECTION000122900000000000000">10.2.9 Calibration and Model libraries</A>
</UL>
<LI><A NAME="tex2html148"
  HREF="node12.html#SECTION000123000000000000000">10.3 <TT>addfims</TT></A>
<UL>
<LI><A NAME="tex2html149"
  HREF="node12.html#SECTION000123100000000000000">10.3.1 Training with FIMs</A>
</UL>
<LI><A NAME="tex2html150"
  HREF="node12.html#SECTION000124000000000000000">10.4 <TT>grabdp</TT></A>
<LI><A NAME="tex2html151"
  HREF="node12.html#SECTION000125000000000000000">10.5 <TT>get_fisher_scores</TT></A>
<LI><A NAME="tex2html152"
  HREF="node12.html#SECTION000126000000000000000">10.6 <TT>modelfromalign</TT></A>
<LI><A NAME="tex2html153"
  HREF="node12.html#SECTION000127000000000000000">10.7 <TT>pathprobs</TT></A>
<LI><A NAME="tex2html154"
  HREF="node12.html#SECTION000128000000000000000">10.8 <TT>predict_track</TT></A>
<LI><A NAME="tex2html155"
  HREF="node12.html#SECTION000129000000000000000">10.9 Model manipulation</A>
<UL>
<LI><A NAME="tex2html156"
  HREF="node12.html#SECTION000129100000000000000">10.9.1 <TT>drawmodel</TT></A>
<LI><A NAME="tex2html157"
  HREF="node12.html#SECTION000129200000000000000">10.9.2 <TT>hmmconvert</TT></A>
<LI><A NAME="tex2html158"
  HREF="node12.html#SECTION000129300000000000000">10.9.3 <TT>makelogo</TT></A>
<LI><A NAME="tex2html159"
  HREF="node12.html#SECTION000129400000000000000">10.9.4 <TT>modifymodel</TT></A>
<LI><A NAME="tex2html160"
  HREF="node12.html#SECTION000129500000000000000">10.9.5 Conversion between SAM and HMMer</A>
<UL>
<LI><A NAME="tex2html161"
  HREF="node12.html#SECTION000129510000000000000">10.9.5.1 Internal HMM Structure</A>
<LI><A NAME="tex2html162"
  HREF="node12.html#SECTION000129520000000000000">10.9.5.2 Extra Information in HMMer</A>
<LI><A NAME="tex2html163"
  HREF="node12.html#SECTION000129530000000000000">10.9.5.3 Extra Information in SAM</A>
</UL>
</UL>
<LI><A NAME="tex2html164"
  HREF="node12.html#SECTION0001210000000000000000">10.10 Plotting Programs</A>
<UL>
<LI><A NAME="tex2html165"
  HREF="node12.html#SECTION0001210100000000000000">10.10.1 <TT>makehist</TT></A>
<LI><A NAME="tex2html166"
  HREF="node12.html#SECTION0001210200000000000000">10.10.2 <TT>makeroc</TT></A>
<LI><A NAME="tex2html167"
  HREF="node12.html#SECTION0001210300000000000000">10.10.3 <TT>makeroc2</TT></A>
<LI><A NAME="tex2html168"
  HREF="node12.html#SECTION0001210400000000000000">10.10.4 <TT>makeroc3</TT></A>
</UL>
<LI><A NAME="tex2html169"
  HREF="node12.html#SECTION0001211000000000000000">10.11 Sequence manipulation</A>
<UL>
<LI><A NAME="tex2html170"
  HREF="node12.html#SECTION0001211100000000000000">10.11.1 <TT>checkseq</TT></A>
<LI><A NAME="tex2html171"
  HREF="node12.html#SECTION0001211200000000000000">10.11.2 <TT>genseq</TT></A>
<LI><A NAME="tex2html172"
  HREF="node12.html#SECTION0001211300000000000000">10.11.3 <TT>permuteseq</TT></A>
<LI><A NAME="tex2html173"
  HREF="node12.html#SECTION0001211400000000000000">10.11.4 <TT>randseq</TT></A>
<LI><A NAME="tex2html174"
  HREF="node12.html#SECTION0001211500000000000000">10.11.5 <TT>splitseq</TT></A>
<LI><A NAME="tex2html175"
  HREF="node12.html#SECTION0001211600000000000000">10.11.6 <TT>sampleseqs</TT></A>
<LI><A NAME="tex2html176"
  HREF="node12.html#SECTION0001211700000000000000">10.11.7 <TT>sortseq</TT></A>
<LI><A NAME="tex2html177"
  HREF="node12.html#SECTION0001211800000000000000">10.11.8 <TT>uniqueseq</TT></A>
<LI><A NAME="tex2html178"
  HREF="node12.html#SECTION0001211900000000000000">10.11.9 <TT>mk_kestrel_db</TT></A>
</UL>
</UL>
<BR>
<LI><A NAME="tex2html179"
  HREF="node13.html">11 System installation</A>
<UL>
<LI><A NAME="tex2html180"
  HREF="node13.html#SECTION000131000000000000000">11.1 Environment variables</A>
<LI><A NAME="tex2html181"
  HREF="node13.html#SECTION000132000000000000000">11.2 Runtime statistics</A>
<LI><A NAME="tex2html182"
  HREF="node13.html#SECTION000133000000000000000">11.3 Reducing runtime</A>
<LI><A NAME="tex2html183"
  HREF="node13.html#SECTION000134000000000000000">11.4 Future Features</A>
<LI><A NAME="tex2html184"
  HREF="node13.html#SECTION000135000000000000000">11.5 Prior versions</A>
<UL>
<LI><A NAME="tex2html185"
  HREF="node13.html#SECTION000135100000000000000">11.5.1 Version 2.2.1</A>
<LI><A NAME="tex2html186"
  HREF="node13.html#SECTION000135200000000000000">11.5.2 Version 2.2</A>
<LI><A NAME="tex2html187"
  HREF="node13.html#SECTION000135300000000000000">11.5.3 Version 2.1.2</A>
<LI><A NAME="tex2html188"
  HREF="node13.html#SECTION000135400000000000000">11.5.4 Version 2.1.1</A>
<LI><A NAME="tex2html189"
  HREF="node13.html#SECTION000135500000000000000">11.5.5 Version 2.1</A>
<LI><A NAME="tex2html190"
  HREF="node13.html#SECTION000135600000000000000">11.5.6 Version 2.0</A>
<LI><A NAME="tex2html191"
  HREF="node13.html#SECTION000135700000000000000">11.5.7 Version 1.4</A>
<LI><A NAME="tex2html192"
  HREF="node13.html#SECTION000135800000000000000">11.5.8 Version 1.3</A>
<LI><A NAME="tex2html193"
  HREF="node13.html#SECTION000135900000000000000">11.5.9 Version 1.2</A>
<LI><A NAME="tex2html194"
  HREF="node13.html#SECTION0001351000000000000000">11.5.10 Version 1.1</A>
<LI><A NAME="tex2html195"
  HREF="node13.html#SECTION0001351100000000000000">11.5.11 Version 1.0</A>
</UL>
</UL>
<BR>
<LI><A NAME="tex2html196"
  HREF="node14.html">12 Parameter descriptions</A>
<LI><A NAME="tex2html197"
  HREF="node15.html">About this document ...</A>
</UL>
<!--End of Table of Child-Links-->
<HR>
<!--Navigation Panel-->
<A NAME="tex2html69"
  HREF="node1.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up_g.png"> 
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev_g.png"> 
<A NAME="tex2html67"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html70"
  HREF="node1.html">Contents</A>
 &nbsp <B>  <A NAME="tex2html68"
  HREF="node1.html">Contents</A></B> 
<!--End of Navigation Panel-->
<ADDRESS>
<a href=http://www.cse.ucsc.edu/research/compbio/sam.html>SAM</a><BR><a href=mailto:sam-info@cse.ucsc.edu>sam-info@cse.ucsc.edu</a><BR><A HREF=http://www.cse.ucsc.edu/research/compbio>UCSC Computational Biology Group</A>
</ADDRESS>
</BODY>
</HTML>
