<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">

<!--Converted with LaTeX2HTML 2K.1beta (1.47)
original version by:  Nikos Drakos, CBLU, University of Leeds
* revised and updated by:  Marcus Hennecke, Ross Moore, Herb Swan
* with significant contributions from:
  Jens Lippmann, Marek Rouchal, Martin Wilck and others -->
<HTML>
<HEAD>
<TITLE>1 Introduction</TITLE>
<META NAME="description" CONTENT="1 Introduction">
<META NAME="keywords" CONTENT="sam_doc">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">

<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META NAME="Generator" CONTENT="LaTeX2HTML v2K.1beta">
<META HTTP-EQUIV="Content-Style-Type" CONTENT="text/css">

<LINK REL="STYLESHEET" HREF="sam_doc.css">

<LINK REL="next" HREF="node3.html">
<LINK REL="previous" HREF="node1.html">
<LINK REL="up" HREF="sam_doc.html">
<LINK REL="next" HREF="node3.html">
</HEAD>

<BODY >
<!--Navigation Panel-->
<A NAME="tex2html285"
  HREF="node3.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<A NAME="tex2html281"
  HREF="sam_doc.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up.png"></A> 
<A NAME="tex2html275"
  HREF="node1.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev.png"></A> 
<A NAME="tex2html283"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html286"
  HREF="node3.html">2 New to this</A>
<B> Up:</B> <A NAME="tex2html282"
  HREF="sam_doc.html">SAM (Sequence Alignment and</A>
<B> Previous:</B> <A NAME="tex2html276"
  HREF="node1.html">Contents</A>
 &nbsp <B>  <A NAME="tex2html284"
  HREF="node1.html">Contents</A></B> 
<BR>
<BR>
<!--End of Navigation Panel-->
<!--Table of Child-Links-->
<A NAME="CHILD_LINKS"><STRONG>Subsections</STRONG></A>

<UL>
<LI><A NAME="tex2html287"
  HREF="#SECTION00021000000000000000">1.1 Acknowledgments</A>
</UL>
<!--End of Table of Child-Links-->
<HR>

<H1><A NAME="SECTION00020000000000000000"></A><A NAME="sec:intro"></A>
<BR>
1 Introduction
</H1>

<P>
The Sequence Alignment and Modeling system (SAM) is a collection of
software tools for creating, refining, and using a type of statistical
model called a linear hidden Markov model for biological sequence
analysis. Linear hidden Markov models only model primary structure
(sequence) information; long-range iterations, such as base pairing
in RNA, require more complex models such as stochastic context-free
grammars, as described by Sakakibara <EM>et.&nbsp;al</EM> (NAR
22(23):5112-5120), also available from the UCSC computational biology
WWW site.

<P>
The algorithms and methods have been described in several papers, some
of which are available  on our WWW site,
<DIV ALIGN="CENTER">
<A NAME="tex2html8"
  HREF="http://www.cse.ucsc.edu/research/compbio/sam.html"><TT>http://www.cse.ucsc.edu/research/compbio/sam.html</TT></A>.

</DIV>

<P>
A tutorial on the use of SAM and the iterative SAM-T98 method
(the direct predecessor of SAM-T99)
is also available at our WWW site,
<DIV ALIGN="CENTER">
<A NAME="tex2html9"
  HREF="http://www.cse.ucsc.edu/research/compbio/ismb99.tutorial.html"><TT>http://www.cse.ucsc.edu/research/compbio/ismb99.tutorial.html</TT></A>.

</DIV>

<P>
SAM-T99 is used at the official SCOP Superfamily server at
<A NAME="tex2html22"
  HREF="http://stash.mrc-lmb.cam.ac.uk/SUPERFAMILY">http://stash.mrc-lmb.cam.ac.uk/SUPERFAMILY</A>
<P>
The primary papers from UCSC (copies of these papers and several
others are available from the SAM WWW site) include:

<P>


<P>

<UL>
<LI>R.&nbsp;Hughey and A.&nbsp;Krogh.  
<A NAME="tex2html10"
  HREF="http://www.cse.ucsc.edu/research/compbio/papers/HughKrogh96.ps.Z">Hidden Markov models for sequence analysis: Extension and analysis of
the basic method</A>,
CABIOS, 12(2):95-107, 1996.

<P>
</LI>
<LI>K. Karplus, C. Barrett, and R. Hughey,
<A NAME="tex2html11"
  HREF="http://www.cse.ucsc.edu/research/compbio/papers/remotehomology.pdf">Hidden Markov Models
for Detecting Remote Protein Homologies</A>,
<EM>Bioinformatics,</EM> 14(10):846-856, 1998.

<P>
</LI>
<LI>A.&nbsp;Krogh, M.&nbsp;Brown, I.&nbsp;S. Mian, K.&nbsp;Sj&#246;lander, and D.&nbsp;Haussler.
<A NAME="tex2html12"
  HREF="ftp://ftp.cse.ucsc.edu/pub/protein/hmm.part1.ps.Z">Hidden Markov models in computational
biology: Applications to 
  protein modeling.</A>
<EM>Journal of Molecular Biology</EM>, 235:1501-1531, February 1994.

<P>
</LI>
<LI>K. Karplus, R. Karchin, C. Barrett, S. Tu, M. Cline,
  M. Diekhans, L. Grate, J. Casper, and R. Hughey,
``What is the value added by human intervention in protein structure
  prediction?,''
 <EM>Proteins: Structure, Function, and Genetics</EM>, 2001.

<P>
</LI>
<LI>R. Wheeler and R. Hughey,
<A NAME="tex2html13"
  HREF="http://www.cse.ucsc.edu/research/kestrel/papers/bio99102.ps">Optimizing Reduced Space Sequence
Analysis,</A>
<EM>BioInformatics</EM>, 16(12): 1082-1090, 2000.

<P>
</LI>
<LI>K. Karplus, C. Barrett, M. Cline, M. Diekhans, L. Grate, R. Hughey,
``Predicting Protein Structure using Only Sequence Information,''  
<EM>Proteins: Structure, Function, and Genetics,</EM>  
Supplement&nbsp;3:121-125, 1999. <A NAME="karplus99"></A> 

<P>
</LI>
<LI>J. Park, K. Karplus, C. Barrett, R. Hughey, D. Haussler, T. Hubbard,
and C. Chothia,
<A NAME="tex2html14"
  HREF="http://www.mrc-lmb.cam.ac.uk/genomes/jong/assess_paper/assess_paperNov.html">Sequence Comparisons Using Multiple Sequences Detect
Twice as many Remote Homologues as Pairwise Methods,</A>
<EM>Journal of
Molecular Biology,</EM> 284(4):1201-1210, 1998.

<P>
</LI>
<LI>R.&nbsp;Karchin and R.&nbsp;Hughey.
<A NAME="tex2html15"
  HREF="http://www.cse.ucsc.edu/research/compbio/html_format_papers/weighting">Weighting hidden Markov models for maximum discrimination</A>
<EM>Bioinformatics,</EM> 14(9):772-782, 1998.

<P>
</LI>
<LI>C.&nbsp;Tarnas and R.&nbsp;Hughey.
<A NAME="tex2html16"
  HREF="http://www.cse.ucsc.edu/research/compbio/papers/samspace.ps.Z">Reduced space hidden Markov model training.</A>
<EM>Bioinformatics,</EM> 14(5):401-406, 1998.

<P>
</LI>
<LI>K.&nbsp;Karplus, Kimmen Sj&#246;lander, C.&nbsp;Barrett, M.&nbsp;Cline,
 D.&nbsp;Haussler, R.&nbsp;Hughey,  L.&nbsp;Holm, and C.&nbsp;Sander,
  ``Predicting protein structure using hidden Markov
  models,'' <EM>Proteins: Structure, Function, and Genetics</EM>,
  Supplement 1, 1997.

<P>
</LI>
<LI>K. Sjolander,
K. Karplus, M. Brown, R. Hughey, A. Krogh, I.S. Mian, and D. Haussler.
<A NAME="tex2html17"
  HREF="http://www.cse.ucsc.edu/research/compbio/dirichlets/index.html">Dirichlet Mixtures: A Method for Improving
Detection of Weak but Significant Protein Sequence Homology.</A>
CABIOS
12(4):327-345, 1996. <A NAME="sj96"></A>
<P>
</LI>
<LI>C.&nbsp;Barrett and R.&nbsp;Hughey and K.&nbsp;Karplus.
<A NAME="tex2html18"
  HREF="http://www.cse.ucsc.edu/research/compbio/papers/nullmod.ps.Z">Scoring Hidden Markov Models.</A>
CABIOS 13(2):191-199, 1997.

<P>
</LI>
<LI>J. A. Grice, R. Hughey, and D. Speck.
<A NAME="tex2html19"
  HREF="http://www.cse.ucsc.edu/research/kestrel/traceback.ps">Reduced space sequence alignment.</A>
CABIOS 13(1):45-53, 1997.

<P>
</LI>
<LI>D.&nbsp;Haussler, A.&nbsp;Krogh, I.&nbsp;S. Mian, and K.&nbsp;Sj&#246;lander.
Protein modeling using hidden Markov models: Analysis of globins.
In <EM>Proceedings of the Hawaii International Conference on System
  Sciences</EM>, volume&nbsp;1, pages 792-802, Los Alamitos, CA, 1993. IEEE Computer
  Society Press.

<P>
</LI>
<LI>A.&nbsp;Krogh, I.&nbsp;S. Mian, and D.&nbsp;Haussler.
A hidden Markov model that finds genes in <EM>E. coli</EM> DNA.
<EM>Nucleic Acids Research</EM>, 1994.

<P>
</LI>
<LI>R.&nbsp;Hughey and A.&nbsp;Krogh.
<A NAME="tex2html20"
  HREF="http://www.cse.ucsc.edu/research/compbio/papers/sam_doc.ps.Z">SAM: Sequence alignment and modeling software system.</A>
Technical Report UCSC-CRL-96-22, University of California,
Santa Cruz, CA, September 1996. 

<P>
</LI>
<LI>K. Karplus. Regularizers for Estimating Distributions of
Amino Acids from Small Samples.  
Technical Report UCSC-CRL-95-11, University of California,
Santa Cruz, CA, 30 March 1995. 

<P>
</LI>
</UL>
We would appreciate references to the first  article in all
work that cites or uses the SAM system, the second for all work that
cites or uses the SAM-T99 method, and the third article in work
that cites or uses HMM methods similar to SAM.

<P>
Because the software is an active research tool, there are a vast
selection of options, many of which have, through experimental study,
been set to reasonable defaults.

<P>
The SAM software and documentation copyright is held by the Regents of
the University of California.  A signed license is required to obtain
a copy of SAM, downloadable from the SAM WWW site,
with no fee for educational research use.  If you have
suggestions for enhancements, new ways of using SAM, or other
comments, please contact us.

<P>
SAM incorporates the readseq package by D. G. Gilbert, who allows it
to be freely copied and used.  The <TT>hmmedit</TT> and <TT>sae</TT>
programs use ACEdb by Richard Durbin and Jean Thierry-Mieg.  The
source code for <TT>hmmedit</TT> and <TT>sae</TT> is available from
<A NAME="tex2html23"
  HREF="ftp://ftp.cse.ucsc.edu/pub/protein/hmmeditsaesrc.tar.Z">ftp://ftp.cse.ucsc.edu/pub/protein/hmmeditsaesrc.tar.Z</A>.

<P>
SAM includes the BLAST matrix library for use with SAM's Smith and
Waterman implementation.  This work of the U. S. Government is
available at
<A NAME="tex2html24"
  HREF="http://www.ncbi.nlm.nig.gov/BLAST">http://www.ncbi.nlm.nig.gov/BLAST</A>
<P>
To be informed of future releases, please send your e-mail
address to <code>sam-info@cse.ucsc.edu</code> for addition to our mailing
list.  Please also use this address for any questions or comments you
may have.

<P>
You will also find Sean Eddy's system,
<A NAME="tex2html25"
  HREF="http://genome.wustl.edu/eddy/hmm.html">HMMER</A>
, to be of interest.

<P>
Martin Madera and Julian Gough have written a perl converter between
SAM and HMMer 2.0 formats. (The SAM programs only work with HMMer
1.7.)
<DIV ALIGN="CENTER">
<A NAME="tex2html21"
  HREF="http://www.mrc-lmb.cam.ac.uk/genomes/julian/convert/convert.html">http://www.mrc-lmb.cam.ac.uk/genomes/julian/convert/convert.html</A>

</DIV>

<P>

<H2><A NAME="SECTION00021000000000000000">
1.1 Acknowledgments</A>
</H2>

<P>
We thank I. Saira Mian and Finn Drablos for their important
early evaluations of the system. Finally, we thank the entire UCSC
Computational Biology Group (now forming the core of a Center for
Biomolecular Engineering), led by David Haussler, who got this whole
thing started. This 
work was supported in part by NSF grants CDA-9115268, IRI-9123692, BIR
94-08579, MIP-9423985, DBI-9808007, and EIA-9905322; DOE grants
94-12-048216 and DE-FG03-99ER62849; ONR grant N00014-91-J-1162; NIH
grant GM17129; a grant from the Danish Natural Science Research
Council; and a gift from Digital Equipment Corporation.

<P>
<HR>
<!--Navigation Panel-->
<A NAME="tex2html285"
  HREF="node3.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/next.png"></A> 
<A NAME="tex2html281"
  HREF="sam_doc.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/up.png"></A> 
<A NAME="tex2html275"
  HREF="node1.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/prev.png"></A> 
<A NAME="tex2html283"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="http://www.cse.ucsc.edu/research/compbio/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html286"
  HREF="node3.html">2 New to this</A>
<B> Up:</B> <A NAME="tex2html282"
  HREF="sam_doc.html">SAM (Sequence Alignment and</A>
<B> Previous:</B> <A NAME="tex2html276"
  HREF="node1.html">Contents</A>
 &nbsp <B>  <A NAME="tex2html284"
  HREF="node1.html">Contents</A></B> 
<!--End of Navigation Panel-->
<ADDRESS>
<a href=http://www.cse.ucsc.edu/research/compbio/sam.html>SAM</a><BR><a href=mailto:sam-info@cse.ucsc.edu>sam-info@cse.ucsc.edu</a><BR><A HREF=http://www.cse.ucsc.edu/research/compbio>UCSC Computational Biology Group</A>
</ADDRESS>
</BODY>
</HTML>
